{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['.DS_Store',\n",
       " 'ASVs_filtered.fasta',\n",
       " 'ASVs_filtered_counts.tsv',\n",
       " 'ASVs_filtered_ref_full.tree',\n",
       " 'ASVs_taxonomy.txt',\n",
       " 'metadata_gyrb_amp_meta_passing_samples.txt',\n",
       " 'moeller_codiv_Bacteroidaceae.fna',\n",
       " 'moeller_codiv_HRclades.txt',\n",
       " 'moeller_codiv_lin_Bt1.tree',\n",
       " 'moeller_codiv_lin_Bt2.tree',\n",
       " 'moeller_codiv_lin_Bt3.tree',\n",
       " 'ref_gyrb_gtdbtk']"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from ete3 import Tree\n",
    "from collections import Counter\n",
    "\n",
    "os.getcwd()\n",
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/results/gyrb/')\n",
    "os.listdir('inputs/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#inputs\n",
    "metadata_file = 'inputs/metadata_gyrb_amp_meta_passing_samples.txt'\n",
    "tax_table_file = 'inputs/ASVs_taxonomy.txt'\n",
    "asv_table_file = 'inputs/ASVs_filtered_counts.tsv'\n",
    "asv_fasta_file = 'inputs/ASVs_filtered.fasta'\n",
    "tree_file = 'inputs/ASVs_filtered_ref_full.tree'\n",
    "full_tree = Tree(tree_file, format=0)\n",
    "moeller_codiv_fasta = 'inputs/codiv_Bacteroidaceae.fna'\n",
    "pident_cutoff = 95"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Blastn moeller co-div seqs against ASVs and filter ASVs matchin X% identity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 07/28/2020 11:23:45\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb/analyses/codiv_moeller_ASVs/codiv_Bacteroidaceae.fna\n",
      "New DB title:  analyses/codiv_moeller_ASVs/codiv_Bacteroidaceae.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb/analyses/codiv_moeller_ASVs/codiv_Bacteroidaceae.fna\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 208 sequences in 0.00704718 seconds.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: analyses/codiv_moeller_ASVs: File exists\n",
      "cp: inputs/codiv_Bacteroidaceae.fna: No such file or directory\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir analyses/codiv_moeller_ASVs\n",
    "cp inputs/codiv_Bacteroidaceae.fna analyses/codiv_moeller_ASVs/codiv_Bacteroidaceae.fna\n",
    "\n",
    "#make blastdb\n",
    "makeblastdb -in analyses/codiv_moeller_ASVs/codiv_Bacteroidaceae.fna -dbtype nucl\n",
    "\n",
    "#blast moeller co-div seqs\n",
    "blastn -query inputs/ASVs_filtered.fasta -db analyses/codiv_moeller_ASVs/codiv_Bacteroidaceae.fna \\\n",
    "-outfmt \"7 qseqid salltitles sseqid pident length qlen evalue\" \\\n",
    "-out analyses/codiv_moeller_ASVs/codiv_blastout_ASVs.txt \\\n",
    "-max_target_seqs 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Subset the matching ASVs "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "314 ASVs matching co-div clades with greater than 95 percent identity\n",
      "      ASV                codiv_clade_seq       codiv_clade lineage\n",
      "0   ASV_1  Human6700665 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "1  ASV_13  Human7859459 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "2  ASV_34  Human9825266 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "3  ASV_37  Human8829943 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "4  ASV_73  Human9825266 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "Bt2_clade1_bonobo     90\n",
      "Bt3_clade1_human      85\n",
      "Bt2_clade2_bonobo     33\n",
      "Bt1_clade1_bonobo     23\n",
      "Bt2_clade1_chimp      20\n",
      "Bt3_clade1_chimp      17\n",
      "Bt3_clade1_bonobo     15\n",
      "Bt1_clade1_gorilla    10\n",
      "Bt1_clade1_chimp       8\n",
      "Bt2_clade2_gorilla     7\n",
      "Bt2_clade2_chimp       6\n",
      "Name: codiv_clade, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#filter blast output to top hit per ASV, remove hits not pass thresholds\n",
    "blast_res = pd.read_csv('analyses/codiv_moeller_ASVs/codiv_blastout_ASVs.txt',sep='\\t',comment='#',header=None)\n",
    "blast_res.columns = ['ASV','codiv_clade_seq','sseqid','pident','length','qlen','evalue']\n",
    "blast_res = blast_res.groupby('ASV').head(1).reset_index(drop=True) #get top hit\n",
    "blast_res_pass = blast_res[blast_res['pident']>pident_cutoff].reset_index(drop=True) #QC filter\n",
    "blast_res_pass = blast_res_pass[blast_res_pass['length']>200].reset_index(drop=True) #QC filter\n",
    "print(len(blast_res_pass),'ASVs matching co-div clades with greater than',pident_cutoff,'percent identity')\n",
    "blast_res_pass['codiv_clade'] = blast_res_pass['codiv_clade_seq'].apply(lambda x: x.split(' ')[1])\n",
    "blast_res_pass['lineage'] = blast_res_pass['codiv_clade'].apply(lambda x: x.split('_')[0]) \n",
    "blast_res_pass = blast_res_pass[['ASV','codiv_clade_seq', 'codiv_clade','lineage']]\n",
    "print(blast_res_pass.head())\n",
    "print(blast_res_pass['codiv_clade'].value_counts())\n",
    "#output hits passing that will be used to determine the mcra of the lineage\n",
    "blast_res_pass.to_csv('analyses/codiv_moeller_ASVs/codiv_clades_ASVs.txt',sep='\\t',index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "f__Bacteroidaceae.g__Bacteroides_B\n"
     ]
    }
   ],
   "source": [
    "#generates three dictionaries where keys are ASVs, and values are metadata\n",
    "#all the samples with a given ASV\n",
    "#all the sample types with a given ASV per metadata info\n",
    "#taxonomic info, family and genus\n",
    "metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "sample_type_dict = dict(zip(metadata['X.SampleID'], metadata['Description']))\n",
    "\n",
    "ASV_sampleName_dict = {}  \n",
    "asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "for ASV,row in asv_table.iterrows():\n",
    "    sampleNames = list(asv_table.columns[row>0])\n",
    "    ASV_sampleName_dict[ASV] = sampleNames"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  cladeName HR_clade  HR_sampleNum  \\\n",
      "0   clade_1    mixed            75   \n",
      "1   clade_2       HR            18   \n",
      "2   clade_3       HR            35   \n",
      "3   clade_4       HR            22   \n",
      "4   clade_5       HR             7   \n",
      "\n",
      "                                                 ASV  ASVsNum  sampleNum  \\\n",
      "0  [ASV_4444, ASV_1277, ASV_6056, ASV_6020, ASV_4...       41         75   \n",
      "1  [ASV_4444, ASV_1277, ASV_6056, ASV_6020, ASV_4...       10         18   \n",
      "2  [ASV_1090, ASV_2134, ASV_482, ASV_5974, ASV_64...        8         35   \n",
      "3  [ASV_4623, ASV_1643, ASV_1616, ASV_4160, ASV_4...       23         22   \n",
      "4                               [ASV_1090, ASV_2134]        2          7   \n",
      "\n",
      "                               sampleTypes  \n",
      "0  [wild_gorilla, wild_bonobo, wild_chimp]  \n",
      "1                           [wild_gorilla]  \n",
      "2                             [wild_chimp]  \n",
      "3                            [wild_bonobo]  \n",
      "4                             [wild_chimp]  \n",
      "3 HR clades\n",
      "41 HR ASVs\n",
      "  cladeName HR_clade  HR_sampleNum  \\\n",
      "0   clade_1    mixed            75   \n",
      "\n",
      "                                                 ASV  ASVsNum  sampleNum  \\\n",
      "0  [ASV_4444, ASV_1277, ASV_6056, ASV_6020, ASV_4...       41         75   \n",
      "\n",
      "                               sampleTypes  \n",
      "0  [wild_gorilla, wild_bonobo, wild_chimp]  \n",
      "Empty DataFrame\n",
      "Columns: [cladeName, HR_clade, HR_sampleNum, ASV, ASVsNum, sampleNum, sampleTypes]\n",
      "Index: [] mixed clades\n",
      "0 mixed clade ASVs\n",
      "0 leftover ASVs\n"
     ]
    }
   ],
   "source": [
    "from pandas.core.common import flatten\n",
    "Bt1_tree = full_tree.get_common_ancestor(list(Bt1_lineage_ASVs['ASV']))\n",
    "\n",
    "def is_HR(sampleNames):\n",
    "    \"\"\"given a list of sample names uses sample type dictionary to determine how many sample types are present\n",
    "    designates clades/nodes as host restricted = 1 sample type or mixed = multiple sample types.\n",
    "    Captive sample types are not considered so some clades will have a 0 sample type length and they can fall within\n",
    "    host-restricted clades or mixed clades or neither\"\"\"\n",
    "    sampleTypes = [sample_type_dict[name] for name in sampleNames]\n",
    "    sampleTypes = [x.replace('non_western_','').replace('western_','') for x in sampleTypes]\n",
    "    neutral_sampleTypes = ['captive_gorilla','captive_bonobo','captive_chimp','captive_orangutan']\n",
    "    select_sampleTypes = [x for x in sampleTypes if x not in neutral_sampleTypes]\n",
    "    unique_sampleTypes = list(set(select_sampleTypes))\n",
    "    \n",
    "    if len(unique_sampleTypes) == 0:\n",
    "        return(unique_sampleTypes,'captive',len(sampleTypes))\n",
    "    if len(unique_sampleTypes)  == 1: #identifies host-restricted clades\n",
    "        HR_clade = select_sampleTypes[0]\n",
    "        HR_sample_num = len([t for t in sampleTypes if t == HR_clade])\n",
    "        return(unique_sampleTypes,'HR',HR_sample_num) #doesn't count captive ape samples\n",
    "    if len(unique_sampleTypes)  > 1:    \n",
    "        return(unique_sampleTypes,'mixed',len(sampleTypes))\n",
    "\n",
    "def search_clades(tree, samples_cutoff, BS_support):\n",
    "    \"\"\"Finds nodes with at least 50% BS support containing ASVs only found in a single wild ape species\n",
    "    ie wild_gorilla, wild_chimp or wild_bonobo, cpat\"\"\"\n",
    "    HR_clades_passing_thresholds = []\n",
    "    counter = 1\n",
    "    for n in tree.traverse():\n",
    "        if n.support > float(BS_support): #makes sure Bootstrap support is over threshold\n",
    "            ASVs = [leaf.name for leaf in n.iter_leaves() if 'ASV' in leaf.name]\n",
    "            ASVsNum = len(ASVs)\n",
    "            sampleNames = list(set(list(flatten([ASV_sampleName_dict[ASV] for ASV in ASVs]))))\n",
    "            \n",
    "            sampleNum = len(sampleNames)\n",
    "            sampleTypes,HR_clade,HR_sampleNum = is_HR(sampleNames)\n",
    "            if HR_sampleNum > samples_cutoff:\n",
    "                cladeName = 'clade_'+str(counter)\n",
    "                counter+=1\n",
    "                clade = [cladeName,HR_clade,HR_sampleNum,ASVs,ASVsNum,sampleNum,sampleTypes]\n",
    "                HR_clades_passing_thresholds.append(clade)  \n",
    "    clades = pd.DataFrame(HR_clades_passing_thresholds, columns = \n",
    "                          ['cladeName','HR_clade','HR_sampleNum',\n",
    "                           'ASV','ASVsNum','sampleNum','sampleTypes'])\n",
    "    return(clades)\n",
    "\n",
    "#initial search contains all sub clades within the largest host restricted clade\n",
    "clades_Bt1 = search_clades(Bt1_tree,samples_cutoff=5,BS_support=.5) \n",
    "print(clades_Bt1.head())\n",
    "\n",
    "CPclades_Bt1 = clades_Bt1[clades_Bt1['HR_clade']=='captive']\n",
    "#print(len(HRclades_Bt1),len(MXclades_Bt1),len(CPclades_Bt1))\n",
    "\n",
    "def eliminate_redundant_clades(search_clades_res):\n",
    "    \"\"\"sorts clades and returns the largest non overlapping clade \"\"\"\n",
    "    df = search_clades_res.sort_values('ASVsNum',ascending=False) #start with the largest clades first\n",
    "    NRclades = []\n",
    "    NR_ASVs = []\n",
    "    for index, row in df.iterrows():\n",
    "        firstASV = row['ASV'][0]    \n",
    "        if firstASV not in NR_ASVs:\n",
    "            NR_ASVs = NR_ASVs + row['ASV']\n",
    "            NRclades.append(row['cladeName']) \n",
    "            #print(row['cladeName'])\n",
    "    res = df[df['cladeName'].isin(NRclades)]   \n",
    "    return(res)\n",
    "\n",
    "HRclades_Bt1 = clades_Bt1[clades_Bt1['HR_clade']=='HR']    \n",
    "HRclades_Bt1 = eliminate_redundant_clades(HRclades_Bt1)\n",
    "HRclades_Bt1.to_csv('analyses/codiv_moeller_ASVs/HRclades_Bt1.txt',sep='\\t',index=False)\n",
    "print(len(HRclades_Bt1),'HR clades')\n",
    "\n",
    "def expand_clades(clades_df):\n",
    "    ASV = clades_df.apply(lambda x: pd.Series(x['ASV']),axis=1).stack().reset_index(level=1, drop=True)\n",
    "    ASV.name = 'ASV'\n",
    "    clades_ASVs_df = clades_df.drop('ASV', axis=1).join(ASV)\n",
    "    return(clades_ASVs_df)\n",
    "\n",
    "HRclades_Bt1_ASVs = expand_clades(HRclades_Bt1)\n",
    "print(len(HRclades_Bt1_ASVs),'HR ASVs')\n",
    "HRclades_Bt1_ASVs.to_csv('analyses/codiv_moeller_ASVs/HRclades_Bt1_ASVs.txt',sep='\\t',index=False)\n",
    "\n",
    "\n",
    "\n",
    "MXclades_Bt1 = clades_Bt1[clades_Bt1['HR_clade']=='mixed']\n",
    "MXclades_Bt1 = MXclades_Bt1[MXclades_Bt1['ASV'].apply(lambda cladeASVs: len(\n",
    "                                                                        list(set(cladeASVs) & set(list(HRclades_Bt1_ASVs['ASV'])))\n",
    "                                                                            ) == 0 )]\n",
    "                                                                                \n",
    "MXclades_Bt1_ASVs = expand_clades(MXclades_Bt1)\n",
    "print(len(MXclades_Bt1_ASVs),'mixed clade ASVs')\n",
    "\n",
    "\n",
    "leftover_ASVs = len(Bt1_lineage_ASVs) - len(HRclades_Bt1_ASVs) - len(MXclades_Bt1_ASVs)\n",
    "print(leftover_ASVs,'leftover ASVs')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 206,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "360 total ASVs\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/usr/local/anaconda3/envs/ete/lib/python3.6/site-packages/ipykernel_launcher.py:7: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
      "  import sys\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "337 HR clade ASVs\n",
      "4 mixed clade ASVs\n",
      "19 leftover ASVs\n",
      "17 1\n"
     ]
    }
   ],
   "source": [
    "full_tree.get_common_ancestor(Bt2_ASVs)\n",
    "\n",
    "Bt2_ASVs = list(Bt2_lineage_ASVs['ASV'])\n",
    "print(len(Bt2_ASVs),'total ASVs')\n",
    "Bt2_tree = full_tree.get_common_ancestor(Bt2_ASVs)\n",
    "\n",
    "clades_Bt2 = search_clades(Bt2_tree,samples_cutoff=5,BS_support=.5) \n",
    "\n",
    "#Define host-restricted \n",
    "HRclades_Bt2 = clades_Bt2[clades_Bt2['HR_clade']!='mixed'][clades_Bt2['HR_clade']!='captive']     \n",
    "HRclades_Bt2 = eliminate_redundant_clades(HRclades_Bt2)\n",
    "HRclades_Bt2.to_csv('analyses/codiv_moeller_ASVs/Bt2_HRclades.txt',sep='\\t',index=False)\n",
    "HRclades_Bt2_ASVs = expand_clades(HRclades_Bt2)\n",
    "HRclades_Bt2_ASVs.to_csv('analyses/codiv_moeller_ASVs/Bt2_HRclades_ASVs.txt',sep='\\t',index=False)\n",
    "print(len(HRclades_Bt2_ASVs),'HR clade ASVs')\n",
    "\n",
    "MXclades_Bt2 = clades_Bt2[clades_Bt2['HR_clade']=='mixed']\n",
    "MXclades_Bt2 = MXclades_Bt2[MXclades_Bt2['ASV'].apply(lambda cladeASVs: is_unique_MX(list(HRclades_Bt2_ASVs['ASV']),cladeASVs))]\n",
    "MXclades_Bt2 = eliminate_redundant_clades(MXclades_Bt2)\n",
    "MXclades_Bt2_ASVs = expand_clades(MXclades_Bt2)\n",
    "print(len(MXclades_Bt2_ASVs),'mixed clade ASVs')\n",
    "\n",
    "leftover_ASVs = len(Bt2_ASVs) - len(HRclades_Bt2_ASVs) - len(MXclades_Bt2_ASVs)\n",
    "print(leftover_ASVs,'leftover ASVs')\n",
    "\n",
    "print(len(HRclades_Bt2),len(MXclades_Bt2))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7148\n",
      "6246 ASVs in 302 HR clades\n",
      "183 ASVs in 51 mixed clades\n",
      "9 ASVs in 6 captive clades\n",
      "710 leftover ASVs\n",
      "{'ASV_10018', 'ASV_10157', 'ASV_7435', 'ASV_6318', 'ASV_6277', 'ASV_9173', 'ASV_7692', 'ASV_6312', 'ASV_9297', 'ASV_1404', 'ASV_10553', 'ASV_3172', 'ASV_4568', 'ASV_4505', 'ASV_4001', 'ASV_8866', 'ASV_9377', 'ASV_10335', 'ASV_7717', 'ASV_7691', 'ASV_2876', 'ASV_6186', 'ASV_10177', 'ASV_2214', 'ASV_8710', 'ASV_6550', 'ASV_6305', 'ASV_6043', 'ASV_10003', 'ASV_7664', 'ASV_7270', 'ASV_10325', 'ASV_9830', 'ASV_9859', 'ASV_9571', 'ASV_9570', 'ASV_5715', 'ASV_9177', 'ASV_4316', 'ASV_10182', 'ASV_7118', 'ASV_9335', 'ASV_7974', 'ASV_10127', 'ASV_9314', 'ASV_1696', 'ASV_9858', 'ASV_8214', 'ASV_2465', 'ASV_2298', 'ASV_10316', 'ASV_8780', 'ASV_10027', 'ASV_9572', 'ASV_9851', 'ASV_6570', 'ASV_9313', 'ASV_9772', 'ASV_9295', 'ASV_6266', 'ASV_1919', 'ASV_2880', 'ASV_10502', 'ASV_1617', 'ASV_6307', 'ASV_10200', 'ASV_8747', 'ASV_9630', 'ASV_6547', 'ASV_7674', 'ASV_9968', 'ASV_5297', 'ASV_4030', 'ASV_2030', 'ASV_7696', 'ASV_10190', 'ASV_2402', 'ASV_10117', 'ASV_10191', 'ASV_8743', 'ASV_5333', 'ASV_6296', 'ASV_4071', 'ASV_10328', 'ASV_10042', 'ASV_3163', 'ASV_6044', 'ASV_5523', 'ASV_9848', 'ASV_9965', 'ASV_2777', 'ASV_10320', 'ASV_6278', 'ASV_9832', 'ASV_5682', 'ASV_10514', 'ASV_3167', 'ASV_9641', 'ASV_4972', 'ASV_9120', 'ASV_8683', 'ASV_8715', 'ASV_4873', 'ASV_2007', 'ASV_10119', 'ASV_9831', 'ASV_6327', 'ASV_9186', 'ASV_9337', 'ASV_9629', 'ASV_6250', 'ASV_10125', 'ASV_9359', 'ASV_10161', 'ASV_2094', 'ASV_7953', 'ASV_8062', 'ASV_2931', 'ASV_2392', 'ASV_6811', 'ASV_9826', 'ASV_9301', 'ASV_10026', 'ASV_9288', 'ASV_9143', 'ASV_4085', 'ASV_9364', 'ASV_1112', 'ASV_5969', 'ASV_10163', 'ASV_2236', 'ASV_6565', 'ASV_4547', 'ASV_8595', 'ASV_9360', 'ASV_10134', 'ASV_10162', 'ASV_2762', 'ASV_10522', 'ASV_1406', 'ASV_9700', 'ASV_1992', 'ASV_4713', 'ASV_8704', 'ASV_2266', 'ASV_9310', 'ASV_6606', 'ASV_2932', 'ASV_2359', 'ASV_8719', 'ASV_8361', 'ASV_10302', 'ASV_10156', 'ASV_5981', 'ASV_10053', 'ASV_1705', 'ASV_6850', 'ASV_10013', 'ASV_9291', 'ASV_7082', 'ASV_1503', 'ASV_9293', 'ASV_5075', 'ASV_1627', 'ASV_10145', 'ASV_10108', 'ASV_9518', 'ASV_1539', 'ASV_2202', 'ASV_6303', 'ASV_10304', 'ASV_3341', 'ASV_8597', 'ASV_7622', 'ASV_10314', 'ASV_5725', 'ASV_10137', 'ASV_9141', 'ASV_2437', 'ASV_1307', 'ASV_9548', 'ASV_4208', 'ASV_9123', 'ASV_6228', 'ASV_5246', 'ASV_10112', 'ASV_10024', 'ASV_10180', 'ASV_9854', 'ASV_6790', 'ASV_9993', 'ASV_7132', 'ASV_5763', 'ASV_5998', 'ASV_7423', 'ASV_2296', 'ASV_5595', 'ASV_6195', 'ASV_6196', 'ASV_6319', 'ASV_5676', 'ASV_2116', 'ASV_7954', 'ASV_5772', 'ASV_7694', 'ASV_4217', 'ASV_2401', 'ASV_7653', 'ASV_9839', 'ASV_2882', 'ASV_10114', 'ASV_3205', 'ASV_3239', 'ASV_9833', 'ASV_7120', 'ASV_2793', 'ASV_10128', 'ASV_7930', 'ASV_1254', 'ASV_10598', 'ASV_10165', 'ASV_6328', 'ASV_3166', 'ASV_9886', 'ASV_8934', 'ASV_1522', 'ASV_6314', 'ASV_9170', 'ASV_6203', 'ASV_10201', 'ASV_10144', 'ASV_9958', 'ASV_10499', 'ASV_1540', 'ASV_10509', 'ASV_5011', 'ASV_10109', 'ASV_7126', 'ASV_9307', 'ASV_10318', 'ASV_6083', 'ASV_9829', 'ASV_7650', 'ASV_3134', 'ASV_10115', 'ASV_10011', 'ASV_5740', 'ASV_2590', 'ASV_3343', 'ASV_2635', 'ASV_2246', 'ASV_6695', 'ASV_8359', 'ASV_9842', 'ASV_2358', 'ASV_9849', 'ASV_9855', 'ASV_9643', 'ASV_10595', 'ASV_5975', 'ASV_3237', 'ASV_8702', 'ASV_10524', 'ASV_6191', 'ASV_3109', 'ASV_5039', 'ASV_9850', 'ASV_6315', 'ASV_9837', 'ASV_2418', 'ASV_9615', 'ASV_3165', 'ASV_1829', 'ASV_10254', 'ASV_5995', 'ASV_6160', 'ASV_10357', 'ASV_7071', 'ASV_10142', 'ASV_1193', 'ASV_9961', 'ASV_6109', 'ASV_10367', 'ASV_8123', 'ASV_6209', 'ASV_9278', 'ASV_2167', 'ASV_6136', 'ASV_6311', 'ASV_9834', 'ASV_6552', 'ASV_5750', 'ASV_6563', 'ASV_8355', 'ASV_10317', 'ASV_7659', 'ASV_8870', 'ASV_10111', 'ASV_6047', 'ASV_10493', 'ASV_4848', 'ASV_10154', 'ASV_9294', 'ASV_5476', 'ASV_2969', 'ASV_7116', 'ASV_10710', 'ASV_6281', 'ASV_10135', 'ASV_3995', 'ASV_3219', 'ASV_7901', 'ASV_10131', 'ASV_1496', 'ASV_9632', 'ASV_5507', 'ASV_8618', 'ASV_10349', 'ASV_1335', 'ASV_6171', 'ASV_2621', 'ASV_8736', 'ASV_9620', 'ASV_10130', 'ASV_2037', 'ASV_2663', 'ASV_5675', 'ASV_9955', 'ASV_3161', 'ASV_6285', 'ASV_3047', 'ASV_8708', 'ASV_8868', 'ASV_8006', 'ASV_2868', 'ASV_8732', 'ASV_1947', 'ASV_8709', 'ASV_6545', 'ASV_2936', 'ASV_2928', 'ASV_3982', 'ASV_7623', 'ASV_10002', 'ASV_6298', 'ASV_3164', 'ASV_4632', 'ASV_10146', 'ASV_3202', 'ASV_9108', 'ASV_6326', 'ASV_5693', 'ASV_6313', 'ASV_5447', 'ASV_9843', 'ASV_10116', 'ASV_6330', 'ASV_6011', 'ASV_7663', 'ASV_6276', 'ASV_8754', 'ASV_410', 'ASV_10336', 'ASV_10253', 'ASV_1396', 'ASV_2181', 'ASV_8905', 'ASV_9635', 'ASV_2439', 'ASV_8678', 'ASV_2297', 'ASV_10286', 'ASV_10192', 'ASV_10167', 'ASV_10319', 'ASV_6279', 'ASV_6097', 'ASV_1634', 'ASV_3026', 'ASV_1570', 'ASV_10313', 'ASV_9828', 'ASV_10505', 'ASV_6398', 'ASV_8720', 'ASV_5163', 'ASV_5925', 'ASV_10133', 'ASV_6125', 'ASV_5122', 'ASV_9950', 'ASV_2034', 'ASV_8745', 'ASV_6691', 'ASV_10500', 'ASV_9169', 'ASV_10259', 'ASV_1336', 'ASV_9846', 'ASV_8121', 'ASV_9374', 'ASV_8738', 'ASV_8751', 'ASV_6546', 'ASV_9852', 'ASV_6321', 'ASV_3208', 'ASV_4185', 'ASV_7272', 'ASV_3287', 'ASV_2901', 'ASV_6158', 'ASV_3271', 'ASV_9311', 'ASV_1553', 'ASV_5394', 'ASV_6098', 'ASV_1777', 'ASV_3206', 'ASV_2255', 'ASV_2270', 'ASV_1924', 'ASV_6070', 'ASV_7672', 'ASV_9136', 'ASV_10260', 'ASV_8362', 'ASV_4703', 'ASV_7078', 'ASV_6479', 'ASV_4339', 'ASV_10118', 'ASV_8740', 'ASV_4470', 'ASV_5801', 'ASV_5990', 'ASV_3190', 'ASV_8750', 'ASV_8877', 'ASV_8357', 'ASV_6309', 'ASV_10150', 'ASV_5104', 'ASV_4516', 'ASV_8621', 'ASV_8744', 'ASV_6571', 'ASV_10139', 'ASV_10490', 'ASV_10132', 'ASV_7975', 'ASV_8218', 'ASV_10346', 'ASV_8365', 'ASV_2467', 'ASV_9989', 'ASV_6151', 'ASV_7081', 'ASV_2484', 'ASV_10322', 'ASV_5767', 'ASV_10312', 'ASV_10056', 'ASV_8867', 'ASV_8692', 'ASV_7080', 'ASV_3691', 'ASV_5953', 'ASV_1377', 'ASV_8739', 'ASV_7915', 'ASV_8363', 'ASV_6477', 'ASV_4183', 'ASV_10296', 'ASV_9827', 'ASV_5111', 'ASV_6282', 'ASV_8396', 'ASV_10004', 'ASV_7703', 'ASV_9137', 'ASV_4651', 'ASV_8781', 'ASV_6537', 'ASV_9176', 'ASV_6481', 'ASV_1538', 'ASV_9988', 'ASV_7693', 'ASV_6478', 'ASV_7140', 'ASV_7117', 'ASV_5115', 'ASV_7701', 'ASV_6248', 'ASV_10324', 'ASV_10202', 'ASV_8735', 'ASV_6329', 'ASV_8215', 'ASV_3294', 'ASV_10527', 'ASV_8727', 'ASV_10126', 'ASV_3270', 'ASV_4292', 'ASV_3344', 'ASV_9969', 'ASV_6046', 'ASV_6749', 'ASV_9315', 'ASV_7046', 'ASV_8760', 'ASV_9856', 'ASV_5739', 'ASV_1552', 'ASV_1946', 'ASV_4455', 'ASV_1785', 'ASV_3243', 'ASV_4536', 'ASV_10523', 'ASV_3642', 'ASV_9857', 'ASV_2705', 'ASV_7981', 'ASV_10129', 'ASV_2594', 'ASV_10043', 'ASV_9823', 'ASV_7273', 'ASV_6183', 'ASV_6199', 'ASV_8707', 'ASV_9637', 'ASV_10176', 'ASV_10020', 'ASV_4211', 'ASV_6295', 'ASV_8364', 'ASV_3290', 'ASV_9308', 'ASV_5095', 'ASV_1505', 'ASV_8755', 'ASV_6021', 'ASV_2601', 'ASV_5261', 'ASV_8596', 'ASV_7130', 'ASV_6306', 'ASV_3286', 'ASV_6648', 'ASV_10323', 'ASV_5494', 'ASV_4862', 'ASV_5982', 'ASV_10138', 'ASV_1996', 'ASV_7673', 'ASV_6476', 'ASV_9621', 'ASV_6566', 'ASV_10531', 'ASV_3160', 'ASV_5633', 'ASV_9809', 'ASV_7125', 'ASV_3235', 'ASV_10288', 'ASV_6569', 'ASV_8395', 'ASV_9618', 'ASV_1662', 'ASV_6301', 'ASV_6055', 'ASV_10555', 'ASV_8402', 'ASV_1647', 'ASV_10303', 'ASV_1998', 'ASV_5923', 'ASV_3207', 'ASV_8101', 'ASV_8674', 'ASV_9845', 'ASV_1086', 'ASV_9391', 'ASV_8716', 'ASV_9627', 'ASV_5325', 'ASV_9425', 'ASV_9835', 'ASV_1262', 'ASV_9636', 'ASV_9734', 'ASV_10356', 'ASV_4781', 'ASV_3162', 'ASV_8352', 'ASV_10287', 'ASV_8403', 'ASV_6573', 'ASV_3201', 'ASV_10497', 'ASV_9172', 'ASV_10297', 'ASV_10256', 'ASV_7434', 'ASV_3120', 'ASV_1948', 'ASV_6023', 'ASV_10014', 'ASV_7408', 'ASV_2003', 'ASV_2878', 'ASV_8746', 'ASV_8733', 'ASV_10513', 'ASV_5980', 'ASV_8952', 'ASV_9622', 'ASV_6555', 'ASV_8737', 'ASV_8775', 'ASV_5812', 'ASV_6169', 'ASV_5205', 'ASV_6251', 'ASV_8752', 'ASV_2406', 'ASV_10023', 'ASV_4454', 'ASV_9174', 'ASV_8063', 'ASV_9449', 'ASV_8762', 'ASV_2350', 'ASV_1082', 'ASV_10148', 'ASV_6810', 'ASV_10504', 'ASV_10136', 'ASV_6273', 'ASV_6383', 'ASV_9581', 'ASV_10347', 'ASV_7916', 'ASV_6325', 'ASV_10507', 'ASV_9847', 'ASV_4647', 'ASV_1467', 'ASV_1355', 'ASV_7129', 'ASV_6240', 'ASV_6551', 'ASV_3117', 'ASV_9601', 'ASV_2407', 'ASV_9992', 'ASV_7141', 'ASV_6304', 'ASV_9334', 'ASV_6308', 'ASV_8765', 'ASV_10216', 'ASV_2260', 'ASV_10152', 'ASV_4221', 'ASV_9971', 'ASV_8356', 'ASV_8779', 'ASV_7070', 'ASV_6270', 'ASV_1256', 'ASV_6564', 'ASV_9638', 'ASV_6287', 'ASV_8764', 'ASV_8766', 'ASV_9574', 'ASV_8124', 'ASV_8360', 'ASV_5086', 'ASV_7683', 'ASV_6475', 'ASV_10295', 'ASV_10113', 'ASV_6320', 'ASV_9185', 'ASV_8763', 'ASV_2292', 'ASV_5977', 'ASV_10329', 'ASV_9357', 'ASV_3842', 'ASV_8878', 'ASV_6241', 'ASV_8358', 'ASV_7111', 'ASV_6480', 'ASV_2879', 'ASV_1334', 'ASV_4416', 'ASV_6166', 'ASV_4912', 'ASV_8689', 'ASV_9952'}\n",
      "       cladeName HR_clade  HR_sampleNum                                   ASV  \\\n",
      "175    clade_176  captive            32  [ASV_300, ASV_411, ASV_727, ASV_365]   \n",
      "398    clade_399  captive             6                             [ASV_935]   \n",
      "1879  clade_1880  captive             6                             [ASV_931]   \n",
      "1881  clade_1882  captive             8                             [ASV_725]   \n",
      "2064  clade_2065  captive             8                             [ASV_726]   \n",
      "\n",
      "      ASVsNum  sampleNum sampleTypes  \n",
      "175         4         32          []  \n",
      "398         1          6          []  \n",
      "1879        1          6          []  \n",
      "1881        1          8          []  \n",
      "2064        1          8          []  \n"
     ]
    }
   ],
   "source": [
    "clades_fulltree = search_clades(full_tree,samples_cutoff=5,BS_support=.5) \n",
    "fulltree_ASVs = [leaf.name for leaf in full_tree.get_leaves() if 'ASV' in leaf.name]\n",
    "print(len(fulltree_ASVs))\n",
    "HRclades_fulltree = clades_fulltree[clades_fulltree['HR_clade']=='HR']     \n",
    "HRclades_fulltree = eliminate_redundant_clades(HRclades_fulltree)\n",
    "HRclades_fulltree_ASVs = expand_clades(HRclades_fulltree)\n",
    "HR_ASVs = [asv for asv in HRclades_fulltree_ASVs['ASV'] if 'ASV' in asv]\n",
    "print(len(HR_ASVs),'ASVs in',len(HRclades_fulltree),'HR clades')\n",
    "\n",
    "MXclades_fulltree = clades_fulltree[clades_fulltree['HR_clade']=='mixed']\n",
    "#make sure mixed clades do not include HR clades\n",
    "MXclades_fulltree = MXclades_fulltree[MXclades_fulltree['ASV'].apply(lambda cladeASVs: len(\n",
    "                                                                        list(set(cladeASVs) & set(list(HRclades_fulltree_ASVs['ASV'])))\n",
    "                                                                            ) == 0 )]\n",
    "MXclades_fulltree = eliminate_redundant_clades(MXclades_fulltree)\n",
    "MXclades_fulltree_ASVs = expand_clades(MXclades_fulltree)\n",
    "MX_ASVs = [asv for asv in MXclades_fulltree_ASVs['ASV'] if 'ASV' in asv]\n",
    "print(len(MX_ASVs),'ASVs in',len(MXclades_fulltree),'mixed clades')\n",
    "\n",
    "CPclades_fulltree = clades_fulltree[clades_fulltree['HR_clade']=='captive']\n",
    "CPclades_fulltree = CPclades_fulltree[CPclades_fulltree['ASV'].apply(lambda cladeASVs: len(\n",
    "                                                                        list(set(cladeASVs) & set(list(HRclades_fulltree_ASVs['ASV'])))\n",
    "                                                                            ) == 0 )]\n",
    "CPclades_fulltree = CPclades_fulltree[CPclades_fulltree['ASV'].apply(lambda cladeASVs: len(\n",
    "                                                                        list(set(cladeASVs) & set(list(MXclades_fulltree_ASVs['ASV'])))\n",
    "                                                                            ) == 0 )]\n",
    "CPclades_fulltree = eliminate_redundant_clades(CPclades_fulltree)\n",
    "CPclades_fulltree_ASVs = expand_clades(CPclades_fulltree)\n",
    "CP_ASVs = [asv for asv in CPclades_fulltree_ASVs['ASV'] if 'ASV' in asv]\n",
    "print(len(CP_ASVs),'ASVs in',len(CPclades_fulltree),'captive clades')\n",
    "\n",
    "leftover_ASVs = set(fulltree_ASVs) - set(HR_ASVs) - set(MX_ASVs) - set(CP_ASVs)\n",
    "print(len(leftover_ASVs),'leftover ASVs')\n",
    "print(leftover_ASVs)\n",
    "\n",
    "print(CPclades_fulltree.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 289,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "359\n",
      "western_human        6805\n",
      "non_western_human     511\n",
      "wild_chimp             69\n",
      "wild_gorilla           37\n",
      "captive_chimp          26\n",
      "wild_bonobo            24\n",
      "captive_gorilla        22\n",
      "captive_bonobo         13\n",
      "captive_orangutan      11\n",
      "Name: Description, dtype: int64\n",
      "Index(['cladeName', 'HR_clade', 'HR_sampleNum', 'ASV', 'ASVsNum', 'ASVsTax',\n",
      "       'sampleNames', 'sampleNum', 'sampleTypes', 'captive_bonobo',\n",
      "       'captive_chimp', 'captive_gorilla', 'captive_orangutan',\n",
      "       'non_western_human', 'western_human', 'wild_bonobo', 'wild_chimp',\n",
      "       'wild_gorilla'],\n",
      "      dtype='object')\n"
     ]
    }
   ],
   "source": [
    "clades = pd.concat([HRclades_fulltree,MXclades_fulltree,CPclades_fulltree])\n",
    "clades.reset_index(drop=True, inplace=True) \n",
    "print(len(clades))\n",
    "sample_type_num = pd.DataFrame()\n",
    "for index,row in clades.iterrows():\n",
    "    #print(row)\n",
    "    sample_names= row['sampleNames']\n",
    "    sample_types = [sample_type_dict[name] for name in sample_names]\n",
    "    sample_types = pd.Series(sample_types).value_counts()\n",
    "    sample_type_num = pd.concat([sample_type_num, \n",
    "                                pd.DataFrame(sample_types)], \n",
    "                                axis=1, sort=False)\n",
    "sample_type_num = sample_type_num.T.fillna(0)\n",
    "sample_type_num.index = clades.index\n",
    "sample_type_counts =  metadata['Description'].value_counts()\n",
    "print(sample_type_counts)\n",
    "sample_type_perc = sample_type_num/sample_type_counts \n",
    "clades_merged = pd.concat([clades,sample_type_perc],axis=1)\n",
    "print(clades_merged.columns)\n",
    "clades_merged = clades_merged[['cladeName', 'HR_clade', 'HR_sampleNum', 'ASVsNum', 'ASVsTax',\n",
    "        'sampleNum', 'sampleTypes', 'captive_bonobo',\n",
    "       'captive_chimp', 'captive_gorilla', 'captive_orangutan',\n",
    "       'non_western_human', 'western_human', 'wild_bonobo', 'wild_chimp',\n",
    "       'wild_gorilla']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 325,
   "metadata": {},
   "outputs": [],
   "source": [
    "def con_Tax(x):\n",
    "    fam = list(set([fam.split('.')[0] for fam in x]))\n",
    "    gen = list(set([gen.split('.')[1] for gen in x])) \n",
    "    if len(fam) == 1:\n",
    "        fam = fam[0].replace('f__','')\n",
    "        if len(gen) == 1:\n",
    "            gen = gen[0].replace('g__','')\n",
    "        else:\n",
    "            gen = 'Unassigned'\n",
    "    else:\n",
    "        fam = 'Unassigned'\n",
    "        gen = 'Unassigned' \n",
    "    return(fam+'_'+gen)\n",
    "    \n",
    "clades_merged['Fam_Gen']=clades_merged['ASVsTax'].apply(lambda x: con_Tax(x))\n",
    "clades_merged['Fam']=clades_merged['Fam_Gen'].apply(lambda x: str(x).split('_')[0])\n",
    "clades_merged['Gen']=clades_merged['Fam_Gen'].apply(lambda x: str(x).split('_')[1])\n",
    "clades_merged['Fam'].value_counts()\n",
    "\n",
    "clades_merged.to_csv('analyses/codiv_moeller_ASVs/ALLclades_fulltree.txt',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [],
   "source": [
    "collasped_tree = Tree(tree_file, format=0)\n",
    "\n",
    "def collapse_node(tree,clade_ASVs,cladeName):\n",
    "    if len(clade_ASVs) > 1:\n",
    "        node = tree.get_common_ancestor(clade_ASVs)\n",
    "        node.name=cladeName\n",
    "        children = node.get_children()\n",
    "        node.remove_child(children[1])\n",
    "        node.remove_child(children[0])\n",
    "        return(tree)\n",
    "    else:\n",
    "        ASV = clade_ASVs[0]\n",
    "        leaf = tree.get_leaves_by_name(name=ASV)[0]\n",
    "        leaf.name = cladeName\n",
    "         \n",
    "clades_merged.apply(lambda row: collapse_node(collasped_tree,row['ASV'],row['cladeName']),axis=1)\n",
    "clade_leaves = [leaf for leaf in collasped_tree.get_leaves() if 'clade' in leaf.name]\n",
    "collasped_tree.prune(clade_leaves)\n",
    "collasped_tree.write(format=2, outfile='analyses/codiv_moeller_ASVs/full_tree_collapsed.tre')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "   /-clade_2\n",
      "--|\n",
      "  |   /-clade_3\n",
      "   \\-|\n",
      "      \\-clade_4\n"
     ]
    }
   ],
   "source": [
    "full_tree = Tree(tree_file, format=0)\n",
    "print(Bt1_tree)\n",
    "\n",
    "def collapse_node(tree,clade_ASVs,cladeName):\n",
    "    node = tree.get_common_ancestor(clade_ASVs)\n",
    "    node.name=cladeName\n",
    "    children = node.get_children()\n",
    "    node.remove_child(children[1])\n",
    "    node.remove_child(children[0])\n",
    "    return(tree)\n",
    "\n",
    "Bt1_tree = full_tree.get_common_ancestor(list(Bt1_lineage_ASVs['ASV']))\n",
    "HRclades_Bt1.apply(lambda row: collapse_node(Bt1_tree,row['ASV'],row['cladeName']),axis=1)\n",
    "HRclades_Bt1.to_csv('analyses/codiv_moeller_ASVs/HR_collapsed.txt',sep='\\t')\n",
    "Bt1_tree.write(format=2, outfile='analyses/codiv_moeller_ASVs/Bt1_collapsed.tre')\n",
    "\n",
    "Bt2_tree = full_tree.get_common_ancestor(list(Bt2_lineage_ASVs['ASV']))\n",
    "HRclades_Bt2.apply(lambda row: collapse_node(Bt2_tree,row['ASV'],row['cladeName']),axis=1)\n",
    "MXclades_Bt2.apply(lambda row: collapse_node(Bt2_tree,row['ASV'],row['cladeName']),axis=1)\n",
    "Bt2_tree.write(format=2, outfile='analyses/codiv_moeller_ASVs/Bt2_collapsed.tre')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'ASV_411', 'ASV_300', 'ASV_725', 'ASV_726', 'ASV_930', 'ASV_935', 'ASV_365', 'ASV_931', 'ASV_727'}\n",
      "       ASV  western_human  non_western_human  captive_chimp  captive_gorilla  \\\n",
      "0  ASV_300            0.0                0.0           13.0              8.0   \n",
      "1  ASV_365            0.0                0.0           15.0              3.0   \n",
      "2  ASV_410            0.0                1.0            5.0              0.0   \n",
      "3  ASV_411            0.0                0.0           14.0              0.0   \n",
      "4  ASV_725            0.0                0.0            5.0              3.0   \n",
      "\n",
      "   captive_bonobo  wild_chimp  captive_orangutan  wild_bonobo  wild_gorilla  \n",
      "0             1.0         0.0                1.0          0.0           0.0  \n",
      "1             0.0         0.0                0.0          0.0           0.0  \n",
      "2             5.0         0.0                5.0          0.0           0.0  \n",
      "3             2.0         0.0                0.0          0.0           0.0  \n",
      "4             0.0         0.0                0.0          0.0           0.0  \n"
     ]
    }
   ],
   "source": [
    "print(set(CPclades_fulltree_ASVs['ASV']) & set(leftover_ASVs))\n",
    "sample_type_num_leftovrs = sample_type_num[sample_type_num.index.isin(leftover_ASVs)]\n",
    "sample_type_num_leftovrs.index.name = 'ASV'\n",
    "sample_type_num_leftovrs.reset_index(inplace=True)\n",
    "print(sample_type_num_leftovrs.head())\n",
    "sample_type_num_leftovrs.to_csv('analyses/codiv_moeller_ASVs/Table_sampletypes_leftovers_ASV.txt',sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Tables summarizing metadata of samples harboring ASVs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       western_human  non_western_human  captive_chimp  captive_gorilla  \\\n",
      "ASV_1         2917.0               64.0            3.0              2.0   \n",
      "ASV_2         2074.0               14.0            9.0              5.0   \n",
      "ASV_3         1907.0               17.0            1.0              2.0   \n",
      "ASV_4         1798.0               18.0            0.0              0.0   \n",
      "ASV_5         1208.0               18.0            0.0              2.0   \n",
      "\n",
      "       captive_bonobo  wild_chimp  captive_orangutan  wild_bonobo  \\\n",
      "ASV_1             1.0         1.0                0.0          0.0   \n",
      "ASV_2             0.0         9.0                4.0          0.0   \n",
      "ASV_3             2.0         1.0                0.0          0.0   \n",
      "ASV_4             0.0         1.0                0.0          0.0   \n",
      "ASV_5             2.0        18.0                4.0          0.0   \n",
      "\n",
      "       wild_gorilla  \n",
      "ASV_1           0.0  \n",
      "ASV_2           0.0  \n",
      "ASV_3           0.0  \n",
      "ASV_4           0.0  \n",
      "ASV_5           0.0  \n",
      "western_human        6805\n",
      "non_western_human     511\n",
      "wild_chimp             69\n",
      "wild_gorilla           37\n",
      "captive_chimp          26\n",
      "wild_bonobo            24\n",
      "captive_gorilla        22\n",
      "captive_bonobo         13\n",
      "captive_orangutan      11\n",
      "Name: Description, dtype: int64\n",
      "       captive_bonobo  captive_chimp  captive_gorilla  captive_orangutan  \\\n",
      "ASV_1        0.076923       0.115385         0.090909           0.000000   \n",
      "ASV_2        0.000000       0.346154         0.227273           0.363636   \n",
      "ASV_3        0.153846       0.038462         0.090909           0.000000   \n",
      "ASV_4        0.000000       0.000000         0.000000           0.000000   \n",
      "ASV_5        0.153846       0.000000         0.090909           0.363636   \n",
      "\n",
      "       non_western_human  western_human  wild_bonobo  wild_chimp  wild_gorilla  \n",
      "ASV_1           0.125245       0.428655          0.0    0.014493           0.0  \n",
      "ASV_2           0.027397       0.304776          0.0    0.130435           0.0  \n",
      "ASV_3           0.033268       0.280235          0.0    0.014493           0.0  \n",
      "ASV_4           0.035225       0.264217          0.0    0.014493           0.0  \n",
      "ASV_5           0.035225       0.177517          0.0    0.260870           0.0  \n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Index(['ASV_1', 'ASV_2', 'ASV_3', 'ASV_4', 'ASV_5', 'ASV_6', 'ASV_7', 'ASV_8',\n",
       "       'ASV_9', 'ASV_10',\n",
       "       ...\n",
       "       'ASV_10701', 'ASV_10702', 'ASV_10703', 'ASV_10704', 'ASV_10706',\n",
       "       'ASV_10707', 'ASV_10708', 'ASV_10709', 'ASV_10710', 'ASV_10711'],\n",
       "      dtype='object', length=7148)"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#creates table where ASVs are rows, columns are sample descriptions (i.e wild_chimp, captive_orangutan)\n",
    "#counts are the number of samples or the percent of samples of a given type over the total number of samples belong to that type\n",
    "\n",
    "sample_type_num = pd.DataFrame()\n",
    "for ASV in ASV_sampleName_dict:\n",
    "        sample_names = ASV_sampleName_dict[ASV]\n",
    "        #sample_type_dict relates sample names to their descriptions \n",
    "        sample_types = [sample_type_dict[name] for name in sample_names]\n",
    "        sample_types = pd.Series(sample_types).value_counts()\n",
    "        row = pd.DataFrame(sample_types)\n",
    "        sample_type_num = pd.concat([sample_type_num, row], axis=1, sort=False)\n",
    "sample_type_num.columns = ASV_sampleName_dict.keys()\n",
    "sample_type_num = sample_type_num.T \n",
    "sample_type_num = sample_type_num.fillna(0)\n",
    "print(sample_type_num.head())\n",
    "sample_type_num.to_csv('analyses/codiv_moeller_ASVs/Table_sampletypes_ASV_count.txt',sep='\\t',index=False)\n",
    "sample_type_counts =  metadata['Description'].value_counts()\n",
    "print(sample_type_counts)\n",
    "sample_type_perc = sample_type_num/sample_type_counts \n",
    "print(sample_type_perc.head())\n",
    "sample_type_perc.to_csv('analyses/codiv_moeller_ASVs/Table_sampletypes_ASV_perc.txt',sep='\\t',index=False)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       western_human_CHN  western_human_DNK  western_human_USA  \\\n",
      "ASV_1              571.0              278.0              269.0   \n",
      "ASV_2              382.0              255.0              134.0   \n",
      "ASV_3              447.0              229.0              174.0   \n",
      "ASV_4              366.0              179.0              160.0   \n",
      "ASV_5              265.0               94.0              128.0   \n",
      "\n",
      "       western_human_SWE  western_human_ISR  western_human_ESP  \\\n",
      "ASV_1              233.0              220.0              204.0   \n",
      "ASV_2              203.0              160.0              141.0   \n",
      "ASV_3              177.0              121.0              129.0   \n",
      "ASV_4              130.0              212.0              128.0   \n",
      "ASV_5              172.0               26.0               44.0   \n",
      "\n",
      "       western_human_NLD  western_human_DEU  western_human_GBR  \\\n",
      "ASV_1              194.0              165.0              137.0   \n",
      "ASV_2              111.0              124.0              126.0   \n",
      "ASV_3               49.0              124.0              119.0   \n",
      "ASV_4              121.0               94.0              137.0   \n",
      "ASV_5               12.0               41.0               68.0   \n",
      "\n",
      "       western_human_FIN  ...  western_human_HUN  wild_gorilla_EK  \\\n",
      "ASV_1              127.0  ...                0.0              0.0   \n",
      "ASV_2               88.0  ...                0.0              0.0   \n",
      "ASV_3               79.0  ...                0.0              0.0   \n",
      "ASV_4               37.0  ...                0.0              0.0   \n",
      "ASV_5              109.0  ...                0.0              0.0   \n",
      "\n",
      "       non_western_human_TZA  wild_gorilla_CP  wild_gorilla_DG  wild_chimp_DG  \\\n",
      "ASV_1                    0.0              0.0              0.0            0.0   \n",
      "ASV_2                    0.0              0.0              0.0            0.0   \n",
      "ASV_3                    0.0              0.0              0.0            0.0   \n",
      "ASV_4                    0.0              0.0              0.0            0.0   \n",
      "ASV_5                    0.0              0.0              0.0            0.0   \n",
      "\n",
      "       wild_chimp_CP  wild_bonobo_LK  wild_bonobo_KR  wild_gorilla_LB  \n",
      "ASV_1            0.0             0.0             0.0              0.0  \n",
      "ASV_2            0.0             0.0             0.0              0.0  \n",
      "ASV_3            0.0             0.0             0.0              0.0  \n",
      "ASV_4            0.0             0.0             0.0              0.0  \n",
      "ASV_5            0.0             0.0             0.0              0.0  \n",
      "\n",
      "[5 rows x 51 columns]\n",
      "western_human        6805\n",
      "non_western_human     511\n",
      "wild_chimp             69\n",
      "wild_gorilla           37\n",
      "captive_chimp          26\n",
      "wild_bonobo            24\n",
      "captive_gorilla        22\n",
      "captive_bonobo         13\n",
      "captive_orangutan      11\n",
      "Name: Description, dtype: int64\n",
      "       captive_bonobo_COLZ  captive_chimp_HOUZ  captive_chimp_PC  \\\n",
      "ASV_1             0.076923            0.333333          0.000000   \n",
      "ASV_2             0.000000            0.555556          0.235294   \n",
      "ASV_3             0.153846            0.111111          0.000000   \n",
      "ASV_4             0.000000            0.000000          0.000000   \n",
      "ASV_5             0.153846            0.000000          0.000000   \n",
      "\n",
      "       captive_gorilla_COLZ  captive_gorilla_HOUZ  captive_orangutan_COLZ  \\\n",
      "ASV_1              0.071429                 0.125                     0.0   \n",
      "ASV_2              0.071429                 0.500                     0.5   \n",
      "ASV_3              0.142857                 0.000                     0.0   \n",
      "ASV_4              0.000000                 0.000                     0.0   \n",
      "ASV_5              0.142857                 0.000                     0.0   \n",
      "\n",
      "       captive_orangutan_HOUZ  non_western_human_FJI  non_western_human_MDG  \\\n",
      "ASV_1                0.000000               0.149194               0.009524   \n",
      "ASV_2                0.285714               0.024194               0.019048   \n",
      "ASV_3                0.000000               0.008065               0.019048   \n",
      "ASV_4                0.000000               0.024194               0.019048   \n",
      "ASV_5                0.571429               0.012097               0.038095   \n",
      "\n",
      "       non_western_human_MNG  ...  wild_chimp_EK  wild_chimp_GM  \\\n",
      "ASV_1               0.400000  ...            0.0       0.020833   \n",
      "ASV_2               0.092308  ...            0.5       0.145833   \n",
      "ASV_3               0.200000  ...            0.0       0.000000   \n",
      "ASV_4               0.153846  ...            0.0       0.020833   \n",
      "ASV_5               0.169231  ...            0.0       0.354167   \n",
      "\n",
      "       wild_chimp_LU  wild_chimp_OP  wild_gorilla_CP  wild_gorilla_DG  \\\n",
      "ASV_1            0.0          0.000              0.0              0.0   \n",
      "ASV_2            0.0          0.125              0.0              0.0   \n",
      "ASV_3            0.0          0.125              0.0              0.0   \n",
      "ASV_4            0.0          0.000              0.0              0.0   \n",
      "ASV_5            0.0          0.125              0.0              0.0   \n",
      "\n",
      "       wild_gorilla_EK  wild_gorilla_LB  wild_gorilla_LU  wild_gorilla_OP  \n",
      "ASV_1              0.0              0.0              0.0              0.0  \n",
      "ASV_2              0.0              0.0              0.0              0.0  \n",
      "ASV_3              0.0              0.0              0.0              0.0  \n",
      "ASV_4              0.0              0.0              0.0              0.0  \n",
      "ASV_5              0.0              0.0              0.0              0.0  \n",
      "\n",
      "[5 rows x 51 columns]\n"
     ]
    }
   ],
   "source": [
    "#creates table where ASVs are rows, columns are sample descriptions plus site \n",
    "#counts are the number of samples or the percent of samples of a given type over the total number of samples belong to that type\n",
    "\n",
    "metadata['Description_site'] = metadata['Description']+'_' +metadata['site_code']\n",
    "sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "\n",
    "captive_only_desc = list(set(metadata['Description_site']))\n",
    "captive_only_desc = [desc for desc in captive_only_desc if 'captive' in str(desc)]\n",
    "captive_only_desc\n",
    "\n",
    "sample_type_site_num = pd.DataFrame()\n",
    "for ASV in ASV_sampleName_dict:\n",
    "        sample_names = ASV_sampleName_dict[ASV]\n",
    "        sample_types = [sample_type_site_dict[name] for name in sample_names]\n",
    "        sample_types = pd.Series(sample_types).value_counts()\n",
    "        row = pd.DataFrame(sample_types)\n",
    "        sample_type_site_num = pd.concat([sample_type_site_num, row], axis=1, sort=False)\n",
    "sample_type_site_num.columns = ASV_sampleName_dict.keys()\n",
    "sample_type_site_num = sample_type_site_num.T \n",
    "sample_type_site_num = sample_type_site_num.fillna(0)\n",
    "print(sample_type_site_num.head())\n",
    "sample_type_site_num.to_csv('analyses/codiv_moeller_ASVs/Table_sampletypes_site_ASV_count.txt',sep='\\t',index=False)\n",
    "sample_type_site_counts =  metadata['Description_site'].value_counts()\n",
    "print(sample_type_counts)\n",
    "sample_type_site_perc = sample_type_site_num/sample_type_site_counts \n",
    "print(sample_type_site_perc.head())\n",
    "sample_type_site_perc.to_csv('analyses/codiv_moeller_ASVs/Table_sampletypes_site_ASV_perc.txt',sep='\\t',index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'sample_type_site_num' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-128-59555d89442a>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m      1\u001b[0m \u001b[0;31m#subset to just captive ape descriptions\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 2\u001b[0;31m \u001b[0msample_type_site_num\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mcolumns\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      3\u001b[0m \u001b[0msample_type_site_num_just_captive\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0msample_type_site_num\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mcaptive_only_desc\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0msample_type_site_num_just_captive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mindex\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;34m'ASV'\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      5\u001b[0m \u001b[0msample_type_site_num_just_captive\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0minplace\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'sample_type_site_num' is not defined"
     ]
    }
   ],
   "source": [
    "#subset to just captive ape descriptions\n",
    "sample_type_site_num.columns\n",
    "sample_type_site_num_just_captive = sample_type_site_num[captive_only_desc]\n",
    "sample_type_site_num_just_captive.index.name = 'ASV'\n",
    "sample_type_site_num_just_captive.reset_index(inplace=True)\n",
    "sample_type_site_num_just_captive.to_csv('analyses/codiv_moeller_ASVs/Table_sampletypes_site_captive_ASV_count.txt',sep='\\t',index=False)\n",
    "sample_type_site_perc_just_captive = sample_type_site_perc[captive_only_desc]\n",
    "sample_type_site_perc_just_captive.index.name = 'ASV'\n",
    "sample_type_site_perc_just_captive.reset_index(inplace=True)\n",
    "sample_type_site_perc_just_captive.to_csv('analyses/codiv_moeller_ASVs/Table_sampletypes_site_captive_ASV_perc.txt',sep='\\t',index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
