{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 278,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/AHN/captive_ape_microbiome/results/16s'"
      ]
     },
     "execution_count": 278,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from ete3 import Tree\n",
    "from collections import Counter\n",
    "\n",
    "os.getcwd()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 279,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gyrb inputs\n",
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/results/gyrb/')\n",
    "metadata_file = 'inputs/metadata_gyrb_amp_meta_passing_samples.txt'\n",
    "tax_table_file = 'inputs/ASVs_taxonomy.txt'\n",
    "asv_table_file = 'inputs/ASVs_filtered_counts.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 280,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human</th>\n",
       "      <th>wild_bonobo</th>\n",
       "      <th>wild_gorilla</th>\n",
       "      <th>wild_chimp</th>\n",
       "      <th>CP</th>\n",
       "      <th>MX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALL</th>\n",
       "      <td>6183.0</td>\n",
       "      <td>353.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP</th>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Prevotella</th>\n",
       "      <td>2450.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Bacteroides</th>\n",
       "      <td>940.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>25.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unassigned</th>\n",
       "      <td>199.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Alistipes</th>\n",
       "      <td>515.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__RC9</th>\n",
       "      <td>363.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Parabacteroides</th>\n",
       "      <td>247.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Bacteroides_A</th>\n",
       "      <td>264.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Porphyromonas</th>\n",
       "      <td>243.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Prevotellamassilia</th>\n",
       "      <td>116.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Bacteroides_B</th>\n",
       "      <td>110.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                        human  wild_bonobo  wild_gorilla  wild_chimp    CP  \\\n",
       "ALL                    6183.0        353.0         258.0       207.0  94.0   \n",
       "CP                       27.0          0.0           0.0         4.0  94.0   \n",
       "g__Prevotella          2450.0        203.0          46.0       114.0  55.0   \n",
       "g__Bacteroides          940.0          1.0           0.0        14.0  12.0   \n",
       "Unassigned              199.0         94.0         181.0        41.0  11.0   \n",
       "g__Alistipes            515.0          0.0           0.0         1.0   0.0   \n",
       "g__RC9                  363.0          0.0           0.0         0.0   0.0   \n",
       "g__Parabacteroides      247.0         38.0           1.0         3.0   3.0   \n",
       "g__Bacteroides_A        264.0          0.0           2.0        17.0   3.0   \n",
       "g__Porphyromonas        243.0          1.0           0.0         2.0   4.0   \n",
       "g__Prevotellamassilia   116.0          0.0           4.0         1.0   0.0   \n",
       "g__Bacteroides_B        110.0          0.0           0.0         0.0   0.0   \n",
       "\n",
       "                         MX  \n",
       "ALL                    53.0  \n",
       "CP                     14.0  \n",
       "g__Prevotella           3.0  \n",
       "g__Bacteroides         25.0  \n",
       "Unassigned             13.0  \n",
       "g__Alistipes            3.0  \n",
       "g__RC9                  0.0  \n",
       "g__Parabacteroides      4.0  \n",
       "g__Bacteroides_A        0.0  \n",
       "g__Porphyromonas        0.0  \n",
       "g__Prevotellamassilia   0.0  \n",
       "g__Bacteroides_B        3.0  "
      ]
     },
     "execution_count": 280,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#read in asv table, get sample names and number for each asv\n",
    "def is_HR(sampleNames,sample_type_dict):\n",
    "    \"\"\"given a list of sample names uses sample type dictionary to determine how many sample types are present\n",
    "    designates ASVs as host restricted = 1 sample type or mixed = multiple sample types.\n",
    "    Captive sample types are not considered so some clades will have a 0 sample type length and they can fall within\n",
    "    host-restricted clades or mixed clades or neither\"\"\"\n",
    "    sampleTypes = [sample_type_dict[name] for name in sampleNames]\n",
    "    sampleTypes = [x.replace('non_western_','').replace('western_','') for x in sampleTypes]\n",
    "    neutral_sampleTypes = ['captive_gorilla','captive_bonobo','captive_chimp','captive_orangutan']\n",
    "    HR_sampleTypes = list(set(sampleTypes) - set(neutral_sampleTypes))\n",
    "    HR_sampleNum = len([x for x in sampleTypes if x not in neutral_sampleTypes])\n",
    "    CP_sampleTypes = list(set(sampleTypes) & set(neutral_sampleTypes))\n",
    "    CP_sampleNum = len([x for x in sampleTypes if x in neutral_sampleTypes])\n",
    "    CP_pres = True if len(CP_sampleTypes) > 0 else False\n",
    "    if len(HR_sampleTypes) == 0:\n",
    "        HR_cat,HR_type='CP','CP'\n",
    "    if len(HR_sampleTypes) == 1: #identifies host-restricted clades\n",
    "        HR_cat,HR_type='HR',HR_sampleTypes[0]  \n",
    "    if len(HR_sampleTypes) > 1:    \n",
    "        HR_cat,HR_type ='MX','MX'\n",
    "    return(HR_sampleTypes,HR_sampleNum,HR_cat,HR_type,CP_pres,CP_sampleTypes,CP_sampleNum)\n",
    "\n",
    "\n",
    "def asv_hr_table(asv_table_file,metadata_file,tax_table_file):\n",
    "    asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "    sampleNames = asv_table.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "    asv_df = sampleNames.reset_index()\n",
    "    asv_df.columns = ['ASV','sampleNames']\n",
    "    asv_df['sampleNum'] = asv_df['sampleNames'].apply(lambda names: len(names))\n",
    "    \n",
    "    #add host restriction info\n",
    "    metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "    sample_type_dict = dict(zip(metadata['X.SampleID'], metadata['Description'])) \n",
    "    hr = asv_df['sampleNames'].apply(lambda x: pd.Series(is_HR(x,sample_type_dict),\n",
    "                                                         index=['HR_sampleTypes','HR_sampleNum','HR_cat','HR_type',\n",
    "                                                                'CP_pres','CP_sampleTypes','CP_sampleNum']))\n",
    "    asv_hr_df = asv_df.merge(hr,left_index=True, right_index=True)\n",
    "    \n",
    "    #add taxonomic info\n",
    "    tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "    tax_table = tax_table[['ASV','Phylum','Order','Family','Genus']]\n",
    "    asv_full = asv_hr_df.merge(tax_table,on='ASV',how='left')\n",
    "    \n",
    "    return(asv_full)\n",
    "    \n",
    "asv_full = asv_hr_table(asv_table_file,metadata_file,tax_table_file)\n",
    "\n",
    "def output_summary_table(asv_full,outfile):\n",
    "     #create summary table\n",
    "    all_asvs = pd.Series(asv_full['HR_type'].value_counts(),name='ALL')\n",
    "    cp_asv = pd.Series(asv_full[asv_full['CP_pres']==True]['HR_type'].value_counts(),name='CP')\n",
    "    num_Gen = len(asv_full['Genus'].value_counts())\n",
    "    top_Gen = pd.DataFrame(\n",
    "            asv_full['Genus'].value_counts().index[:10],\n",
    "            columns=['Genus'])\n",
    "    gen_asvs = pd.DataFrame(top_Gen['Genus'].apply(\n",
    "            lambda Genus: \n",
    "            pd.Series(asv_full[asv_full['Genus']==Genus]['HR_type'].value_counts(),name=Genus)\n",
    "            )).T    \n",
    "    gen_asvs.columns = list(top_Gen['Genus'])  \n",
    "    res = pd.concat([all_asvs,cp_asv,gen_asvs],axis=1).fillna(0).T\n",
    "    res.to_csv(outfile,sep='\\t')\n",
    "    return(res)\n",
    "    \n",
    "output_summary_table(asv_full,'analyses/codiv_moeller_ASVs/gyrb_hr_asv_table.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 281,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human</th>\n",
       "      <th>wild_gorilla</th>\n",
       "      <th>wild_chimp</th>\n",
       "      <th>wild_bonobo</th>\n",
       "      <th>MX</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALL</th>\n",
       "      <td>22.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>13.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP</th>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Prevotella</th>\n",
       "      <td>11.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WCHB1-69_Unassigned</th>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Unassigned</th>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Bacteroides</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tannerellaceae_Parabacteroides</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unassigned_Unassigned</th>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rikenellaceae_Alistipes</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tannerellaceae_Unassigned</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Bacteroides_B</th>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Dysgonomonadaceae_Dysgonomonas</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                human  wild_gorilla  wild_chimp  wild_bonobo  \\\n",
       "ALL                              22.0          19.0        15.0         14.0   \n",
       "CP                                7.0           0.0         1.0          0.0   \n",
       "Bacteroidaceae_Prevotella        11.0           4.0         8.0          8.0   \n",
       "WCHB1-69_Unassigned               0.0           7.0         0.0          0.0   \n",
       "Bacteroidaceae_Unassigned         2.0           3.0         0.0          3.0   \n",
       "Bacteroidaceae_Bacteroides        1.0           0.0         1.0          0.0   \n",
       "Tannerellaceae_Parabacteroides    1.0           0.0         1.0          1.0   \n",
       "Unassigned_Unassigned             1.0           3.0         0.0          0.0   \n",
       "Rikenellaceae_Alistipes           2.0           0.0         0.0          0.0   \n",
       "Tannerellaceae_Unassigned         0.0           1.0         2.0          0.0   \n",
       "Bacteroidaceae_Bacteroides_B      1.0           0.0         0.0          0.0   \n",
       "Dysgonomonadaceae_Dysgonomonas    0.0           1.0         0.0          0.0   \n",
       "\n",
       "                                  MX  \n",
       "ALL                             13.0  \n",
       "CP                               8.0  \n",
       "Bacteroidaceae_Prevotella        2.0  \n",
       "WCHB1-69_Unassigned              3.0  \n",
       "Bacteroidaceae_Unassigned        0.0  \n",
       "Bacteroidaceae_Bacteroides       4.0  \n",
       "Tannerellaceae_Parabacteroides   1.0  \n",
       "Unassigned_Unassigned            0.0  \n",
       "Rikenellaceae_Alistipes          1.0  \n",
       "Tannerellaceae_Unassigned        0.0  \n",
       "Bacteroidaceae_Bacteroides_B     1.0  \n",
       "Dysgonomonadaceae_Dysgonomonas   1.0  "
      ]
     },
     "execution_count": 281,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clades_df=pd.read_csv('analyses/codiv_moeller_ASVs/full_tree_clades_collapsed_table.txt',sep='\\t')\n",
    "clades_df=clades_df.rename(columns={\"cladeTax\": \"Genus\"})\n",
    "output_summary_table(clades_df,'analyses/codiv_moeller_ASVs/gyrb_hr_clade_table.txt')\n",
    "clades_df_15perc =  clades_df.loc[(clades_df.wild_bonobo > 0.15) | \n",
    "              (clades_df.wild_chimp > 0.15)  |\n",
    "              (clades_df.wild_gorilla > 0.15)|\n",
    "              (clades_df.western_human > 0.15)|\n",
    "              (clades_df.non_western_human > 0.15)] \n",
    "output_summary_table(clades_df_15perc,'analyses/codiv_moeller_ASVs/gyrb_hr_clade_15perc_table.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 288,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   numEnclosure          multi_site_sp  count\n",
      "0             1  single_site_single_sp     97\n",
      "1             2    multi_site_multi_sp      8\n",
      "2             2   multi_site_single_sp      2\n",
      "3             2   single_site_multi_sp      9\n",
      "4             3    multi_site_multi_sp     10\n",
      "5             4    multi_site_multi_sp      5\n",
      "6             5    multi_site_multi_sp      3\n",
      "7             6    multi_site_multi_sp      4\n",
      "8             7    multi_site_multi_sp      1\n"
     ]
    }
   ],
   "source": [
    "def multi_site_sp(cp_desc):\n",
    "    sites = list(set([x.split('_')[2] for x in cp_desc]))\n",
    "    sitesNum = len(sites)\n",
    "    sitesMulti = 'multi_site' if sitesNum > 1 else 'single_site'      \n",
    "    species = list(set([x.split('_')[1] for x in cp_desc]))\n",
    "    speciesNum = len(species)\n",
    "    speciessMulti = 'multi_sp' if speciesNum > 1 else 'single_sp' \n",
    "    return(sitesMulti + '_' + speciessMulti)\n",
    "multi_site_sp(['captive_chimp_HOUZ','captive_gorilla_HOUZ'])\n",
    "\n",
    "def captive_apes_asv_summary(asv_full,metadata_file):\n",
    "    asv_cp = asv_full[asv_full['CP_pres']==True]\n",
    "    metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "    metadata['Description_site'] = metadata['Description']+'_' +metadata['site_code']\n",
    "    sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "\n",
    "    description_df = asv_cp['sampleNames'].apply(lambda l: pd.Series(\n",
    "        [sample_type_site_dict[name] for name in l]).value_counts())\n",
    "    description_df = description_df.fillna(0)  \n",
    "    capt_desc = list(set(metadata['Description_site'][metadata['captivity_status']=='captive']))\n",
    "    description_df = description_df[capt_desc]\n",
    "    description_df['CP_sp_loc'] = description_df.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "    description_df['numEnclosure'] = description_df['CP_sp_loc'].apply(lambda x: len(x))\n",
    "    description_df['multi_site_sp'] = description_df['CP_sp_loc'].apply(lambda x:  \n",
    "                                                                        multi_site_sp(x))\n",
    "    asv_cp_table = asv_cp.merge(description_df,left_index=True,right_index=True)\n",
    "    return(asv_cp_table)\n",
    "\n",
    "asv_cp_table = captive_apes_asv_summary(asv_full,metadata_file)\n",
    "asv_cp_table_summary = asv_cp_table.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "print(asv_cp_table_summary)\n",
    "asv_cp_table_summary.to_csv('analyses/codiv_moeller_ASVs/numEnclosures_table.txt',sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "asv_table = asv_table.reset_index().rename(columns={\"index\": \"ASV\"})\n",
    "tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "asv_table_gen = asv_table.merge(tax_table[['ASV','Phylum','Family','Genus']],on='ASV',how='left')\n",
    "asv_table_gen = asv_table_gen[asv_table_gen['Genus']!='Unassigned']\n",
    "\n",
    "def shared_ASVs_common_gen(ind1,ind2,asv_table_gen):\n",
    "    pw = asv_table_gen[['Phylum','Family','Genus','ASV',ind1,ind2]][\n",
    "        (asv_table_gen[ind1]>0)|(asv_table_gen[ind2]>0)\n",
    "        ]\n",
    "    #print(pw)\n",
    "    common_gen = list(\n",
    "                 set(pw[pw[ind1]>0]['Genus']) & \n",
    "                 set(pw[pw[ind2]>0]['Genus']))\n",
    "    pw= pw[pw['Genus'].isin(common_gen)]\n",
    "    ind1_ASVs = pw[ind1].sum()\n",
    "    ind2_ASVs = pw[ind2].sum()\n",
    "    common_ASVs = len(pw[(pw[ind1]>0)&(pw[ind2]>0)])\n",
    "    res = pd.Series([ind1,ind2,ind1_ASVs,ind2_ASVs,common_ASVs],\n",
    "              index=['ind1','ind2','ind1_ASVs','ind2_ASVs','common_ASVs'])\n",
    "    return(res)\n",
    "\n",
    "shared_ASVs_common_gen('cp.bon.COLZ.01.Bt','cp.bon.COLZ.02.Bt',asv_table_gen)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 263,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 201,
   "metadata": {},
   "outputs": [],
   "source": [
    "pw_shared_ASV_all_genera = pw_df.merge(shared_ASV_all_genera,on=['ind1','ind2'])\n",
    "\n",
    "def prop_shared(row):\n",
    "    if row['common_ASVs'] > 0:\n",
    "        return(row['common_ASVs']/min(row['ind1_ASVs'],row['ind2_ASVs']))\n",
    "    else:\n",
    "        return(0)\n",
    "                                   \n",
    "pw_shared_ASV_all_genera['prop_shared_ASVs'] = pw_shared_ASV_all_genera.apply(\n",
    "    lambda row: prop_shared(row),axis=1)\n",
    "pw_shared_ASV_all_genera.to_csv('analyses/codiv_moeller_ASVs/gyrb_hr_clade_15perc_table.txt',sep='\\t',index=Faluse)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 202,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_table_Prevotella = asv_table_gen[asv_table_gen['Genus']=='g__Prevotella']\n",
    "shared_ASV_Prevotella = pw_df.apply(lambda row: \n",
    "                    shared_ASVs_common_gen(row['ind1'],row['ind2'],asv_table_Prevotella),axis=1)\n",
    "pw_shared_ASV_Prevotella = pw_df.merge(shared_ASV_Prevotella,on=['ind1','ind2'])\n",
    "pw_shared_ASV_Prevotella['prop_shared_ASVs'] = pw_shared_ASV_Prevotella.apply(\n",
    "    lambda row: prop_shared(row),axis=1)\n",
    "pw_shared_ASV_all_genera.to_csv('analyses/codiv_moeller_ASVs/gyrb_prop_shared_ASVs_Prevotella_table.txt',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 16S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 271,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/results/16s/')\n",
    "metadata_file = 'inputs/16S_metadata.txt'\n",
    "tax_table_file = 'inputs/ASVs_taxonomy.txt'\n",
    "asv_table_file = 'inputs/ASV_tab.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 272,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 272,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#sanity check make sure all sample names overlap\n",
    "asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "set(asv_table.columns) - set(metadata['X.SampleID'])\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 273,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>human</th>\n",
       "      <th>MX</th>\n",
       "      <th>wild_chimp</th>\n",
       "      <th>wild_gorilla</th>\n",
       "      <th>CP</th>\n",
       "      <th>wild_bonobo</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALL</th>\n",
       "      <td>211.0</td>\n",
       "      <td>88.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>22.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP</th>\n",
       "      <td>83.0</td>\n",
       "      <td>35.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella_9</th>\n",
       "      <td>38.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroides</th>\n",
       "      <td>33.0</td>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rikenellaceae_RC9_gut_group</th>\n",
       "      <td>16.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella_7</th>\n",
       "      <td>4.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alloprevotella</th>\n",
       "      <td>9.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella_2</th>\n",
       "      <td>20.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parabacteroides</th>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alistipes</th>\n",
       "      <td>12.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotellaceae_NK3B31_group</th>\n",
       "      <td>5.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotellaceae_UCG-001</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             human    MX  wild_chimp  wild_gorilla    CP  \\\n",
       "ALL                          211.0  88.0        50.0          34.0  29.0   \n",
       "CP                            83.0  35.0         0.0           2.0  29.0   \n",
       "Prevotella_9                  38.0  11.0         7.0           2.0   0.0   \n",
       "Bacteroides                   33.0  16.0         0.0           0.0   0.0   \n",
       "Rikenellaceae_RC9_gut_group   16.0  13.0         7.0           8.0   0.0   \n",
       "Prevotella_7                   4.0   4.0         7.0           6.0   0.0   \n",
       "Alloprevotella                 9.0   6.0         4.0           1.0   0.0   \n",
       "Prevotella_2                  20.0   1.0         0.0           0.0   0.0   \n",
       "Parabacteroides                6.0   9.0         0.0           0.0   0.0   \n",
       "Alistipes                     12.0   3.0         0.0           0.0   0.0   \n",
       "Prevotellaceae_NK3B31_group    5.0   2.0         1.0           2.0   0.0   \n",
       "Prevotellaceae_UCG-001         3.0   1.0         4.0           0.0   2.0   \n",
       "\n",
       "                             wild_bonobo  \n",
       "ALL                                 22.0  \n",
       "CP                                   0.0  \n",
       "Prevotella_9                         2.0  \n",
       "Bacteroides                          0.0  \n",
       "Rikenellaceae_RC9_gut_group          0.0  \n",
       "Prevotella_7                         2.0  \n",
       "Alloprevotella                       2.0  \n",
       "Prevotella_2                         0.0  \n",
       "Parabacteroides                      1.0  \n",
       "Alistipes                            0.0  \n",
       "Prevotellaceae_NK3B31_group          1.0  \n",
       "Prevotellaceae_UCG-001               1.0  "
      ]
     },
     "execution_count": 273,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_16s = asv_hr_table(asv_table_file,metadata_file,tax_table_file)\n",
    "output_summary_table(asv_16s,'analyses/tables/16s_hr_all_asvs_table.txt')\n",
    "\n",
    "asv_Bifidobacteriales = asv_16s[asv_16s['Order']=='Bifidobacteriales']\n",
    "output_summary_table(asv_Bifidobacteriales,'analyses/tables/16s_hr_Bifidobacteriales_asvs_table.txt')\n",
    "\n",
    "asv_Bacteroidales = asv_16s[asv_16s['Order']=='Bacteroidales']\n",
    "output_summary_table(asv_Bacteroidales,'analyses/tables/16s_hr_Bacteroidales_asvs_table.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'asv_16s' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-17b16e47063c>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0masv_16s\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0masv_16s\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'HR_type'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;34m'human'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'CP_sampleTypes'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mvalue_counts\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'asv_16s' is not defined"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 276,
   "metadata": {},
   "outputs": [],
   "source": [
    "def captive_apes_asv_summary(asv_full,metadata_file):\n",
    "    asv_cp = asv_full[asv_full['CP_pres']==True]\n",
    "    metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "    metadata['Description_site'] = metadata['Description']+'_' +metadata['site_code']\n",
    "    sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "\n",
    "    description_df = asv_cp['sampleNames'].apply(lambda l: pd.Series(\n",
    "        [sample_type_site_dict[name] for name in l]).value_counts())\n",
    "    description_df = description_df.fillna(0)\n",
    "    captive_desc = list(set([x for x in metadata['Description_site'] if 'captive' in x]))\n",
    "    print(captive_desc)\n",
    "    description_df = description_df[captive_desc]\n",
    "    description_df['CP_sp_loc'] = description_df.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "    description_df['numEnclosure'] = description_df['CP_sp_loc'].apply(lambda x: len(x))\n",
    "    description_df['multi_site_sp'] = description_df['CP_sp_loc'].apply(lambda x:  \n",
    "                                                                        multi_site_sp(x))\n",
    "    asv_cp_table = asv_cp.merge(description_df,left_index=True,right_index=True)\n",
    "    return(asv_cp_table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 277,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['captive_gorilla_COLZ', 'captive_orangutan_COMZ', 'captive_orangutan_HOUZ', 'captive_chimp_HOUZ', 'captive_orangutan_ATLZ', 'captive_gorilla_COMZ', 'captive_orangutan_COLZ', 'captive_gorilla_HOUZ', 'captive_chimp_PC', 'captive_bonobo_COLZ']\n",
      "    numEnclosure          multi_site_sp  count\n",
      "0              1  single_site_single_sp    168\n",
      "1              2    multi_site_multi_sp     71\n",
      "2              2   multi_site_single_sp     30\n",
      "3              2   single_site_multi_sp     19\n",
      "4              3    multi_site_multi_sp     92\n",
      "5              3   multi_site_single_sp     10\n",
      "6              3   single_site_multi_sp     11\n",
      "7              4    multi_site_multi_sp     72\n",
      "8              5    multi_site_multi_sp     72\n",
      "9              6    multi_site_multi_sp     63\n",
      "10             7    multi_site_multi_sp     56\n",
      "11             8    multi_site_multi_sp     66\n",
      "12             9    multi_site_multi_sp     66\n",
      "13            10    multi_site_multi_sp     75\n",
      "   numEnclosure        multi_site_sp  count\n",
      "0             2  multi_site_multi_sp      2\n",
      "    numEnclosure          multi_site_sp  count\n",
      "0              1  single_site_single_sp     30\n",
      "1              2    multi_site_multi_sp     11\n",
      "2              2   multi_site_single_sp      5\n",
      "3              2   single_site_multi_sp      4\n",
      "4              3    multi_site_multi_sp     13\n",
      "5              3   multi_site_single_sp      2\n",
      "6              3   single_site_multi_sp      2\n",
      "7              4    multi_site_multi_sp     17\n",
      "8              5    multi_site_multi_sp     10\n",
      "9              6    multi_site_multi_sp     12\n",
      "10             7    multi_site_multi_sp     10\n",
      "11             8    multi_site_multi_sp     11\n",
      "12             9    multi_site_multi_sp     14\n",
      "13            10    multi_site_multi_sp      8\n"
     ]
    }
   ],
   "source": [
    "asv_cp_table = captive_apes_asv_summary(asv_16s,metadata_file)\n",
    "asv_cp_table_summary = asv_cp_table.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "print(asv_cp_table_summary)\n",
    "asv_cp_table_summary.to_csv('analyses/tables/numEnclosures_table.txt',sep='\\t',index=False)\n",
    "\n",
    "cp_Bifidobacteriales = asv_cp_table[asv_cp_table['Order']=='Bifidobacteriales']\n",
    "cp_Bifidobacteriales_summary = cp_Bifidobacteriales.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "print(cp_Bifidobacteriales_summary)\n",
    "cp_Bifidobacteriales_summary.to_csv('analyses/tables/numEnclosures_Bifidobacteriales_table.txt',sep='\\t',index=False)\n",
    "\n",
    "cp_Bacteroidales = asv_cp_table[asv_cp_table['Order']=='Bacteroidales']\n",
    "cp_Bacteroidales_summary = cp_Bacteroidales.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "print(cp_Bacteroidales_summary)\n",
    "cp_Bacteroidales_summary.to_csv('analyses/tables/numEnclosures_Bacteroidales_table.txt',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
