{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/Volumes/AHN/captive_ape_microbiome/scripts/gyrb/analyses'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from ete3 import Tree\n",
    "from collections import Counter\n",
    "\n",
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gyrb inputs\n",
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/results/gyrb/')\n",
    "metadata_file = 'inputs/metadata_gyrb_amp_meta_passing_samples.txt'\n",
    "tax_table_file = 'inputs/ASVs_taxonomy.txt'\n",
    "asv_table_file = 'inputs/ASVs_filtered_counts.tsv'\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read in asv table, get sample names and number for each asv\n",
    "def is_HR(sampleNames,sample_type_dict):\n",
    "    \"\"\"given a list of sample names uses sample type dictionary to determine how many sample types are present\n",
    "    designates ASVs as host restricted = 1 sample type or mixed = multiple sample types.\n",
    "    Captive sample types are not considered so some clades will have a 0 sample type length and they can fall within\n",
    "    host-restricted clades or mixed clades or neither\"\"\"\n",
    "    sampleTypes = [sample_type_dict[name] for name in sampleNames]\n",
    "    sampleTypes = [x.replace('non_western_','').replace('western_','') for x in sampleTypes]\n",
    "    neutral_sampleTypes = ['captive_gorilla','captive_bonobo','captive_chimp','captive_orangutan']\n",
    "    HR_sampleTypes = list(set(sampleTypes) - set(neutral_sampleTypes))\n",
    "    HR_sampleNum = len([x for x in sampleTypes if x not in neutral_sampleTypes])\n",
    "    CP_sampleTypes = list(set(sampleTypes) & set(neutral_sampleTypes))\n",
    "    CP_sampleNum = len([x for x in sampleTypes if x in neutral_sampleTypes])\n",
    "    CP_pres = True if len(CP_sampleTypes) > 0 else False\n",
    "    if len(HR_sampleTypes) == 0:\n",
    "        HR_cat,HR_type='Unique_CP','Unique_CP'\n",
    "    if len(HR_sampleTypes) == 1: #identifies host-restricted clades\n",
    "        HR_cat,HR_type='HR','HR_'+HR_sampleTypes[0]  \n",
    "    if len(HR_sampleTypes) > 1: \n",
    "        HR_cat = 'NHR'\n",
    "        if len(HR_sampleTypes) == 2:\n",
    "            if 'human' in HR_sampleTypes:\n",
    "                HR_type = 'NHR_human_single_wild_ape'\n",
    "            else:\n",
    "                HR_type = 'NHR_2_wild_apes'\n",
    "        if len(HR_sampleTypes) == 3:\n",
    "            if 'human' in HR_sampleTypes:\n",
    "                HR_type = 'NHR_human_2_wild_apes'\n",
    "            else:\n",
    "                HR_type = 'NHR_3_wild_apes'\n",
    "        if len(HR_sampleTypes) == 4:\n",
    "            HR_type = 'NHR_4_hominids'\n",
    "        \n",
    "    return(HR_sampleTypes,HR_sampleNum,HR_cat,HR_type,CP_pres,CP_sampleTypes,CP_sampleNum)\n",
    "\n",
    "\n",
    "def asv_hr_table(asv_table_file,metadata_file,tax_table_file):\n",
    "    asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "    sampleNames = asv_table.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "    asv_df = sampleNames.reset_index()\n",
    "    asv_df.columns = ['ASV','sampleNames']\n",
    "    asv_df['sampleNum'] = asv_df['sampleNames'].apply(lambda names: len(names))\n",
    "    \n",
    "    #add host restriction info\n",
    "    metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "    sample_type_dict = dict(zip(metadata['X.SampleID'], metadata['Description'])) \n",
    "    hr = asv_df['sampleNames'].apply(lambda x: pd.Series(is_HR(x,sample_type_dict),\n",
    "                                                         index=['HR_sampleTypes','HR_sampleNum','HR_cat','HR_type',\n",
    "                                                                'CP_pres','CP_sampleTypes','CP_sampleNum']))\n",
    "    asv_hr_df = asv_df.merge(hr,left_index=True, right_index=True)\n",
    "    \n",
    "    #add taxonomic info\n",
    "    tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "    tax_table = tax_table[['ASV','Phylum','Order','Family','Genus']]\n",
    "    asv_full = asv_hr_df.merge(tax_table,on='ASV',how='left')\n",
    "    \n",
    "    return(asv_full)\n",
    "    \n",
    "def output_summary_table(asv_full,outfile):\n",
    "     #create summary table\n",
    "    all_asvs = pd.Series(asv_full['HR_type'].value_counts(),name='ALL')\n",
    "    cp_asv = pd.Series(asv_full[asv_full['CP_pres']==True]['HR_type'].value_counts(),name='CP')\n",
    "    num_Gen = len(asv_full['Genus'].value_counts())\n",
    "    top_Gen = pd.DataFrame(\n",
    "            asv_full['Genus'].value_counts().index[:10],\n",
    "            columns=['Genus'])\n",
    "    gen_asvs = pd.DataFrame(top_Gen['Genus'].apply(\n",
    "            lambda Genus: \n",
    "            pd.Series(asv_full[asv_full['Genus']==Genus]['HR_type'].value_counts(),name=Genus)\n",
    "            )).T    \n",
    "    gen_asvs.columns = list(top_Gen['Genus'])  \n",
    "    res = pd.concat([all_asvs,cp_asv,gen_asvs],axis=1).fillna(0).T\n",
    "    order_col = sorted(list(res.columns))\n",
    "    print(order_col)\n",
    "    res = res[order_col]\n",
    "    res.to_csv(outfile,sep='\\t')\n",
    "    return(res)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HR_human', 'HR_wild_bonobo', 'HR_wild_chimp', 'HR_wild_gorilla', 'NHR_2_wild_apes', 'NHR_human_2_wild_apes', 'NHR_human_single_wild_ape', 'Unique_CP']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HR_human</th>\n",
       "      <th>HR_wild_bonobo</th>\n",
       "      <th>HR_wild_chimp</th>\n",
       "      <th>HR_wild_gorilla</th>\n",
       "      <th>NHR_2_wild_apes</th>\n",
       "      <th>NHR_human_2_wild_apes</th>\n",
       "      <th>NHR_human_single_wild_ape</th>\n",
       "      <th>Unique_CP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALL</th>\n",
       "      <td>6183.0</td>\n",
       "      <td>353.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>94.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP</th>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>94.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Prevotella</th>\n",
       "      <td>2450.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>55.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Bacteroides</th>\n",
       "      <td>940.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Unassigned</th>\n",
       "      <td>199.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>41.0</td>\n",
       "      <td>181.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>11.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Alistipes</th>\n",
       "      <td>515.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__RC9</th>\n",
       "      <td>363.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Parabacteroides</th>\n",
       "      <td>247.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Bacteroides_A</th>\n",
       "      <td>264.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Porphyromonas</th>\n",
       "      <td>243.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Prevotellamassilia</th>\n",
       "      <td>116.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>g__Bacteroides_B</th>\n",
       "      <td>110.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                       HR_human  HR_wild_bonobo  HR_wild_chimp  \\\n",
       "ALL                      6183.0           353.0          207.0   \n",
       "CP                         27.0             0.0            4.0   \n",
       "g__Prevotella            2450.0           203.0          114.0   \n",
       "g__Bacteroides            940.0             1.0           14.0   \n",
       "Unassigned                199.0            94.0           41.0   \n",
       "g__Alistipes              515.0             0.0            1.0   \n",
       "g__RC9                    363.0             0.0            0.0   \n",
       "g__Parabacteroides        247.0            38.0            3.0   \n",
       "g__Bacteroides_A          264.0             0.0           17.0   \n",
       "g__Porphyromonas          243.0             1.0            2.0   \n",
       "g__Prevotellamassilia     116.0             0.0            1.0   \n",
       "g__Bacteroides_B          110.0             0.0            0.0   \n",
       "\n",
       "                       HR_wild_gorilla  NHR_2_wild_apes  \\\n",
       "ALL                              258.0             14.0   \n",
       "CP                                 0.0              2.0   \n",
       "g__Prevotella                     46.0              1.0   \n",
       "g__Bacteroides                     0.0              0.0   \n",
       "Unassigned                       181.0             11.0   \n",
       "g__Alistipes                       0.0              0.0   \n",
       "g__RC9                             0.0              0.0   \n",
       "g__Parabacteroides                 1.0              1.0   \n",
       "g__Bacteroides_A                   2.0              0.0   \n",
       "g__Porphyromonas                   0.0              0.0   \n",
       "g__Prevotellamassilia              4.0              0.0   \n",
       "g__Bacteroides_B                   0.0              0.0   \n",
       "\n",
       "                       NHR_human_2_wild_apes  NHR_human_single_wild_ape  \\\n",
       "ALL                                      5.0                       34.0   \n",
       "CP                                       2.0                       10.0   \n",
       "g__Prevotella                            0.0                        2.0   \n",
       "g__Bacteroides                           5.0                       20.0   \n",
       "Unassigned                               0.0                        2.0   \n",
       "g__Alistipes                             0.0                        3.0   \n",
       "g__RC9                                   0.0                        0.0   \n",
       "g__Parabacteroides                       0.0                        3.0   \n",
       "g__Bacteroides_A                         0.0                        0.0   \n",
       "g__Porphyromonas                         0.0                        0.0   \n",
       "g__Prevotellamassilia                    0.0                        0.0   \n",
       "g__Bacteroides_B                         0.0                        3.0   \n",
       "\n",
       "                       Unique_CP  \n",
       "ALL                         94.0  \n",
       "CP                          94.0  \n",
       "g__Prevotella               55.0  \n",
       "g__Bacteroides              12.0  \n",
       "Unassigned                  11.0  \n",
       "g__Alistipes                 0.0  \n",
       "g__RC9                       0.0  \n",
       "g__Parabacteroides           3.0  \n",
       "g__Bacteroides_A             3.0  \n",
       "g__Porphyromonas             4.0  \n",
       "g__Prevotellamassilia        0.0  \n",
       "g__Bacteroides_B             0.0  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_gyrb = asv_hr_table(asv_table_file,metadata_file,tax_table_file)\n",
    "output_summary_table(asv_gyrb,'analyses/codiv_moeller_ASVs/gyrb_hr_asv_table.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CP', 'MX', 'human', 'wild_bonobo', 'wild_chimp', 'wild_gorilla']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CP</th>\n",
       "      <th>MX</th>\n",
       "      <th>human</th>\n",
       "      <th>wild_bonobo</th>\n",
       "      <th>wild_chimp</th>\n",
       "      <th>wild_gorilla</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALL</th>\n",
       "      <td>6.0</td>\n",
       "      <td>51.0</td>\n",
       "      <td>249.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>19.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP</th>\n",
       "      <td>6.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Bacteroides</th>\n",
       "      <td>4.0</td>\n",
       "      <td>25.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Prevotella</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>75.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tannerellaceae_Parabacteroides</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rikenellaceae_Alistipes</th>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Unassigned</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Bacteroides_B</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tannerellaceae_Unassigned</th>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WCHB1-69_Unassigned</th>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Porphyromonadaceae_Porphyromonas</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Bacteroides_A</th>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                   CP    MX  human  wild_bonobo  wild_chimp  \\\n",
       "ALL                               6.0  51.0  249.0         14.0        20.0   \n",
       "CP                                6.0  19.0   25.0          0.0         2.0   \n",
       "Bacteroidaceae_Bacteroides        4.0  25.0   75.0          0.0         1.0   \n",
       "Bacteroidaceae_Prevotella         0.0   5.0   75.0          8.0         9.0   \n",
       "Tannerellaceae_Parabacteroides    0.0   3.0   18.0          1.0         2.0   \n",
       "Rikenellaceae_Alistipes           0.0   5.0   17.0          0.0         0.0   \n",
       "Bacteroidaceae_Unassigned         0.0   1.0   13.0          3.0         1.0   \n",
       "Bacteroidaceae_Bacteroides_B      0.0   3.0   15.0          0.0         0.0   \n",
       "Tannerellaceae_Unassigned         0.0   2.0    3.0          0.0         4.0   \n",
       "WCHB1-69_Unassigned               0.0   3.0    0.0          0.0         0.0   \n",
       "Porphyromonadaceae_Porphyromonas  0.0   0.0    9.0          0.0         0.0   \n",
       "Bacteroidaceae_Bacteroides_A      0.0   1.0    6.0          0.0         1.0   \n",
       "\n",
       "                                  wild_gorilla  \n",
       "ALL                                       19.0  \n",
       "CP                                         0.0  \n",
       "Bacteroidaceae_Bacteroides                 0.0  \n",
       "Bacteroidaceae_Prevotella                  4.0  \n",
       "Tannerellaceae_Parabacteroides             0.0  \n",
       "Rikenellaceae_Alistipes                    0.0  \n",
       "Bacteroidaceae_Unassigned                  3.0  \n",
       "Bacteroidaceae_Bacteroides_B               0.0  \n",
       "Tannerellaceae_Unassigned                  1.0  \n",
       "WCHB1-69_Unassigned                        7.0  \n",
       "Porphyromonadaceae_Porphyromonas           0.0  \n",
       "Bacteroidaceae_Bacteroides_A               0.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clades_df=pd.read_csv('analyses/codiv_moeller_ASVs/full_tree_clades_collapsed_table.txt',sep='\\t')\n",
    "clades_df=clades_df.rename(columns={\"cladeTax\": \"Genus\"})\n",
    "output_summary_table(clades_df,'analyses/codiv_moeller_ASVs/gyrb_hr_clade_table.txt')\n",
    "#clades_df_15perc =  clades_df.loc[(clades_df.wild_bonobo > 0.15) | \n",
    "#              (clades_df.wild_chimp > 0.15)  |\n",
    "#              (clades_df.wild_gorilla > 0.15)|\n",
    "#              (clades_df.western_human > 0.15)|\n",
    "#              (clades_df.non_western_human > 0.15)] \n",
    "#output_summary_table(clades_df_15perc,'analyses/codiv_moeller_ASVs/gyrb_hr_clade_15perc_table.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'asv_gyrb' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-58-ca3012eb1b52>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     27\u001b[0m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0masv_cp_table\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     28\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 29\u001b[0;31m \u001b[0masv_cp_table\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mcaptive_apes_asv_summary\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0masv_gyrb\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mmetadata_file\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     30\u001b[0m \u001b[0masv_cp_table_summary\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0masv_cp_table\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mgroupby\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'numEnclosure'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'multi_site_sp'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msize\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mreset_index\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mname\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;34m\"count\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     31\u001b[0m \u001b[0mprint\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0masv_cp_table_summary\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mNameError\u001b[0m: name 'asv_gyrb' is not defined"
     ]
    }
   ],
   "source": [
    "def multi_site_sp(cp_desc):\n",
    "    sites = list(set([x.split('_')[2] for x in cp_desc]))\n",
    "    sitesNum = len(sites)\n",
    "    sitesMulti = 'multi_site' if sitesNum > 1 else 'single_site'      \n",
    "    species = list(set([x.split('_')[1] for x in cp_desc]))\n",
    "    speciesNum = len(species)\n",
    "    speciessMulti = 'multi_sp' if speciesNum > 1 else 'single_sp' \n",
    "    return(sitesMulti + '_' + speciessMulti)\n",
    "multi_site_sp(['captive_chimp_HOUZ','captive_gorilla_HOUZ'])\n",
    "\n",
    "def captive_apes_asv_summary(asv_full,metadata_file):\n",
    "    asv_cp = asv_full[asv_full['CP_pres']==True]\n",
    "    metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "    metadata['Description_site'] = metadata['Description']+'_' +metadata['site_code']\n",
    "    sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "\n",
    "    description_df = asv_cp['sampleNames'].apply(lambda l: pd.Series(\n",
    "        [sample_type_site_dict[name] for name in l]).value_counts())\n",
    "    description_df = description_df.fillna(0)  \n",
    "    capt_desc = list(set(metadata['Description_site'][metadata['captivity_status']=='captive']))\n",
    "    description_df = description_df[capt_desc]\n",
    "    description_df['CP_sp_loc'] = description_df.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "    description_df['numEnclosure'] = description_df['CP_sp_loc'].apply(lambda x: len(x))\n",
    "    description_df['multi_site_sp'] = description_df['CP_sp_loc'].apply(lambda x:  \n",
    "                                                                        multi_site_sp(x))\n",
    "    asv_cp_table = asv_cp.merge(description_df,left_index=True,right_index=True)\n",
    "    return(asv_cp_table)\n",
    "\n",
    "asv_cp_table = captive_apes_asv_summary(asv_gyrb,metadata_file)\n",
    "asv_cp_table_summary = asv_cp_table.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "print(asv_cp_table_summary)\n",
    "asv_cp_table_summary.to_csv('analyses/codiv_moeller_ASVs/numEnclosures_table.txt',sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "\"['cp.bon.COLZ.01.Bt', 'cp.bon.COLZ.02.Bt'] not in index\"",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-59-4d2d0fc95daf>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     21\u001b[0m     \u001b[0;32mreturn\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mres\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     22\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 23\u001b[0;31m \u001b[0mshared_ASVs_common_gen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m'cp.bon.COLZ.01.Bt'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0;34m'cp.bon.COLZ.02.Bt'\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0masv_table_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     24\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     25\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mprop_shared\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mrow\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m<ipython-input-59-4d2d0fc95daf>\u001b[0m in \u001b[0;36mshared_ASVs_common_gen\u001b[0;34m(ind1, ind2, asv_table_gen)\u001b[0m\n\u001b[1;32m      6\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m \u001b[0;32mdef\u001b[0m \u001b[0mshared_ASVs_common_gen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mind1\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mind2\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0masv_table_gen\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 8\u001b[0;31m     pw = asv_table_gen[['Phylum','Family','Genus','ASV',ind1,ind2]][\n\u001b[0m\u001b[1;32m      9\u001b[0m         \u001b[0;34m(\u001b[0m\u001b[0masv_table_gen\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mind1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m>\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m|\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0masv_table_gen\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mind2\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m>\u001b[0m\u001b[0;36m0\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     10\u001b[0m         ]\n",
      "\u001b[0;32m/usr/local/anaconda3/envs/ete/lib/python3.6/site-packages/pandas/core/frame.py\u001b[0m in \u001b[0;36m__getitem__\u001b[0;34m(self, key)\u001b[0m\n\u001b[1;32m   2804\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mis_iterator\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2805\u001b[0m                 \u001b[0mkey\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 2806\u001b[0;31m             \u001b[0mindexer\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mloc\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_listlike_indexer\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0maxis\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0;32mTrue\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   2807\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   2808\u001b[0m         \u001b[0;31m# take() does not accept boolean indexers\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/anaconda3/envs/ete/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_get_listlike_indexer\u001b[0;34m(self, key, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1551\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1552\u001b[0m         self._validate_read_indexer(\n\u001b[0;32m-> 1553\u001b[0;31m             \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mo\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_get_axis_number\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0maxis\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m=\u001b[0m\u001b[0mraise_missing\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1554\u001b[0m         )\n\u001b[1;32m   1555\u001b[0m         \u001b[0;32mreturn\u001b[0m \u001b[0mkeyarr\u001b[0m\u001b[0;34m,\u001b[0m \u001b[0mindexer\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m/usr/local/anaconda3/envs/ete/lib/python3.6/site-packages/pandas/core/indexing.py\u001b[0m in \u001b[0;36m_validate_read_indexer\u001b[0;34m(self, key, indexer, axis, raise_missing)\u001b[0m\n\u001b[1;32m   1644\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0;34m(\u001b[0m\u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mname\u001b[0m \u001b[0;34m==\u001b[0m \u001b[0;34m\"loc\"\u001b[0m \u001b[0;32mand\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0mraise_missing\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1645\u001b[0m                 \u001b[0mnot_found\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mlist\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mkey\u001b[0m\u001b[0;34m)\u001b[0m \u001b[0;34m-\u001b[0m \u001b[0mset\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0max\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m-> 1646\u001b[0;31m                 \u001b[0;32mraise\u001b[0m \u001b[0mKeyError\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34mf\"{not_found} not in index\"\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m   1647\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m   1648\u001b[0m             \u001b[0;31m# we skip the warning on Categorical/Interval\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyError\u001b[0m: \"['cp.bon.COLZ.01.Bt', 'cp.bon.COLZ.02.Bt'] not in index\""
     ]
    }
   ],
   "source": [
    "asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "asv_table = asv_table.reset_index().rename(columns={\"index\": \"ASV\"})\n",
    "tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "asv_table_gen = asv_table.merge(tax_table[['ASV','Phylum','Family','Genus']],on='ASV',how='left')\n",
    "asv_table_gen = asv_table_gen[asv_table_gen['Genus']!='Unassigned']\n",
    "\n",
    "def shared_ASVs_common_gen(ind1,ind2,asv_table_gen):\n",
    "    pw = asv_table_gen[['Phylum','Family','Genus','ASV',ind1,ind2]][\n",
    "        (asv_table_gen[ind1]>0)|(asv_table_gen[ind2]>0)\n",
    "        ]\n",
    "    #print(pw)\n",
    "    common_gen = list(\n",
    "                 set(pw[pw[ind1]>0]['Genus']) & \n",
    "                 set(pw[pw[ind2]>0]['Genus']))\n",
    "    pw= pw[pw['Genus'].isin(common_gen)]\n",
    "    ind1_ASVs = len(pw[(pw[ind1]>0)])\n",
    "    ind2_ASVs = len(pw[(pw[ind2]>0)])\n",
    "    common_ASVs = len(pw[(pw[ind1]>0)&(pw[ind2]>0)])\n",
    "    res = pd.Series([ind1,ind2,common_gen,ind1_ASVs,ind2_ASVs,common_ASVs],\n",
    "              index=['ind1','ind2','common_gen','ind1_ASVs','ind2_ASVs','common_ASVs'])\n",
    "    return(res)\n",
    "\n",
    "shared_ASVs_common_gen('cp.bon.COLZ.01.Bt','cp.bon.COLZ.02.Bt',asv_table_gen)\n",
    "\n",
    "def prop_shared(row):\n",
    "    if row['common_ASVs'] > 0:\n",
    "        return(row['common_ASVs']/min(row['ind1_ASVs'],row['ind2_ASVs']))\n",
    "    else:\n",
    "        return(0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                  ind1                 ind2          desc_site_ind1  \\\n",
      "0  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.115.16s  captive_orangutan_ATLZ   \n",
      "1  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.116.16s  captive_orangutan_ATLZ   \n",
      "2  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.117.16s  captive_orangutan_ATLZ   \n",
      "3  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.118.16s  captive_orangutan_ATLZ   \n",
      "4  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.119.16s  captive_orangutan_ATLZ   \n",
      "\n",
      "           desc_site_ind2                                    full_desc_comp  \\\n",
      "0  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
      "1  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
      "2  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
      "3  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
      "4  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
      "\n",
      "          sp_site_comp  \n",
      "0  same_spec_same_site  \n",
      "1  same_spec_same_site  \n",
      "2  same_spec_same_site  \n",
      "3  same_spec_same_site  \n",
      "4  same_spec_same_site  \n",
      "1977\n"
     ]
    }
   ],
   "source": [
    "from itertools import combinations\n",
    "\n",
    "def get_sp_site_comp(ind1,ind2):\n",
    "    sp = ('same_spec' if ind1.split('_')[1] == ind2.split('_')[1] else 'diff_spec')\n",
    "    st = ('same_site' if ind1.split('_')[2] == ind2.split('_')[2] else 'diff_site')\n",
    "    return(sp+'_'+st)\n",
    "get_sp_site_comp('captive_bonobo_COLZ','captive_chimp_COLZ')\n",
    "\n",
    "\n",
    "def pw_metadata_capt_samples(metadata_file):\n",
    "    metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "    metadata['Description_site'] = metadata['Description']+'_' +metadata['site_code']\n",
    "    sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "\n",
    "    cp_samples = metadata['X.SampleID'][metadata['captivity_status']=='captive']\n",
    "    pw_df = pd.DataFrame(combinations(cp_samples, 2),columns=['ind1','ind2'])\n",
    "    pw_df['desc_site_ind1'] = pw_df['ind1'].apply(lambda x:sample_type_site_dict[x])\n",
    "    pw_df['desc_site_ind2'] = pw_df['ind2'].apply(lambda x:sample_type_site_dict[x])\n",
    "    pw_df['full_desc_comp'] = pw_df['desc_site_ind1'] + '_vs_' + pw_df['desc_site_ind2']\n",
    "    pw_df['sp_site_comp'] = pw_df.apply(lambda row: \n",
    "                                        get_sp_site_comp(row['desc_site_ind1'],row['desc_site_ind2']),\n",
    "                                        axis=1)\n",
    "    return(pw_df)\n",
    "\n",
    "def add_tax_to_asv_table(asv_table_file,tax_table_file):\n",
    "    asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "    asv_table = asv_table.reset_index().rename(columns={\"index\": \"ASV\"})\n",
    "    tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "    asv_table_tax = asv_table.merge(tax_table[['ASV','Phylum','Order','Family','Genus']],on='ASV',how='left')\n",
    "    return(asv_table_tax)\n",
    "    \n",
    "asv_table_tax = add_tax_to_asv_table(asv_table_file,tax_table_file)\n",
    "print(len(asv_table_tax))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [],
   "source": [
    "pw_df = pw_metadata_capt_samples(metadata_file)\n",
    "asv_table_tax = add_tax_to_asv_table(asv_table_file,tax_table_file)\n",
    "\n",
    "asv_table_tax = asv_table_tax[asv_table_tax['Genus']!='Unassigned']\n",
    "shared_ASV_all_genera = pw_df.apply(lambda row: \n",
    "                    shared_ASVs_common_gen(row['ind1'],row['ind2'],asv_table_tax),axis=1)\n",
    "pw_shared_ASV_all_genera = pw_df.merge(shared_ASV_all_genera,on=['ind1','ind2'])                       \n",
    "pw_shared_ASV_all_genera['prop_shared_ASVs'] = pw_shared_ASV_all_genera.apply(\n",
    "    lambda row: prop_shared(row),axis=1)\n",
    "pw_shared_ASV_all_genera.to_csv(\n",
    "    'analyses/codiv_moeller_ASVs/gyrb_prop_shared_ASVs_all_gen_table.txt',\n",
    "    sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_table_Prevotella = asv_table_gen[asv_table_gen['Genus']=='g__Prevotella']\n",
    "shared_ASV_Prevotella = pw_df.apply(lambda row: \n",
    "                    shared_ASVs_common_gen(row['ind1'],row['ind2'],asv_table_Prevotella),axis=1)\n",
    "pw_shared_ASV_Prevotella = pw_df.merge(shared_ASV_Prevotella,on=['ind1','ind2'])\n",
    "pw_shared_ASV_Prevotella['prop_shared_ASVs'] = pw_shared_ASV_Prevotella.apply(\n",
    "    lambda row: prop_shared(row),axis=1)\n",
    "pw_shared_ASV_all_genera.to_csv('analyses/codiv_moeller_ASVs/gyrb_prop_shared_ASVs_Prevotella_table.txt',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 16S"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/results/16s/')\n",
    "metadata_file = 'inputs/16S_metadata.txt'\n",
    "tax_table_file = 'inputs/ASVs_taxonomy.txt'\n",
    "asv_table_file = 'inputs/ASV_tab.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HR_human', 'HR_wild_bonobo', 'HR_wild_chimp', 'HR_wild_gorilla', 'NHR_2_wild_apes', 'NHR_3_wild_apes', 'NHR_4_hominids', 'NHR_human_2_wild_apes', 'NHR_human_single_wild_ape', 'Unique_CP']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HR_human</th>\n",
       "      <th>HR_wild_bonobo</th>\n",
       "      <th>HR_wild_chimp</th>\n",
       "      <th>HR_wild_gorilla</th>\n",
       "      <th>NHR_2_wild_apes</th>\n",
       "      <th>NHR_3_wild_apes</th>\n",
       "      <th>NHR_4_hominids</th>\n",
       "      <th>NHR_human_2_wild_apes</th>\n",
       "      <th>NHR_human_single_wild_ape</th>\n",
       "      <th>Unique_CP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALL</th>\n",
       "      <td>553.0</td>\n",
       "      <td>57.0</td>\n",
       "      <td>105.0</td>\n",
       "      <td>210.0</td>\n",
       "      <td>227.0</td>\n",
       "      <td>104.0</td>\n",
       "      <td>189.0</td>\n",
       "      <td>157.0</td>\n",
       "      <td>269.0</td>\n",
       "      <td>106.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP</th>\n",
       "      <td>227.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>166.0</td>\n",
       "      <td>128.0</td>\n",
       "      <td>187.0</td>\n",
       "      <td>106.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ruminococcaceae_UCG-014</th>\n",
       "      <td>18.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>19.0</td>\n",
       "      <td>11.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella_9</th>\n",
       "      <td>38.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroides</th>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rikenellaceae_RC9_gut_group</th>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Treponema_2</th>\n",
       "      <td>14.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>12.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Faecalibacterium</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ruminococcus_1</th>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>7.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ruminococcaceae_UCG-005</th>\n",
       "      <td>7.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Ruminococcaceae_UCG-010</th>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella_7</th>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             HR_human  HR_wild_bonobo  HR_wild_chimp  \\\n",
       "ALL                             553.0            57.0          105.0   \n",
       "CP                              227.0             0.0           10.0   \n",
       "Ruminococcaceae_UCG-014          18.0             0.0            1.0   \n",
       "Prevotella_9                     38.0             2.0            7.0   \n",
       "Bacteroides                      33.0             0.0            0.0   \n",
       "Rikenellaceae_RC9_gut_group      16.0             0.0            7.0   \n",
       "Treponema_2                      14.0             0.0            4.0   \n",
       "Faecalibacterium                 15.0             0.0            2.0   \n",
       "Ruminococcus_1                    6.0             0.0            1.0   \n",
       "Ruminococcaceae_UCG-005           7.0             0.0            2.0   \n",
       "Ruminococcaceae_UCG-010           2.0             0.0            0.0   \n",
       "Prevotella_7                      4.0             2.0            7.0   \n",
       "\n",
       "                             HR_wild_gorilla  NHR_2_wild_apes  \\\n",
       "ALL                                    210.0            227.0   \n",
       "CP                                       8.0             20.0   \n",
       "Ruminococcaceae_UCG-014                  1.0              2.0   \n",
       "Prevotella_9                             2.0              1.0   \n",
       "Bacteroides                              0.0              0.0   \n",
       "Rikenellaceae_RC9_gut_group              8.0              8.0   \n",
       "Treponema_2                              6.0              1.0   \n",
       "Faecalibacterium                         1.0              1.0   \n",
       "Ruminococcus_1                           0.0              2.0   \n",
       "Ruminococcaceae_UCG-005                  1.0              1.0   \n",
       "Ruminococcaceae_UCG-010                  1.0              0.0   \n",
       "Prevotella_7                             6.0              2.0   \n",
       "\n",
       "                             NHR_3_wild_apes  NHR_4_hominids  \\\n",
       "ALL                                    104.0           189.0   \n",
       "CP                                      19.0           166.0   \n",
       "Ruminococcaceae_UCG-014                  3.0            19.0   \n",
       "Prevotella_9                             1.0             0.0   \n",
       "Bacteroides                              0.0             1.0   \n",
       "Rikenellaceae_RC9_gut_group              1.0             1.0   \n",
       "Treponema_2                              1.0             0.0   \n",
       "Faecalibacterium                         1.0             3.0   \n",
       "Ruminococcus_1                           0.0             7.0   \n",
       "Ruminococcaceae_UCG-005                  1.0             5.0   \n",
       "Ruminococcaceae_UCG-010                  0.0             7.0   \n",
       "Prevotella_7                             1.0             0.0   \n",
       "\n",
       "                             NHR_human_2_wild_apes  NHR_human_single_wild_ape  \\\n",
       "ALL                                          157.0                      269.0   \n",
       "CP                                           128.0                      187.0   \n",
       "Ruminococcaceae_UCG-014                       11.0                       14.0   \n",
       "Prevotella_9                                   0.0                        9.0   \n",
       "Bacteroides                                    2.0                       13.0   \n",
       "Rikenellaceae_RC9_gut_group                    1.0                        2.0   \n",
       "Treponema_2                                    0.0                        3.0   \n",
       "Faecalibacterium                               6.0                        9.0   \n",
       "Ruminococcus_1                                 9.0                        4.0   \n",
       "Ruminococcaceae_UCG-005                        4.0                        8.0   \n",
       "Ruminococcaceae_UCG-010                        6.0                        7.0   \n",
       "Prevotella_7                                   0.0                        1.0   \n",
       "\n",
       "                             Unique_CP  \n",
       "ALL                              106.0  \n",
       "CP                               106.0  \n",
       "Ruminococcaceae_UCG-014            3.0  \n",
       "Prevotella_9                       0.0  \n",
       "Bacteroides                        0.0  \n",
       "Rikenellaceae_RC9_gut_group        0.0  \n",
       "Treponema_2                       12.0  \n",
       "Faecalibacterium                   1.0  \n",
       "Ruminococcus_1                     7.0  \n",
       "Ruminococcaceae_UCG-005            1.0  \n",
       "Ruminococcaceae_UCG-010            2.0  \n",
       "Prevotella_7                       0.0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_16s = asv_hr_table(asv_table_file,metadata_file,tax_table_file)\n",
    "asv_16s.head()\n",
    "\n",
    "output_summary_table(asv_16s,'analyses/tables/16s_hr_asv_table.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "HR_Human_ASVs = asv_16s[asv_16s['HR_type']=='HR_human']\n",
    "#how many captive ape species are they present in?\n",
    "HR_Human_ASVs['CP_sampleTypes'].apply(lambda x: len(x)).value_counts()\n",
    "#how many captive ape species are they present in?\n",
    "\n",
    "HR_wild_ASVs = asv_16s[asv_16s['HR_type'].isin(['HR_wild_bonobo','HR_wild_chimp','HR_wild_gorilla'])]\n",
    "\n",
    "HR_wild_ASVs_CP = HR_wild_ASVs[HR_wild_ASVs['CP_pres']==True]\n",
    "HR_wild_ASVs_CP['HR_type'].value_counts()\n",
    "HR_wild_ASVs_CP['CP_sampleTypes'].apply(lambda x: len(x)).value_counts()\n",
    "HR_wild_ASVs_CP.to_csv('analyses/tables/16s_hr_in_cp_asv_table.txt',sep='\\t')\n",
    "#HR_wild_ASVs_CP['HR_type','CP_sampleTypes']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['HR_human', 'HR_wild_bonobo', 'HR_wild_chimp', 'HR_wild_gorilla', 'NHR_2_wild_apes', 'NHR_3_wild_apes', 'NHR_4_hominids', 'NHR_human_2_wild_apes', 'NHR_human_single_wild_ape', 'Unique_CP']\n",
      "['HR_human', 'NHR_4_hominids', 'NHR_human_2_wild_apes', 'NHR_human_single_wild_ape']\n",
      "['HR_human', 'HR_wild_bonobo', 'HR_wild_chimp', 'HR_wild_gorilla', 'NHR_2_wild_apes', 'NHR_3_wild_apes', 'NHR_4_hominids', 'NHR_human_2_wild_apes', 'NHR_human_single_wild_ape', 'Unique_CP']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HR_human</th>\n",
       "      <th>HR_wild_bonobo</th>\n",
       "      <th>HR_wild_chimp</th>\n",
       "      <th>HR_wild_gorilla</th>\n",
       "      <th>NHR_2_wild_apes</th>\n",
       "      <th>NHR_3_wild_apes</th>\n",
       "      <th>NHR_4_hominids</th>\n",
       "      <th>NHR_human_2_wild_apes</th>\n",
       "      <th>NHR_human_single_wild_ape</th>\n",
       "      <th>Unique_CP</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALL</th>\n",
       "      <td>211.0</td>\n",
       "      <td>22.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>24.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>50.0</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP</th>\n",
       "      <td>83.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>28.0</td>\n",
       "      <td>29.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella_9</th>\n",
       "      <td>38.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroides</th>\n",
       "      <td>33.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rikenellaceae_RC9_gut_group</th>\n",
       "      <td>16.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella_7</th>\n",
       "      <td>4.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alloprevotella</th>\n",
       "      <td>9.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella_2</th>\n",
       "      <td>20.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parabacteroides</th>\n",
       "      <td>6.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Alistipes</th>\n",
       "      <td>12.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotellaceae_UCG-001</th>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotellaceae_NK3B31_group</th>\n",
       "      <td>5.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                             HR_human  HR_wild_bonobo  HR_wild_chimp  \\\n",
       "ALL                             211.0            22.0           50.0   \n",
       "CP                               83.0             0.0            0.0   \n",
       "Prevotella_9                     38.0             2.0            7.0   \n",
       "Bacteroides                      33.0             0.0            0.0   \n",
       "Rikenellaceae_RC9_gut_group      16.0             0.0            7.0   \n",
       "Prevotella_7                      4.0             2.0            7.0   \n",
       "Alloprevotella                    9.0             2.0            4.0   \n",
       "Prevotella_2                     20.0             0.0            0.0   \n",
       "Parabacteroides                   6.0             1.0            0.0   \n",
       "Alistipes                        12.0             0.0            0.0   \n",
       "Prevotellaceae_UCG-001            3.0             1.0            4.0   \n",
       "Prevotellaceae_NK3B31_group       5.0             1.0            1.0   \n",
       "\n",
       "                             HR_wild_gorilla  NHR_2_wild_apes  \\\n",
       "ALL                                     34.0             24.0   \n",
       "CP                                       2.0              3.0   \n",
       "Prevotella_9                             2.0              1.0   \n",
       "Bacteroides                              0.0              0.0   \n",
       "Rikenellaceae_RC9_gut_group              8.0              8.0   \n",
       "Prevotella_7                             6.0              2.0   \n",
       "Alloprevotella                           1.0              2.0   \n",
       "Prevotella_2                             0.0              0.0   \n",
       "Parabacteroides                          0.0              0.0   \n",
       "Alistipes                                0.0              0.0   \n",
       "Prevotellaceae_UCG-001                   0.0              1.0   \n",
       "Prevotellaceae_NK3B31_group              2.0              0.0   \n",
       "\n",
       "                             NHR_3_wild_apes  NHR_4_hominids  \\\n",
       "ALL                                      8.0             2.0   \n",
       "CP                                       0.0             1.0   \n",
       "Prevotella_9                             1.0             0.0   \n",
       "Bacteroides                              0.0             1.0   \n",
       "Rikenellaceae_RC9_gut_group              1.0             1.0   \n",
       "Prevotella_7                             1.0             0.0   \n",
       "Alloprevotella                           0.0             0.0   \n",
       "Prevotella_2                             0.0             0.0   \n",
       "Parabacteroides                          1.0             0.0   \n",
       "Alistipes                                0.0             0.0   \n",
       "Prevotellaceae_UCG-001                   0.0             0.0   \n",
       "Prevotellaceae_NK3B31_group              0.0             0.0   \n",
       "\n",
       "                             NHR_human_2_wild_apes  NHR_human_single_wild_ape  \\\n",
       "ALL                                            4.0                       50.0   \n",
       "CP                                             3.0                       28.0   \n",
       "Prevotella_9                                   0.0                        9.0   \n",
       "Bacteroides                                    2.0                       13.0   \n",
       "Rikenellaceae_RC9_gut_group                    1.0                        2.0   \n",
       "Prevotella_7                                   0.0                        1.0   \n",
       "Alloprevotella                                 0.0                        4.0   \n",
       "Prevotella_2                                   0.0                        1.0   \n",
       "Parabacteroides                                0.0                        8.0   \n",
       "Alistipes                                      1.0                        2.0   \n",
       "Prevotellaceae_UCG-001                         0.0                        0.0   \n",
       "Prevotellaceae_NK3B31_group                    0.0                        2.0   \n",
       "\n",
       "                             Unique_CP  \n",
       "ALL                               29.0  \n",
       "CP                                29.0  \n",
       "Prevotella_9                       0.0  \n",
       "Bacteroides                        0.0  \n",
       "Rikenellaceae_RC9_gut_group        0.0  \n",
       "Prevotella_7                       0.0  \n",
       "Alloprevotella                     0.0  \n",
       "Prevotella_2                       0.0  \n",
       "Parabacteroides                    0.0  \n",
       "Alistipes                          0.0  \n",
       "Prevotellaceae_UCG-001             2.0  \n",
       "Prevotellaceae_NK3B31_group        0.0  "
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_16s = asv_hr_table(asv_table_file,metadata_file,tax_table_file)\n",
    "output_summary_table(asv_16s,'analyses/tables/16s_hr_all_asvs_table.txt')\n",
    "\n",
    "asv_Bifidobacteriales = asv_16s[asv_16s['Order']=='Bifidobacteriales']\n",
    "output_summary_table(asv_Bifidobacteriales,'analyses/tables/16s_hr_Bifidobacteriales_asvs_table.txt')\n",
    "\n",
    "asv_Bacteroidales = asv_16s[asv_16s['Order']=='Bacteroidales']\n",
    "output_summary_table(asv_Bacteroidales,'analyses/tables/16s_hr_Bacteroidales_asvs_table.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "multi_site_multi_sp      398\n",
      "single_site_single_sp     86\n",
      "multi_site_single_sp      24\n",
      "single_site_multi_sp      12\n",
      "Name: multi_site_sp, dtype: int64\n",
      "multi_site_multi_sp      165\n",
      "single_site_single_sp     64\n",
      "multi_site_single_sp      10\n",
      "single_site_multi_sp       6\n",
      "Name: multi_site_sp, dtype: int64\n",
      "multi_site_multi_sp      70\n",
      "single_site_single_sp    18\n",
      "single_site_multi_sp     12\n",
      "multi_site_single_sp      6\n",
      "Name: multi_site_sp, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(asv_cp_table[asv_cp_table['HR_cat']=='NHR'][\"multi_site_sp\"].value_counts())\n",
    "print(asv_cp_table[asv_cp_table['HR_cat']=='HR'][\"multi_site_sp\"].value_counts())\n",
    "print(asv_cp_table[asv_cp_table['HR_cat']=='Unique_CP'][\"multi_site_sp\"].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "    numEnclosure          multi_site_sp  count\n",
      "0              1  single_site_single_sp    168\n",
      "1              2    multi_site_multi_sp     71\n",
      "2              2   multi_site_single_sp     30\n",
      "3              2   single_site_multi_sp     19\n",
      "4              3    multi_site_multi_sp     92\n",
      "5              3   multi_site_single_sp     10\n",
      "6              3   single_site_multi_sp     11\n",
      "7              4    multi_site_multi_sp     72\n",
      "8              5    multi_site_multi_sp     72\n",
      "9              6    multi_site_multi_sp     63\n",
      "10             7    multi_site_multi_sp     56\n",
      "11             8    multi_site_multi_sp     66\n",
      "12             9    multi_site_multi_sp     66\n",
      "13            10    multi_site_multi_sp     75\n",
      "   numEnclosure        multi_site_sp  count\n",
      "0             2  multi_site_multi_sp      2\n",
      "    numEnclosure          multi_site_sp  count\n",
      "0              1  single_site_single_sp     30\n",
      "1              2    multi_site_multi_sp     11\n",
      "2              2   multi_site_single_sp      5\n",
      "3              2   single_site_multi_sp      4\n",
      "4              3    multi_site_multi_sp     13\n",
      "5              3   multi_site_single_sp      2\n",
      "6              3   single_site_multi_sp      2\n",
      "7              4    multi_site_multi_sp     17\n",
      "8              5    multi_site_multi_sp     10\n",
      "9              6    multi_site_multi_sp     12\n",
      "10             7    multi_site_multi_sp     10\n",
      "11             8    multi_site_multi_sp     11\n",
      "12             9    multi_site_multi_sp     14\n",
      "13            10    multi_site_multi_sp      8\n"
     ]
    }
   ],
   "source": [
    "asv_cp_table = captive_apes_asv_summary(asv_16s,metadata_file)\n",
    "asv_cp_table_summary = asv_cp_table.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "print(asv_cp_table_summary)\n",
    "asv_cp_table_summary.to_csv('analyses/tables/numEnclosures_table.txt',sep='\\t',index=False)\n",
    "\n",
    "cp_Bifidobacteriales = asv_cp_table[asv_cp_table['Order']=='Bifidobacteriales']\n",
    "cp_Bifidobacteriales_summary = cp_Bifidobacteriales.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "print(cp_Bifidobacteriales_summary)\n",
    "cp_Bifidobacteriales_summary.to_csv('analyses/tables/numEnclosures_Bifidobacteriales_table.txt',sep='\\t',index=False)\n",
    "\n",
    "cp_Bacteroidales = asv_cp_table[asv_cp_table['Order']=='Bacteroidales']\n",
    "cp_Bacteroidales_summary = cp_Bacteroidales.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "print(cp_Bacteroidales_summary)\n",
    "cp_Bacteroidales_summary.to_csv('analyses/tables/numEnclosures_Bacteroidales_table.txt',sep='\\t',index=False) \n",
    "\n",
    "Prevotella_genera = ['Prevotella','Prevotella_2','Prevotella_7','Prevotella_1', 'Prevotella_9']\n",
    "cp_Prevotella = asv_cp_table[asv_cp_table['Genus'].isin(Prevotella_genera)]\n",
    "cp_Prevotella_summary = cp_Prevotella.groupby(['numEnclosure','multi_site_sp']).size().reset_index(name=\"count\")\n",
    "cp_Prevotella_summary.to_csv('analyses/tables/numEnclosures_Prevotella_table.txt',sep='\\t',index=False) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>sampleNames</th>\n",
       "      <th>sampleNum</th>\n",
       "      <th>HR_sampleTypes</th>\n",
       "      <th>HR_sampleNum</th>\n",
       "      <th>HR_cat</th>\n",
       "      <th>HR_type</th>\n",
       "      <th>CP_pres</th>\n",
       "      <th>CP_sampleTypes</th>\n",
       "      <th>CP_sampleNum</th>\n",
       "      <th>...</th>\n",
       "      <th>captive_orangutan_ATLZ</th>\n",
       "      <th>captive_orangutan_HOUZ</th>\n",
       "      <th>captive_orangutan_COLZ</th>\n",
       "      <th>captive_gorilla_HOUZ</th>\n",
       "      <th>captive_bonobo_COLZ</th>\n",
       "      <th>captive_chimp_HOUZ</th>\n",
       "      <th>captive_gorilla_COLZ</th>\n",
       "      <th>CP_sp_loc</th>\n",
       "      <th>numEnclosure</th>\n",
       "      <th>multi_site_sp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>ASV_224</td>\n",
       "      <td>[cp.ora.ATLZ.114.16s, cp.ora.ATLZ.115.16s, cp....</td>\n",
       "      <td>293</td>\n",
       "      <td>[wild_gorilla, wild_chimp, wild_bonobo, human]</td>\n",
       "      <td>238</td>\n",
       "      <td>NHR</td>\n",
       "      <td>NHR_4_hominids</td>\n",
       "      <td>True</td>\n",
       "      <td>[captive_gorilla, captive_orangutan, captive_b...</td>\n",
       "      <td>55</td>\n",
       "      <td>...</td>\n",
       "      <td>6.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>[captive_orangutan_COMZ, captive_chimp_PC, cap...</td>\n",
       "      <td>8</td>\n",
       "      <td>multi_site_multi_sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>ASV_840</td>\n",
       "      <td>[cp.ora.COMZ.MB31.16s, cp.ora.COMZ.MB7.16s, cp...</td>\n",
       "      <td>37</td>\n",
       "      <td>[human]</td>\n",
       "      <td>34</td>\n",
       "      <td>HR</td>\n",
       "      <td>HR_human</td>\n",
       "      <td>True</td>\n",
       "      <td>[captive_orangutan]</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[captive_orangutan_COMZ]</td>\n",
       "      <td>1</td>\n",
       "      <td>single_site_single_sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>ASV_1451</td>\n",
       "      <td>[cp.ora.ATLZ.117.16s, cp.ora.ATLZ.118.16s, cp....</td>\n",
       "      <td>34</td>\n",
       "      <td>[wild_gorilla, wild_chimp, human]</td>\n",
       "      <td>25</td>\n",
       "      <td>NHR</td>\n",
       "      <td>NHR_human_2_wild_apes</td>\n",
       "      <td>True</td>\n",
       "      <td>[captive_gorilla, captive_orangutan, captive_c...</td>\n",
       "      <td>9</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>[captive_chimp_PC, captive_orangutan_ATLZ, cap...</td>\n",
       "      <td>5</td>\n",
       "      <td>multi_site_multi_sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>ASV_546</td>\n",
       "      <td>[cp.ora.ATLZ.114.16s, cp.ora.ATLZ.115.16s, cp....</td>\n",
       "      <td>102</td>\n",
       "      <td>[wild_chimp, human]</td>\n",
       "      <td>48</td>\n",
       "      <td>NHR</td>\n",
       "      <td>NHR_human_single_wild_ape</td>\n",
       "      <td>True</td>\n",
       "      <td>[captive_gorilla, captive_orangutan, captive_b...</td>\n",
       "      <td>54</td>\n",
       "      <td>...</td>\n",
       "      <td>8.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>[captive_gorilla_COMZ, captive_orangutan_COMZ,...</td>\n",
       "      <td>9</td>\n",
       "      <td>multi_site_multi_sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>ASV_1846</td>\n",
       "      <td>[cp.ora.ATLZ.116.16s, cp.ora.ATLZ.123.16s, wd....</td>\n",
       "      <td>23</td>\n",
       "      <td>[wild_chimp, human]</td>\n",
       "      <td>20</td>\n",
       "      <td>NHR</td>\n",
       "      <td>NHR_human_single_wild_ape</td>\n",
       "      <td>True</td>\n",
       "      <td>[captive_orangutan]</td>\n",
       "      <td>3</td>\n",
       "      <td>...</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[captive_orangutan_ATLZ, captive_orangutan_COLZ]</td>\n",
       "      <td>2</td>\n",
       "      <td>multi_site_single_sp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 27 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ASV                                        sampleNames  sampleNum  \\\n",
       "5   ASV_224  [cp.ora.ATLZ.114.16s, cp.ora.ATLZ.115.16s, cp....        293   \n",
       "6   ASV_840  [cp.ora.COMZ.MB31.16s, cp.ora.COMZ.MB7.16s, cp...         37   \n",
       "7  ASV_1451  [cp.ora.ATLZ.117.16s, cp.ora.ATLZ.118.16s, cp....         34   \n",
       "8   ASV_546  [cp.ora.ATLZ.114.16s, cp.ora.ATLZ.115.16s, cp....        102   \n",
       "9  ASV_1846  [cp.ora.ATLZ.116.16s, cp.ora.ATLZ.123.16s, wd....         23   \n",
       "\n",
       "                                   HR_sampleTypes  HR_sampleNum HR_cat  \\\n",
       "5  [wild_gorilla, wild_chimp, wild_bonobo, human]           238    NHR   \n",
       "6                                         [human]            34     HR   \n",
       "7               [wild_gorilla, wild_chimp, human]            25    NHR   \n",
       "8                             [wild_chimp, human]            48    NHR   \n",
       "9                             [wild_chimp, human]            20    NHR   \n",
       "\n",
       "                     HR_type  CP_pres  \\\n",
       "5             NHR_4_hominids     True   \n",
       "6                   HR_human     True   \n",
       "7      NHR_human_2_wild_apes     True   \n",
       "8  NHR_human_single_wild_ape     True   \n",
       "9  NHR_human_single_wild_ape     True   \n",
       "\n",
       "                                      CP_sampleTypes  CP_sampleNum  ...  \\\n",
       "5  [captive_gorilla, captive_orangutan, captive_b...            55  ...   \n",
       "6                                [captive_orangutan]             3  ...   \n",
       "7  [captive_gorilla, captive_orangutan, captive_c...             9  ...   \n",
       "8  [captive_gorilla, captive_orangutan, captive_b...            54  ...   \n",
       "9                                [captive_orangutan]             3  ...   \n",
       "\n",
       "  captive_orangutan_ATLZ captive_orangutan_HOUZ captive_orangutan_COLZ  \\\n",
       "5                    6.0                    2.0                    0.0   \n",
       "6                    0.0                    0.0                    0.0   \n",
       "7                    4.0                    0.0                    0.0   \n",
       "8                    8.0                    6.0                    0.0   \n",
       "9                    2.0                    0.0                    1.0   \n",
       "\n",
       "  captive_gorilla_HOUZ  captive_bonobo_COLZ  captive_chimp_HOUZ  \\\n",
       "5                  7.0                  5.0                 8.0   \n",
       "6                  0.0                  0.0                 0.0   \n",
       "7                  1.0                  0.0                 1.0   \n",
       "8                  5.0                  6.0                 6.0   \n",
       "9                  0.0                  0.0                 0.0   \n",
       "\n",
       "   captive_gorilla_COLZ                                          CP_sp_loc  \\\n",
       "5                   8.0  [captive_orangutan_COMZ, captive_chimp_PC, cap...   \n",
       "6                   0.0                           [captive_orangutan_COMZ]   \n",
       "7                   1.0  [captive_chimp_PC, captive_orangutan_ATLZ, cap...   \n",
       "8                   4.0  [captive_gorilla_COMZ, captive_orangutan_COMZ,...   \n",
       "9                   0.0   [captive_orangutan_ATLZ, captive_orangutan_COLZ]   \n",
       "\n",
       "   numEnclosure          multi_site_sp  \n",
       "5             8    multi_site_multi_sp  \n",
       "6             1  single_site_single_sp  \n",
       "7             5    multi_site_multi_sp  \n",
       "8             9    multi_site_multi_sp  \n",
       "9             2   multi_site_single_sp  \n",
       "\n",
       "[5 rows x 27 columns]"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_cp_table.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [],
   "source": [
    "def output_proportion_shared_ASVs(pw_df,asv_table_tax):\n",
    "    shared = pw_df.apply(lambda row: \n",
    "                    shared_ASVs_common_gen(row['ind1'],row['ind2'],asv_table_tax),axis=1)\n",
    "    pw_shared = pw_df.merge(shared,on=['ind1','ind2'])                       #pw_shared_ASV_all_genera\n",
    "\n",
    "    pw_shared['prop_shared_ASVs'] = pw_shared.apply(\n",
    "        lambda row: prop_shared(row),axis=1)\n",
    "    return(pw_shared)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [],
   "source": [
    "pw_df = pw_metadata_capt_samples(metadata_file)\n",
    "asv_table_tax = add_tax_to_asv_table(asv_table_file,tax_table_file)\n",
    "asv_table_tax = asv_table_tax[~asv_table_tax['Genus'].isna()]\n",
    "\n",
    "\n",
    "pw_shared_ASV_all_genera = output_proportion_shared_ASVs(pw_df,asv_table_tax)\n",
    "pw_shared_ASV_all_genera.to_csv(\n",
    "    'analyses/tables/16s_prop_shared_ASVs_all_gen_table.txt',\n",
    "    sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_table_Bacteroidales = asv_table_tax[asv_table_tax['Order']=='Bacteroidales']\n",
    "pw_shared_ASV_Bacteroidales = output_proportion_shared_ASVs(pw_df,asv_table_Bacteroidales)\n",
    "pw_shared_ASV_Bacteroidales.to_csv(\n",
    "    'analyses/tables/16s_prop_shared_ASVs_Bacteroidales_table.txt',\n",
    "    sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'Prevotella', 'Alloprevotella', 'Prevotellaceae_UCG-003', 'Prevotella_2', 'dgA-11_gut_group', 'Bacteroides', 'Barnesiella', 'Butyricimonas', 'Prevotellaceae_UCG-001', 'Prevotella_7', 'Rikenellaceae_RC9_gut_group', 'Prevotellaceae_NK3B31_group', 'Prevotellaceae_UCG-004', 'Alistipes', 'Prevotellaceae_Ga6A1_group', 'Coprobacter', 'CAG-873', 'Parabacteroides', 'Prevotella_1', 'Prevotella_9', 'Paraprevotella', 'Odoribacter'}\n"
     ]
    }
   ],
   "source": [
    "print(set(asv_table_Bacteroidales['Genus']))\n",
    "Prevotella_genera = ['Prevotella','Prevotella_2','Prevotella_7','Prevotella_1', 'Prevotella_9']\n",
    "asv_table_Prevotella = asv_table_tax[asv_table_tax['Genus'].isin(Prevotella_genera)]\n",
    "pw_shared_ASV_Prevotella = output_proportion_shared_ASVs(pw_df,asv_table_Prevotella)\n",
    "pw_shared_ASV_Prevotella.to_csv(\n",
    "    'analyses/tables/16s_prop_shared_ASVs_Prevotella_table.txt',\n",
    "    sep='\\t',index=False)\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [],
   "source": [
    "def prop_shared_by_order(order):\n",
    "    asv_table_order = asv_table_tax[asv_table_tax['Order']==order]\n",
    "    asv_table_order = output_proportion_shared_ASVs(pw_df,asv_table_order)\n",
    "    asv_table_order['common_gen'].apply(lambda x: len(x)) #remove comparisons where two ind don't share gen\n",
    "    asv_table_order = asv_table_order[asv_table_order['common_gen'].apply(lambda x: len(x)) > 0]\n",
    "    asv_table_order['order']=order\n",
    "    return(asv_table_order)\n",
    "\n",
    "Clostridales = prop_shared_by_order('Clostridiales')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 200,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Index(['Clostridiales', 'Bacteroidales', 'Mollicutes_RF39',\n",
      "       'Erysipelotrichales', 'Spirochaetales', 'Gastranaerophilales',\n",
      "       'Selenomonadales', 'Betaproteobacteriales', 'Coriobacteriales',\n",
      "       'Lactobacillales'],\n",
      "      dtype='object')\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "2926"
      ]
     },
     "execution_count": 200,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(asv_16s[asv_16s['CP_pres']==True]['Order'].value_counts()[:10].index)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 205,
   "metadata": {},
   "outputs": [],
   "source": [
    "Bacteroidales = prop_shared_by_order('Bacteroidales')\n",
    "Mollicutes_RF39 = prop_shared_by_order('Mollicutes_RF39')\n",
    "Erysipelotrichales = prop_shared_by_order('Erysipelotrichales')\n",
    "Spirochaetales = prop_shared_by_order('Spirochaetales')\n",
    "Gastranaerophilales = prop_shared_by_order('Gastranaerophilales')\n",
    "Selenomonadales = prop_shared_by_order('Selenomonadales')\n",
    "Betaproteobacteriales = prop_shared_by_order('Selenomonadales')\n",
    "Coriobacteriales = prop_shared_by_order('Coriobacteriales')\n",
    "Lactobacillales = prop_shared_by_order('Lactobacillales')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "pw_shared_ASV_all_genera['order']='all'\n",
    "pw_shared_ASV_top10_gen = pd.concat([pw_shared_ASV_all_genera,Clostridales,Bacteroidales,Mollicutes_RF39,Erysipelotrichales,Spirochaetales,\n",
    " Gastranaerophilales,Selenomonadales,Betaproteobacteriales,Coriobacteriales,Lactobacillales])\n",
    "pw_shared_ASV_top10_gen.to_csv(\n",
    "    'analyses/tables/16s_prop_shared_ASVs_top10_table.txt',\n",
    "    sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind1</th>\n",
       "      <th>ind2</th>\n",
       "      <th>desc_site_ind1</th>\n",
       "      <th>desc_site_ind2</th>\n",
       "      <th>full_desc_comp</th>\n",
       "      <th>sp_site_comp</th>\n",
       "      <th>common_gen</th>\n",
       "      <th>ind1_ASVs</th>\n",
       "      <th>ind2_ASVs</th>\n",
       "      <th>common_ASVs</th>\n",
       "      <th>prop_shared_ASVs</th>\n",
       "      <th>order</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cp.ora.ATLZ.114.16s</td>\n",
       "      <td>cp.ora.ATLZ.115.16s</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ</td>\n",
       "      <td>same_spec_same_site</td>\n",
       "      <td>[Intestinimonas, Ruminococcaceae_UCG-005, Rumi...</td>\n",
       "      <td>136</td>\n",
       "      <td>118</td>\n",
       "      <td>91</td>\n",
       "      <td>0.771186</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cp.ora.ATLZ.114.16s</td>\n",
       "      <td>cp.ora.ATLZ.116.16s</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ</td>\n",
       "      <td>same_spec_same_site</td>\n",
       "      <td>[Intestinimonas, Ruminococcaceae_UCG-005, Rumi...</td>\n",
       "      <td>141</td>\n",
       "      <td>131</td>\n",
       "      <td>96</td>\n",
       "      <td>0.732824</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>cp.ora.ATLZ.114.16s</td>\n",
       "      <td>cp.ora.ATLZ.117.16s</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ</td>\n",
       "      <td>same_spec_same_site</td>\n",
       "      <td>[Intestinimonas, Ruminococcaceae_UCG-005, Rumi...</td>\n",
       "      <td>138</td>\n",
       "      <td>114</td>\n",
       "      <td>96</td>\n",
       "      <td>0.842105</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>cp.ora.ATLZ.114.16s</td>\n",
       "      <td>cp.ora.ATLZ.118.16s</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ</td>\n",
       "      <td>same_spec_same_site</td>\n",
       "      <td>[Intestinimonas, Catenibacterium, Ruminococcac...</td>\n",
       "      <td>124</td>\n",
       "      <td>100</td>\n",
       "      <td>85</td>\n",
       "      <td>0.850000</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>cp.ora.ATLZ.114.16s</td>\n",
       "      <td>cp.ora.ATLZ.119.16s</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ</td>\n",
       "      <td>captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ</td>\n",
       "      <td>same_spec_same_site</td>\n",
       "      <td>[Intestinimonas, Ruminococcaceae_UCG-005, Rumi...</td>\n",
       "      <td>104</td>\n",
       "      <td>72</td>\n",
       "      <td>54</td>\n",
       "      <td>0.750000</td>\n",
       "      <td>all</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  ind1                 ind2          desc_site_ind1  \\\n",
       "0  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.115.16s  captive_orangutan_ATLZ   \n",
       "1  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.116.16s  captive_orangutan_ATLZ   \n",
       "2  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.117.16s  captive_orangutan_ATLZ   \n",
       "3  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.118.16s  captive_orangutan_ATLZ   \n",
       "4  cp.ora.ATLZ.114.16s  cp.ora.ATLZ.119.16s  captive_orangutan_ATLZ   \n",
       "\n",
       "           desc_site_ind2                                    full_desc_comp  \\\n",
       "0  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
       "1  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
       "2  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
       "3  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
       "4  captive_orangutan_ATLZ  captive_orangutan_ATLZ_vs_captive_orangutan_ATLZ   \n",
       "\n",
       "          sp_site_comp                                         common_gen  \\\n",
       "0  same_spec_same_site  [Intestinimonas, Ruminococcaceae_UCG-005, Rumi...   \n",
       "1  same_spec_same_site  [Intestinimonas, Ruminococcaceae_UCG-005, Rumi...   \n",
       "2  same_spec_same_site  [Intestinimonas, Ruminococcaceae_UCG-005, Rumi...   \n",
       "3  same_spec_same_site  [Intestinimonas, Catenibacterium, Ruminococcac...   \n",
       "4  same_spec_same_site  [Intestinimonas, Ruminococcaceae_UCG-005, Rumi...   \n",
       "\n",
       "   ind1_ASVs  ind2_ASVs  common_ASVs  prop_shared_ASVs order  \n",
       "0        136        118           91          0.771186   all  \n",
       "1        141        131           96          0.732824   all  \n",
       "2        138        114           96          0.842105   all  \n",
       "3        124        100           85          0.850000   all  \n",
       "4        104         72           54          0.750000   all  "
      ]
     },
     "execution_count": 210,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pw_shared_ASV_top10_gen.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 191,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gen</th>\n",
       "      <th>prop_shared_ASVs</th>\n",
       "      <th>num_comp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0\n",
       "0  Candidat...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0\n",
       "0  Candidat...</td>\n",
       "      <td>0.927483</td>\n",
       "      <td>1772</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0\n",
       "0  Candidat...</td>\n",
       "      <td>0.575687</td>\n",
       "      <td>1297</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0\n",
       "0  Candidat...</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>401</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>0\n",
       "0  Candidat...</td>\n",
       "      <td>0.491477</td>\n",
       "      <td>528</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>0\n",
       "0  Candidat...</td>\n",
       "      <td>0.868993</td>\n",
       "      <td>1712</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                 gen  prop_shared_ASVs  \\\n",
       "0                                   0\n",
       "0  Candidat...               NaN   \n",
       "1                                   0\n",
       "0  Candidat...          0.927483   \n",
       "2                                   0\n",
       "0  Candidat...          0.575687   \n",
       "3                                   0\n",
       "0  Candidat...          1.000000   \n",
       "4                                   0\n",
       "0  Candidat...          0.491477   \n",
       "5                                   0\n",
       "0  Candidat...          0.868993   \n",
       "\n",
       "   num_comp  \n",
       "0         0  \n",
       "1      1772  \n",
       "2      1297  \n",
       "3       401  \n",
       "4       528  \n",
       "5      1712  "
      ]
     },
     "execution_count": 191,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
