{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from ete3 import Tree\n",
    "from collections import Counter\n",
    "import numpy as np\n",
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.getcwd()\n",
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/results/gyrb/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [],
   "source": [
    "#inputs\n",
    "metadata_file = 'inputs/metadata_gyrb_amp_meta_passing_samples.txt'\n",
    "tax_table_file = 'inputs/ASVs_taxonomy.txt'\n",
    "asv_table_file = 'inputs/ASVs_filtered_counts.tsv'\n",
    "asv_fasta_file = 'inputs/ASVs_filtered.fasta'\n",
    "full_tree_file = 'inputs/ASVs_filtered_ref_full.tree'\n",
    "#read tree with ete3 \n",
    "full_tree = Tree(full_tree_file, format=0)\n",
    "moeller_codiv_fasta = 'inputs/codiv_Bacteroidaceae.fna'\n",
    "#set blast percent identity cutoff \n",
    "pident_cutoff = 95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "\n",
    "sampleNames = asv_table.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "ASV_sampleName_dict = dict(zip(sampleNames.index,sampleNames))\n",
    "\n",
    "#sample to sample type category\n",
    "metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "sample_type_dict = dict(zip(metadata['X.SampleID'], metadata['Description']))\n",
    "\n",
    "#taxonomic info, family and genus\n",
    "tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "tax_fam_dict = dict(zip(tax_table['ASV'], tax_table['Family']))\n",
    "tax_gen_dict = dict(zip(tax_table['ASV'], tax_table['Genus']))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 09/09/2020 09:53:01\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb/analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "New DB title:  analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb/analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 208 sequences in 0.009305 seconds.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: analyses/codiv_moeller_ASVs: File exists\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir analyses/codiv_moeller_ASVs\n",
    "cp inputs/moeller_codiv_Bacteroidaceae.fna analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
    "\n",
    "#make blastdb\n",
    "makeblastdb -in analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna -dbtype nucl\n",
    "\n",
    "#blast moeller co-div seqs\n",
    "blastn -query inputs/ASVs_filtered.fasta -db analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna \\\n",
    "-outfmt \"7 qseqid salltitles sseqid pident length qlen evalue\" \\\n",
    "-out analyses/codiv_moeller_ASVs/codiv_blastout_ASVs.txt \\\n",
    "-max_target_seqs 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "314 ASVs matching co-div clades with greater than 95 percent identity\n",
      "Bt2_clade1_bonobo     90\n",
      "Bt3_clade1_human      85\n",
      "Bt2_clade2_bonobo     33\n",
      "Bt1_clade1_bonobo     23\n",
      "Bt2_clade1_chimp      20\n",
      "Bt3_clade1_chimp      17\n",
      "Bt3_clade1_bonobo     15\n",
      "Bt1_clade1_gorilla    10\n",
      "Bt1_clade1_chimp       8\n",
      "Bt2_clade1_gorilla     7\n",
      "Bt2_clade2_chimp       6\n",
      "Name: codiv_clade, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#filter blast output to top hit per ASV, remove hits not pass thresholds\n",
    "blast_res = pd.read_csv('analyses/codiv_moeller_ASVs/codiv_blastout_ASVs.txt',sep='\\t',comment='#',header=None)\n",
    "blast_res.columns = ['ASV','codiv_clade_seq','sseqid','pident','length','qlen','evalue']\n",
    "blast_res = blast_res.groupby('ASV').head(1).reset_index(drop=True) #get top hit\n",
    "blast_res_pass = blast_res[blast_res['pident']>pident_cutoff].reset_index(drop=True) #QC filter\n",
    "blast_res_pass = blast_res_pass[blast_res_pass['length']>200].reset_index(drop=True) #QC filter\n",
    "print(len(blast_res_pass),'ASVs matching co-div clades with greater than',pident_cutoff,'percent identity')\n",
    "blast_res_pass['codiv_clade'] = blast_res_pass['codiv_clade_seq'].apply(lambda x: x.split(' ')[1])\n",
    "blast_res_pass['lineage'] = blast_res_pass['codiv_clade'].apply(lambda x: x.split('_')[0]) \n",
    "blast_res_pass = blast_res_pass[['ASV','codiv_clade_seq', 'codiv_clade','lineage']]\n",
    "#print(blast_res_pass.head())\n",
    "print(blast_res_pass['codiv_clade'].value_counts())\n",
    "#output hits passing that will be used to determine the mcra of the lineage\n",
    "blast_res_pass.to_csv('analyses/codiv_moeller_ASVs/codiv_clades_ASVs.txt',sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# defining host restricted clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(['human'], 4, 'HR', 'HR_human', False, [], 0)\n",
      "Bacteroidaceae_Unassigned\n"
     ]
    }
   ],
   "source": [
    "def is_HR(sampleNames,sample_type_dict):\n",
    "    \"\"\"given a list of sample names uses sample type dictionary to determine how many sample types are present\n",
    "    designates ASVs as host restricted = 1 sample type or mixed = multiple sample types.\n",
    "    Captive sample types are not considered so some clades will have a 0 sample type length and they can fall within\n",
    "    host-restricted clades or mixed clades or neither\"\"\"\n",
    "    sampleTypes = [sample_type_dict[name] for name in sampleNames]\n",
    "    sampleTypes = [x.replace('non_western_','').replace('western_','') for x in sampleTypes]\n",
    "    neutral_sampleTypes = ['captive_gorilla','captive_bonobo','captive_chimp','captive_orangutan']\n",
    "    HR_sampleTypes = list(set(sampleTypes) - set(neutral_sampleTypes))\n",
    "    HR_sampleNum = len([x for x in sampleTypes if x not in neutral_sampleTypes])\n",
    "    CP_sampleTypes = list(set(sampleTypes) & set(neutral_sampleTypes))\n",
    "    CP_sampleNum = len([x for x in sampleTypes if x in neutral_sampleTypes])\n",
    "    CP_pres = True if len(CP_sampleTypes) > 0 else False\n",
    "    if len(HR_sampleTypes) == 0:\n",
    "        HR_cat,HR_type='CP','CP'\n",
    "    if len(HR_sampleTypes) == 1: #identifies host-restricted clades\n",
    "        HR_cat,HR_type='HR',HR_sampleTypes[0]  \n",
    "    if len(HR_sampleTypes) > 1:    \n",
    "        HR_cat,HR_type ='MX','MX'\n",
    "    return(HR_sampleTypes,HR_sampleNum,HR_cat,HR_type,CP_pres,CP_sampleTypes,CP_sampleNum)\n",
    "\n",
    "def is_HR(sampleNames,sample_type_dict):\n",
    "    \"\"\"given a list of sample names uses sample type dictionary to determine how many sample types are present\n",
    "    designates ASVs as host restricted = 1 sample type or mixed = multiple sample types.\n",
    "    Captive sample types are not considered so some clades will have a 0 sample type length and they can fall within\n",
    "    host-restricted clades or mixed clades or neither\"\"\"\n",
    "    sampleTypes = [sample_type_dict[name] for name in sampleNames]\n",
    "    sampleTypes = [x.replace('non_western_','').replace('western_','') for x in sampleTypes]\n",
    "    neutral_sampleTypes = ['captive_gorilla','captive_bonobo','captive_chimp','captive_orangutan']\n",
    "    HR_sampleTypes = list(set(sampleTypes) - set(neutral_sampleTypes))\n",
    "    HR_sampleNum = len([x for x in sampleTypes if x not in neutral_sampleTypes])\n",
    "    CP_sampleTypes = list(set(sampleTypes) & set(neutral_sampleTypes))\n",
    "    CP_sampleNum = len([x for x in sampleTypes if x in neutral_sampleTypes])\n",
    "    CP_pres = True if len(CP_sampleTypes) > 0 else False\n",
    "    if len(HR_sampleTypes) == 0:\n",
    "        HR_cat,HR_type='Unique_CP','Unique_CP'\n",
    "    if len(HR_sampleTypes) == 1: #identifies host-restricted clades\n",
    "        HR_cat,HR_type='HR','HR_'+HR_sampleTypes[0] \n",
    "    if len(HR_sampleTypes) > 1:    \n",
    "        HR_cat='MX'\n",
    "        if len(HR_sampleTypes) == 2:\n",
    "            if 'human' in HR_sampleTypes:\n",
    "                HR_type = 'MX_human_single_wild_ape'\n",
    "            else:\n",
    "                HR_type = 'MX_2_wild_apes'\n",
    "        if len(HR_sampleTypes) == 3:\n",
    "            if 'human' in HR_sampleTypes:\n",
    "                HR_type = 'MX_human_2_wild_apes'\n",
    "            else:\n",
    "                HR_type = 'MX_3_wild_apes'\n",
    "        if len(HR_sampleTypes) == 4:\n",
    "            HR_type = 'MX_4_hominids'\n",
    "            \n",
    "    return(HR_sampleTypes,HR_sampleNum,HR_cat,HR_type,CP_pres,CP_sampleTypes,CP_sampleNum)\n",
    "\n",
    "    #if len(HR_sampleTypes) == 0:\n",
    "    #    HR_cat,HR_type='Unique_CP','Unique_CP'\n",
    "   # if len(HR_sampleTypes) == 1: #identifies host-restricted clades\n",
    "   #     HR_cat,HR_type='HR','HR_'+HR_sampleTypes[0]  \n",
    "    #if len(HR_sampleTypes) > 1: \n",
    "    #    HR_cat = 'NHR'\n",
    "    #    if len(HR_sampleTypes) == 2:\n",
    "     #       if 'human' in HR_sampleTypes:\n",
    "     #           HR_type = 'NHR_human_single_wild_ape'\n",
    "     #       else:\n",
    "    #            HR_type = 'NHR_2_wild_apes'\n",
    "    #    if len(HR_sampleTypes) == 3:\n",
    "    #        if 'human' in HR_sampleTypes:\n",
    "    #            HR_type = 'NHR_human_2_wild_apes'\n",
    "    #        else:\n",
    "    #            HR_type = 'NHR_3_wild_apes'\n",
    "    #    if len(HR_sampleTypes) == 4:\n",
    "    #        HR_type = 'NHR_4_hominids'\n",
    "        \n",
    "    #return(HR_sampleTypes,HR_sampleNum,HR_cat,HR_type,CP_pres,CP_sampleTypes,CP_sampleNum)\n",
    "           \n",
    "           \n",
    "print(is_HR(['CosteaPI_2017__SID713A076-11-0-0','LiSS_2016__FAT_024-22-42-0', \n",
    "       'LiSS_2016__FAT_024-22-84-0', 'LiSS_2016__FAT_DON_11-22-0-4'],sample_type_dict)) \n",
    "\n",
    "def get_consensus_taxonomy(listASVs,tax_fam_dict,tax_gen_dict):\n",
    "    \"\"\"Because multiple ASVs in a single clade, we determine the taxonomy of the clade by consensus.\n",
    "    If all ASVs in the clade belong to a given bacterial taxonomic family/genus the clade is assigned that taxonomy\n",
    "    If the clade contains ASVs assigned to multiple families/genera the taxonomy is labeled as unassigned.\"\"\"\n",
    "    fam = list(set([tax_fam_dict[ASV] for ASV in listASVs])-set(['Unassigned']))\n",
    "    gen = list(set([tax_gen_dict[ASV] for ASV in listASVs])-set(['Unassigned']))\n",
    "    fam = fam[0].split('__')[1] if len(fam) == 1 else 'Unassigned'\n",
    "    if fam == 'Unassigned':\n",
    "        gen='Unassigned'\n",
    "    else:\n",
    "        gen = gen[0].split('__')[1] if len(gen) == 1 else 'Unassigned'  \n",
    "    return(fam+'_'+gen)   \n",
    "print(get_consensus_taxonomy(['ASV_1','ASV_2','ASV_1078'],tax_fam_dict,tax_gen_dict))\n",
    "\n",
    "def search_clades(tree, samples_cutoff, BS_support,ASV_sampleName_dict):\n",
    "    \"\"\"Finds nodes with at least 50% BS support containing ASVs only found in a single wild ape species\n",
    "    ie wild_gorilla, wild_chimp or wild_bonobo, cpat\"\"\"\n",
    "    clades_prelim = []\n",
    "    counter = 1\n",
    "    for n in tree.traverse():\n",
    "        if n.support > float(BS_support): #makes sure Bootstrap support is over threshold\n",
    "            ASVs = [leaf.name for leaf in n.iter_leaves() if 'ASV' in leaf.name]\n",
    "            listoflists = [ASV_sampleName_dict[ASV] for ASV in ASVs]\n",
    "            sampleNames = list(set(list(itertools.chain.from_iterable(listoflists))))\n",
    "            cladeName='clade_'+str(counter)\n",
    "            counter+=1\n",
    "            clade = [cladeName,ASVs,sampleNames]\n",
    "            if len(sampleNames)>samples_cutoff:\n",
    "                clades_prelim.append(clade)\n",
    "    clades_prelim = pd.DataFrame(clades_prelim, columns = \n",
    "                          ['cladeName','ASVs','sampleNames'])\n",
    "    clades_prelim['sampleNum'] = clades_prelim['sampleNames'].apply(lambda x: len(x))\n",
    "    clades_prelim['ASVsNum'] = clades_prelim['ASVs'].apply(lambda x: len(x))\n",
    "    return(clades_prelim)\n",
    "\n",
    "def eliminate_redundant_clades(clades_df,offlimits_ASVs):\n",
    "    \"\"\"sorts clades and returns the largest non overlapping clade \"\"\"\n",
    "    df = clades_df.sort_values('ASVsNum',ascending=False) #start with the largest clades first\n",
    "    NRclades = []\n",
    "    for index, row in df.iterrows():\n",
    "        if len(set(row['ASVs']) & set(offlimits_ASVs)) == 0: \n",
    "            offlimits_ASVs = offlimits_ASVs + row['ASVs']\n",
    "            NRclades.append(row['cladeName']) \n",
    "    res = df[df['cladeName'].isin(NRclades)]   \n",
    "    return(res)  \n",
    "\n",
    "def expand_clades(clades_df):\n",
    "    ASV = clades_df.apply(lambda x: pd.Series(x['ASVs']),axis=1).stack().reset_index(level=1, drop=True)\n",
    "    ASV.name = 'ASVs'\n",
    "    clades_ASVs_df = clades_df.drop('ASVs', axis=1).join(ASV)\n",
    "    return(clades_ASVs_df)\n",
    "                                 \n",
    "def host_restricted_clades(asv_table_file,metadata_file,tax_table_file):\n",
    "    asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "    sampleNames = asv_table.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "    ASV_sampleName_dict = dict(zip(sampleNames.index,sampleNames))\n",
    "\n",
    "    #sample to sample type category\n",
    "    metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "    sample_type_dict = dict(zip(metadata['X.SampleID'], metadata['Description']))\n",
    "\n",
    "    #taxonomic info, family and genus\n",
    "    tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "    tax_fam_dict = dict(zip(tax_table['ASV'], tax_table['Family']))\n",
    "    tax_gen_dict = dict(zip(tax_table['ASV'], tax_table['Genus']))\n",
    "    \n",
    "    #search tree for clades\n",
    "    clades_prelim = search_clades(full_tree, 5, .5, ASV_sampleName_dict)\n",
    "    hr =clades_prelim['sampleNames'].apply(lambda x:  pd.Series(is_HR(x,sample_type_dict),\n",
    "                index=['HR_sampleTypes','HR_sampleNum','HR_cat','HR_type','CP_pres','CP_sampleTypes','CP_sampleNum']))\n",
    "    clades_prelim_hr = clades_prelim.merge(hr,right_index=True,left_index=True)\n",
    "\n",
    "    #identify HR clades\n",
    "    HR_clades = clades_prelim_hr[clades_prelim_hr['HR_cat']=='HR']\n",
    "    HR_clades = HR_clades[HR_clades['HR_sampleNum']>5]\n",
    "    HR_clades = eliminate_redundant_clades(HR_clades,[])\n",
    "    HR_clades_ASVs = expand_clades(HR_clades)\n",
    "\n",
    "    #identify MX clades that don't contain any HR clades\n",
    "    MX_clades = clades_prelim_hr[clades_prelim_hr['HR_cat']=='MX']\n",
    "    MX_clades = MX_clades[MX_clades['sampleNum']>5]\n",
    "    MX_clades = eliminate_redundant_clades(MX_clades,offlimits_ASVs=list(HR_clades_ASVs['ASVs']))\n",
    "    MX_clades_ASVs = expand_clades(MX_clades)\n",
    "    \n",
    "    #identify MX clades that don't contain any HR clades\n",
    "    CP_clades = clades_prelim_hr[clades_prelim_hr['HR_cat']=='Unique_CP']\n",
    "    CP_clades = CP_clades[CP_clades['sampleNum']>5]\n",
    "    HR_MX_ASVs = list(HR_clades_ASVs['ASVs']) + list(MX_clades_ASVs['ASVs'])\n",
    "    CP_clades = eliminate_redundant_clades(CP_clades,offlimits_ASVs=HR_MX_ASVs)\n",
    "    CP_clades_ASVs = expand_clades(CP_clades)\n",
    "    \n",
    "    #merge dataframes\n",
    "    clades = pd.concat([HR_clades,MX_clades,CP_clades])\n",
    "    clades.reset_index(drop=True, inplace=True)\n",
    "    clades_ASVs = pd.concat([HR_clades_ASVs,MX_clades_ASVs,CP_clades_ASVs])\n",
    "    clades_ASVs.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    \n",
    "    clades['cladeTax']=clades['ASVs'].apply(lambda x: \n",
    "        get_consensus_taxonomy(x,tax_fam_dict,tax_gen_dict))\n",
    "    \n",
    "    sample_type_counts =  metadata['Description'].value_counts()\n",
    "    description_df = clades['sampleNames'].apply(lambda l: pd.Series(\n",
    "    [sample_type_dict[name] for name in l]).value_counts())\n",
    "    description_df = description_df.fillna(0)  \n",
    "    sample_type_percent = description_df/sample_type_counts\n",
    "    clades=clades.merge(sample_type_percent,left_index=True,right_index=True)  \n",
    "    \n",
    "    return(clades,clades_ASVs)\n",
    "    \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Run function on gyrb data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clades_df,clades_ASVs_df = host_restricted_clades(asv_table_file,metadata_file,tax_table_file)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(359, 25)\n",
      "(6438, 12)\n",
      "How many of the clades are present in captive apes?\n",
      "52\n",
      "Are these clades host-restricted, mixed host, or unique to captive?\n",
      "HR           27\n",
      "MX           19\n",
      "Unique_CP     6\n",
      "Name: HR_cat, dtype: int64\n",
      "False    328\n",
      "True      31\n",
      "Name: CP_prominent, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(clades_df.shape)\n",
    "#print(clades_df_sh.head())\n",
    "\n",
    "#Used for Figure3 tips on tree are ASVs, not clades\n",
    "print(clades_ASVs_df.shape)\n",
    "#print(clades_ASVs_df.head())\n",
    "\n",
    "print('How many of the clades are present in captive apes?')\n",
    "captive_clades = clades_df[clades_df['CP_pres']==True]\n",
    "print(len(captive_clades))\n",
    "print('Are these clades host-restricted, mixed host, or unique to captive?')\n",
    "print(captive_clades['HR_cat'].value_counts())\n",
    "\n",
    "\n",
    "def just_captive_samples(sampleNames):\n",
    "    return([name for name in sampleNames if 'captive' in sample_type_dict[name]])    \n",
    "clades_df['captiveNames'] = clades_df['sampleNames'].apply(lambda x: just_captive_samples(x))  \n",
    "\n",
    "prominent_clades = clades_df.loc[(clades_df.captive_bonobo > 0.15) | \n",
    "              (clades_df.captive_chimp > 0.15)  |\n",
    "              (clades_df.captive_gorilla > 0.15)|\n",
    "              (clades_df.captive_orangutan > 0.15)]\n",
    "clades_df['CP_prominent'] = clades_df['cladeName'].apply(lambda x: x in list(prominent_clades['cladeName']))\n",
    "print(clades_df['CP_prominent'].value_counts())   \n",
    "\n",
    "#clades_df_sh = clades_df.drop(['ASVs','sampleNames'], axis=1)\n",
    "clades_df_sh = clades_df[['cladeName', 'cladeTax', 'sampleNum', 'ASVsNum', \n",
    "      'HR_sampleTypes', 'HR_sampleNum','HR_cat', 'HR_type', \n",
    "      'CP_pres','CP_prominent', 'CP_sampleTypes', 'CP_sampleNum','captiveNames',\n",
    "       'captive_bonobo', 'captive_chimp', 'captive_gorilla',\n",
    "       'captive_orangutan', 'non_western_human', 'western_human',\n",
    "       'wild_bonobo', 'wild_chimp', 'wild_gorilla']]\n",
    "       \n",
    "\n",
    "clades_df_sh.to_csv('analyses/codiv_moeller_ASVs/HRclades_Figure2_table.txt',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### search full tree for host-restricted clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "collasped_tree = Tree(full_tree_file, format=0)\n",
    "\n",
    "def collapse_node(tree,clade_ASVs,cladeName):\n",
    "    if len(clade_ASVs) > 1:\n",
    "        node = tree.get_common_ancestor(clade_ASVs)\n",
    "        node.name = cladeName\n",
    "        children = node.get_children()\n",
    "        node.remove_child(children[1])\n",
    "        node.remove_child(children[0])\n",
    "    else:\n",
    "        ASV = clade_ASVs[0]\n",
    "        leaf = tree.get_leaves_by_name(name=ASV)[0]\n",
    "        leaf.name = cladeName\n",
    "\n",
    "for clade,row in clades_df.iterrows():\n",
    "    collapse_node(collasped_tree,row['ASVs'],row['cladeName'])\n",
    "\n",
    "clade_leaves = [leaf for leaf in collasped_tree.get_leaves() if 'clade' in leaf.name] \n",
    "collasped_tree.prune(clade_leaves) #eliminate leftover ASVs and ref taxa\n",
    "collasped_tree.write(format=2, outfile='analyses/codiv_moeller_ASVs/HRclades_Figure2.tre')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# ASV info for subtree figures"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7148\n",
      "  cladeName                                        sampleNames  sampleNum  \\\n",
      "0  clade_41  [Obregon-TitoAJ_2015__SM41, CosteaPI_2017__SID...        788   \n",
      "1  clade_41  [Obregon-TitoAJ_2015__SM41, CosteaPI_2017__SID...        788   \n",
      "2  clade_41  [Obregon-TitoAJ_2015__SM41, CosteaPI_2017__SID...        788   \n",
      "3  clade_41  [Obregon-TitoAJ_2015__SM41, CosteaPI_2017__SID...        788   \n",
      "4  clade_41  [Obregon-TitoAJ_2015__SM41, CosteaPI_2017__SID...        788   \n",
      "\n",
      "   ASVsNum HR_sampleTypes  HR_sampleNum HR_cat   HR_type  CP_pres  \\\n",
      "0      366        [human]           788     HR  HR_human    False   \n",
      "1      366        [human]           788     HR  HR_human    False   \n",
      "2      366        [human]           788     HR  HR_human    False   \n",
      "3      366        [human]           788     HR  HR_human    False   \n",
      "4      366        [human]           788     HR  HR_human    False   \n",
      "\n",
      "  CP_sampleTypes  CP_sampleNum      ASVs  \n",
      "0             []             0  ASV_2623  \n",
      "1             []             0  ASV_1788  \n",
      "2             []             0  ASV_7289  \n",
      "3             []             0  ASV_7288  \n",
      "4             []             0  ASV_7176  \n",
      "6438\n",
      "      ASV                codiv_clade_seq       codiv_clade lineage\n",
      "0   ASV_1  Human6700665 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "1  ASV_13  Human7859459 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "2  ASV_34  Human9825266 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "3  ASV_37  Human8829943 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "4  ASV_73  Human9825266 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "314\n",
      "         ASV   cladeName                                        sampleNames  \\\n",
      "0  ASV_10387  clade_8751                           [SchirmerM_2016__G88949]   \n",
      "1    ASV_408  clade_8751  [CosteaPI_2017__SID713A037-11-0-0, CosteaPI_20...   \n",
      "2   ASV_3299  clade_8751  [LeChatelierE_2013__MH0085, NielsenHB_2014__MH...   \n",
      "3  ASV_10369  clade_8751                           [SchirmerM_2016__G89027]   \n",
      "4  ASV_10428  clade_8751                           [SchirmerM_2016__G89097]   \n",
      "\n",
      "   sampleNum  ASVsNum HR_sampleTypes  HR_sampleNum HR_cat   HR_type CP_pres  \\\n",
      "0     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "1     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "2     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "3     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "4     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "\n",
      "   ... lineage             Family          Genus captive_chimp_HOUZ  \\\n",
      "0  ...     NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "1  ...     NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "2  ...     NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "3  ...     NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "4  ...     NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "\n",
      "  captive_gorilla_HOUZ captive_orangutan_HOUZ captive_bonobo_COLZ  \\\n",
      "0                  0.0                    0.0                 0.0   \n",
      "1                  0.0                    0.0                 0.0   \n",
      "2                  0.0                    0.0                 0.0   \n",
      "3                  0.0                    0.0                 0.0   \n",
      "4                  0.0                    0.0                 0.0   \n",
      "\n",
      "  captive_gorilla_COLZ  captive_orangutan_COLZ  captive_chimp_PC  \n",
      "0                  0.0                     0.0               0.0  \n",
      "1                  0.0                     0.0               0.0  \n",
      "2                  0.0                     0.0               0.0  \n",
      "3                  0.0                     0.0               0.0  \n",
      "4                  0.0                     0.0               0.0  \n",
      "\n",
      "[5 rows x 25 columns]\n"
     ]
    }
   ],
   "source": [
    "allASVs = [leaf.name for leaf in full_tree.get_leaves() if 'ASV' in leaf.name]\n",
    "allASVs = pd.DataFrame(allASVs,columns=['ASV'])\n",
    "print(len(allASVs))\n",
    "\n",
    "#Host restricted clades \n",
    "print(clades_ASVs_df.head())\n",
    "print(len(clades_ASVs_df))\n",
    "#Moeller codiv clades\n",
    "print(blast_res_pass.head())\n",
    "print(len(blast_res_pass))\n",
    "\n",
    "#combine\n",
    "allASVs_HRclade = allASVs.merge(clades_ASVs_df, how='left', left_on='ASV',right_on='ASVs')\n",
    "allASVs_HRclade_codiv = allASVs_HRclade.merge(blast_res_pass, how='left', on='ASV')\n",
    "allASVs_HRclade_codiv.head()\n",
    "\n",
    "#add taxonomy\n",
    "allASVs_HRclade_codiv['Family'] = allASVs_HRclade_codiv['ASV'].apply(lambda ASV: tax_fam_dict[ASV])\n",
    "allASVs_HRclade_codiv['Genus'] = allASVs_HRclade_codiv['ASV'].apply(lambda ASV: tax_gen_dict[ASV])\n",
    "\n",
    "#add captive sp and site info\n",
    "metadata['Description_site'] = metadata['Description']+'_' +metadata['site_code']\n",
    "sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "\n",
    "allASVs_HRclade_codiv['sampleNames'] = allASVs_HRclade_codiv['ASV'].apply(lambda x: ASV_sampleName_dict[x])\n",
    "\n",
    "description_df = allASVs_HRclade_codiv['sampleNames'].apply(lambda l: pd.Series(\n",
    "    [sample_type_site_dict[name] for name in l]).value_counts())\n",
    "description_df = description_df.fillna(0) \n",
    "description_cp_df= description_df[['captive_chimp_HOUZ','captive_gorilla_HOUZ','captive_orangutan_HOUZ',\n",
    "'captive_bonobo_COLZ','captive_gorilla_COLZ','captive_orangutan_COLZ','captive_chimp_PC']]\n",
    "allASVs_HRclade_codiv_cp = allASVs_HRclade_codiv.merge(description_cp_df, left_index=True,right_index=True)\n",
    "\n",
    "allASVs_HRclade_codiv_cp.to_csv('analyses/codiv_moeller_ASVs/full_tree_ASV_table.txt',sep='\\t',index=False)\n",
    "print(allASVs_HRclade_codiv_cp.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
