{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from ete3 import Tree\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/scripts/analyses')\n",
    "%run functions.ipynb\n",
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/results/gyrb')\n",
    "\n",
    "#results/gyrb/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#inputs\n",
    "metadata_file = 'inputs/metadata_gyrb_amp_meta_passing_samples.txt'\n",
    "tax_table_file = 'inputs/ASVs_taxonomy.txt'\n",
    "asv_table_file = 'inputs/ASVs_filtered_counts.tsv'\n",
    "asv_fasta_file = 'inputs/ASVs_filtered.fasta'\n",
    "full_tree_file = 'inputs/ASVs_filtered_ref_full.tree'\n",
    "#read tree with ete3 \n",
    "full_tree = Tree(full_tree_file, format=0)\n",
    "moeller_codiv_fasta = 'inputs/codiv_Bacteroidaceae.fna'\n",
    "#set blast percent identity cutoff \n",
    "pident_cutoff = 95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "g__Prevotella            2871\n",
       "g__Bacteroides            992\n",
       "Unassigned                539\n",
       "g__Alistipes              519\n",
       "g__RC9                    363\n",
       "g__Parabacteroides        296\n",
       "g__Bacteroides_A          286\n",
       "g__Porphyromonas          250\n",
       "g__Prevotellamassilia     121\n",
       "g__Bacteroides_B          113\n",
       "Name: Genus, dtype: int64"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gyrb_asvs = asv_hr_table(asv_table_file,metadata_file,tax_table_file)\n",
    "gyrb_asvs['Genus'].value_counts()[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def output_summary_table(asv_hr_table_output):\n",
    "     #create summary table\n",
    "    Bacteroides_genera=['g__Bacteroides','g__Bacteroides_A','g__Bacteroides_B']\n",
    "    all_asvs = pd.Series(asv_hr_table_output['HR_type'].value_counts(),name='ALL')\n",
    "    cp_asvs = pd.Series(asv_hr_table_output[asv_hr_table_output['CP_pres']==True]['HR_type'].value_counts(),name='CP')\n",
    "    Bacteroidales_asvs = pd.Series(asv_hr_table_output[asv_hr_table_output['Order']=='o__Bacteroidales']['HR_type'].value_counts(),name='Bacteroidales')\n",
    "    Prevotella_asvs = pd.Series(asv_hr_table_output[asv_hr_table_output['Genus']=='g__Prevotella']['HR_type'].value_counts(),name='Prevotella')\n",
    "    Bacteroides_asvs = pd.Series(asv_hr_table_output[asv_hr_table_output['Genus'].isin(Bacteroides_genera)]['HR_type'].value_counts(),name='Bacteroides')\n",
    "    Parabacteroides_asvs = pd.Series(asv_hr_table_output[asv_hr_table_output['Genus']=='g__Parabacteroides']['HR_type'].value_counts(),name='Parabacteroides')\n",
    "\n",
    "    res = pd.concat([all_asvs,cp_asvs,Bacteroidales_asvs,\n",
    "                     Prevotella_asvs,Bacteroides_asvs,Parabacteroides_asvs],\n",
    "                     axis=1).fillna(0).T\n",
    "    res['Total'] = res.sum(axis=1)\n",
    "    res['MX_Total'] = res.loc[:,'MX_human_single_wild_ape':'MX_human_2_wild_apes'].sum(axis=1)\n",
    "    return(res)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Host restricted ASVs totals\n",
      "ALL                7001.0\n",
      "CP                   31.0\n",
      "Bacteroidales      7001.0\n",
      "Prevotella         2813.0\n",
      "Bacteroides        1348.0\n",
      "Parabacteroides     289.0\n",
      "dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>HR_human</th>\n",
       "      <th>HR_wild_bonobo</th>\n",
       "      <th>HR_wild_gorilla</th>\n",
       "      <th>HR_wild_chimp</th>\n",
       "      <th>Unique_CP</th>\n",
       "      <th>MX_human_single_wild_ape</th>\n",
       "      <th>MX_2_wild_apes</th>\n",
       "      <th>MX_human_2_wild_apes</th>\n",
       "      <th>Total</th>\n",
       "      <th>MX_Total</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>ALL</th>\n",
       "      <td>6183.0</td>\n",
       "      <td>353.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7148.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CP</th>\n",
       "      <td>27.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>10.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>139.0</td>\n",
       "      <td>14.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidales</th>\n",
       "      <td>6183.0</td>\n",
       "      <td>353.0</td>\n",
       "      <td>258.0</td>\n",
       "      <td>207.0</td>\n",
       "      <td>94.0</td>\n",
       "      <td>34.0</td>\n",
       "      <td>14.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>7148.0</td>\n",
       "      <td>53.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Prevotella</th>\n",
       "      <td>2450.0</td>\n",
       "      <td>203.0</td>\n",
       "      <td>46.0</td>\n",
       "      <td>114.0</td>\n",
       "      <td>55.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2871.0</td>\n",
       "      <td>3.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroides</th>\n",
       "      <td>1314.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>31.0</td>\n",
       "      <td>15.0</td>\n",
       "      <td>23.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>1391.0</td>\n",
       "      <td>28.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Parabacteroides</th>\n",
       "      <td>247.0</td>\n",
       "      <td>38.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>296.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                 HR_human  HR_wild_bonobo  HR_wild_gorilla  HR_wild_chimp  \\\n",
       "ALL                6183.0           353.0            258.0          207.0   \n",
       "CP                   27.0             0.0              0.0            4.0   \n",
       "Bacteroidales      6183.0           353.0            258.0          207.0   \n",
       "Prevotella         2450.0           203.0             46.0          114.0   \n",
       "Bacteroides        1314.0             1.0              2.0           31.0   \n",
       "Parabacteroides     247.0            38.0              1.0            3.0   \n",
       "\n",
       "                 Unique_CP  MX_human_single_wild_ape  MX_2_wild_apes  \\\n",
       "ALL                   94.0                      34.0            14.0   \n",
       "CP                    94.0                      10.0             2.0   \n",
       "Bacteroidales         94.0                      34.0            14.0   \n",
       "Prevotella            55.0                       2.0             1.0   \n",
       "Bacteroides           15.0                      23.0             0.0   \n",
       "Parabacteroides        3.0                       3.0             1.0   \n",
       "\n",
       "                 MX_human_2_wild_apes   Total  MX_Total  \n",
       "ALL                               5.0  7148.0      53.0  \n",
       "CP                                2.0   139.0      14.0  \n",
       "Bacteroidales                     5.0  7148.0      53.0  \n",
       "Prevotella                        0.0  2871.0       3.0  \n",
       "Bacteroides                       5.0  1391.0      28.0  \n",
       "Parabacteroides                   0.0   296.0       4.0  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gyrb_asv_summary = output_summary_table(gyrb_asvs)\n",
    "print('Host restricted ASVs totals')\n",
    "print(gyrb_asv_summary.loc[:,'HR_human':'HR_wild_chimp'].sum(axis=1))\n",
    "gyrb_asv_summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "western_human        6805\n",
      "non_western_human     511\n",
      "wild_chimp             69\n",
      "wild_gorilla           37\n",
      "captive_chimp          26\n",
      "wild_bonobo            24\n",
      "captive_gorilla        22\n",
      "captive_bonobo         13\n",
      "captive_orangutan      11\n",
      "Name: Description, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "clades_df,clades_ASVs_df = host_restricted_clades(asv_table_file,metadata_file,tax_table_file) \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(359, 23)\n",
      "numbers of clades in various host types\n",
      "HR_human                    249\n",
      "MX_human_single_wild_ape     35\n",
      "HR_wild_chimp                20\n",
      "HR_wild_gorilla              19\n",
      "HR_wild_bonobo               14\n",
      "MX_2_wild_apes                8\n",
      "MX_human_2_wild_apes          7\n",
      "Unique_CP                     6\n",
      "MX_3_wild_apes                1\n",
      "Name: HR_type, dtype: int64\n",
      "(6438, 13)\n",
      "number of ASVs falling into various clades\n",
      "HR_human                    5625\n",
      "HR_wild_bonobo               342\n",
      "HR_wild_gorilla              168\n",
      "HR_wild_chimp                111\n",
      "MX_human_single_wild_ape      92\n",
      "MX_2_wild_apes                43\n",
      "MX_human_2_wild_apes          41\n",
      "Unique_CP                      9\n",
      "MX_3_wild_apes                 7\n",
      "Name: HR_type, dtype: int64\n",
      "How many of the clades are present in captive apes?\n",
      "52\n",
      "Are these clades host-restricted, mixed host, or unique to captive?\n",
      "HR           27\n",
      "MX           19\n",
      "Unique_CP     6\n",
      "Name: HR_cat, dtype: int64\n",
      "31 captive clades in 0.15 of captive samples of any species\n",
      "HR_human                    12\n",
      "MX_human_single_wild_ape     8\n",
      "Unique_CP                    6\n",
      "MX_2_wild_apes               3\n",
      "MX_human_2_wild_apes         1\n",
      "HR_wild_chimp                1\n",
      "Name: HR_type, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#host-restricted, mixed-host, and unique-to-captive clades\n",
    "print(clades_df.shape)\n",
    "print('numbers of clades in various host types')\n",
    "print(clades_df['HR_type'].value_counts())\n",
    "\n",
    "#Used for Figure3, where tips on tree are ASVs not clades\n",
    "print(clades_ASVs_df.shape)\n",
    "print('number of ASVs falling into various clades')\n",
    "print(clades_ASVs_df['HR_type'].value_counts())\n",
    "\n",
    "#Whats the breakdown of clades present in captive ape samples?\n",
    "print('How many of the clades are present in captive apes?')\n",
    "captive_clades = clades_df[clades_df['CP_pres']==True]\n",
    "print(len(captive_clades))\n",
    "print('Are these clades host-restricted, mixed host, or unique to captive?')\n",
    "print(captive_clades['HR_cat'].value_counts())\n",
    "\n",
    "#Whats the breakdown of clades present in 15% of captive ape samples of any species?\n",
    "threshold=0.15\n",
    "prominent_clades = clades_df.loc[(clades_df.captive_bonobo > threshold) | \n",
    "              (clades_df.captive_chimp > threshold)  |\n",
    "              (clades_df.captive_gorilla > threshold)|\n",
    "              (clades_df.captive_orangutan > threshold)]\n",
    "print(len(prominent_clades),'captive clades in',threshold,'of captive samples of any species')\n",
    "clades_df['CP_prominent'] = clades_df['cladeName'].apply(lambda x: x in list(prominent_clades['cladeName']))\n",
    "clades_df_CP_prominent = clades_df[clades_df['CP_prominent']==True]\n",
    "print(clades_df_CP_prominent['HR_type'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "threshold = .15\n",
    "clades_df['heatmap_col1'] = clades_df.apply(\n",
    "    lambda row: 'MX' \n",
    "        if 'MX' in row['HR_type'] else row['HR_type'],axis=1)\n",
    "clades_df['heatmap_col2'] = clades_df.apply(\n",
    "    lambda row: row['heatmap_col1'] \n",
    "        if max(row['captive_bonobo':'wild_gorilla'])>.15 else 'Blank',axis=1)\n",
    "clades_df['heatmap_col3'] = clades_df.apply(\n",
    "    lambda row: row['heatmap_col1'] \n",
    "        if max(row['captive_bonobo':'captive_orangutan'])>.15 else 'Blank',axis=1)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine ASVs that hit to Moeller co-div clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 09/15/2020 06:46:46\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb/analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "New DB title:  analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb/analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 208 sequences in 0.00554895 seconds.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: analyses/codiv_moeller_ASVs: File exists\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir analyses/codiv_moeller_ASVs\n",
    "cp inputs/moeller_codiv_Bacteroidaceae.fna analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
    "\n",
    "#make blastdb\n",
    "makeblastdb -in analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna -dbtype nucl\n",
    "\n",
    "#blast moeller co-div seqs\n",
    "blastn -query inputs/ASVs_filtered.fasta -db analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna \\\n",
    "-outfmt \"7 qseqid salltitles sseqid pident length qlen evalue\" \\\n",
    "-out analyses/codiv_moeller_ASVs/codiv_blastout_ASVs.txt \\\n",
    "-max_target_seqs 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "314 ASVs matching co-div clades with greater than 95 percent identity\n",
      "Bt2_clade1_bonobo     90\n",
      "Bt3_clade1_human      85\n",
      "Bt2_clade2_bonobo     33\n",
      "Bt1_clade1_bonobo     23\n",
      "Bt2_clade1_chimp      20\n",
      "Bt3_clade1_chimp      17\n",
      "Bt3_clade1_bonobo     15\n",
      "Bt1_clade1_gorilla    10\n",
      "Bt1_clade1_chimp       8\n",
      "Bt2_clade1_gorilla     7\n",
      "Bt2_clade2_chimp       6\n",
      "Name: codiv_clade, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#filter blast output to top hit per ASV, remove hits not pass thresholds\n",
    "blast_res = pd.read_csv('analyses/codiv_moeller_ASVs/codiv_blastout_ASVs.txt',sep='\\t',comment='#',header=None)\n",
    "blast_res.columns = ['ASV','codiv_clade_seq','sseqid','pident','length','qlen','evalue']\n",
    "blast_res = blast_res.groupby('ASV').head(1).reset_index(drop=True) #get top hit\n",
    "blast_res_pass = blast_res[blast_res['pident']>pident_cutoff].reset_index(drop=True) #QC filter\n",
    "blast_res_pass = blast_res_pass[blast_res_pass['length']>200].reset_index(drop=True) #QC filter\n",
    "print(len(blast_res_pass),'ASVs matching co-div clades with greater than',pident_cutoff,'percent identity')\n",
    "blast_res_pass['codiv_clade'] = blast_res_pass['codiv_clade_seq'].apply(lambda x: x.split(' ')[1])\n",
    "blast_res_pass['lineage'] = blast_res_pass['codiv_clade'].apply(lambda x: x.split('_')[0]) \n",
    "blast_res_pass = blast_res_pass[['ASV','codiv_clade_seq', 'codiv_clade','lineage']]\n",
    "#print(blast_res_pass.head())\n",
    "print(blast_res_pass['codiv_clade'].value_counts())\n",
    "#output hits passing that will be used to determine the mcra of the lineage\n",
    "blast_res_pass.to_csv('analyses/codiv_moeller_ASVs/codiv_clades_ASVs.txt',sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine HR clades that hit to Moeller co-div clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(359, 27)\n",
      "codiv_clade         lineage\n",
      "Bt1_clade1_bonobo   Bt1         1\n",
      "Bt1_clade1_chimp    Bt1         1\n",
      "Bt1_clade1_gorilla  Bt1         1\n",
      "Bt2_clade1_bonobo   Bt2         1\n",
      "Bt2_clade1_chimp    Bt2         2\n",
      "Bt2_clade1_gorilla  Bt2         1\n",
      "Bt2_clade2_bonobo   Bt2         1\n",
      "Bt2_clade2_chimp    Bt2         1\n",
      "Bt3_clade1_bonobo   Bt3         1\n",
      "Bt3_clade1_chimp    Bt3         2\n",
      "Bt3_clade1_human    Bt3        15\n",
      "dtype: int64\n"
     ]
    }
   ],
   "source": [
    "print(clades_df.shape)\n",
    "clades_ASVs_codiv = clades_ASVs_df.merge(blast_res_pass, how='left', left_on='ASVs',right_on='ASV')\n",
    "clades_ASVs_codiv = clades_ASVs_codiv.drop(['ASV'],axis=1)\n",
    "clades_ASVs_codiv = clades_ASVs_codiv[~clades_ASVs_codiv['lineage'].isna()]\n",
    "clades_ASVs_codiv = clades_ASVs_codiv[['cladeName','codiv_clade','lineage']].drop_duplicates()\n",
    "clades_df_codiv = clades_df.merge(clades_ASVs_codiv, how='left',on='cladeName')\n",
    "print(clades_df_codiv.groupby(['codiv_clade','lineage']).size())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Output table for Figure 2 \n",
    "clades_df_sh = clades_df_codiv[['cladeName', 'cladeTax', 'sampleNum', 'ASVsNum', \n",
    "      'HR_sampleTypes', 'HR_sampleNum','HR_cat', 'HR_type', \n",
    "      'CP_pres','CP_prominent', 'CP_sampleTypes', 'CP_sampleNum','captiveNames',\n",
    "       'captive_bonobo', 'captive_chimp', 'captive_gorilla',\n",
    "       'captive_orangutan', 'non_western_human', 'western_human',\n",
    "       'wild_bonobo', 'wild_chimp', 'wild_gorilla',\n",
    "       'heatmap_col1','heatmap_col2','heatmap_col3',\n",
    "        'codiv_clade','lineage']]\n",
    "clades_df_sh.to_csv('analyses/figures/HRclades_Figure2_table.txt',sep='\\t',index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### search full tree for host-restricted clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "collasped_tree = Tree(full_tree_file, format=0)\n",
    "\n",
    "def collapse_node(tree,clade_ASVs,cladeName):\n",
    "    if len(clade_ASVs) > 1:\n",
    "        node = tree.get_common_ancestor(clade_ASVs)\n",
    "        node.name = cladeName\n",
    "        children = node.get_children()\n",
    "        node.remove_child(children[1])\n",
    "        node.remove_child(children[0])\n",
    "    else:\n",
    "        ASV = clade_ASVs[0]\n",
    "        leaf = tree.get_leaves_by_name(name=ASV)[0]\n",
    "        leaf.name = cladeName\n",
    "\n",
    "for clade,row in clades_df.iterrows():\n",
    "    collapse_node(collasped_tree,row['ASVs'],row['cladeName'])\n",
    "\n",
    "clade_leaves = [leaf for leaf in collasped_tree.get_leaves() if 'clade' in leaf.name] \n",
    "collasped_tree.prune(clade_leaves) #eliminate leftover ASVs and ref taxa\n",
    "collasped_tree.write(format=2, outfile='analyses/figures/HRclades_Figure2.tre')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Table 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "359.0\n",
      "cladeTax\n",
      "Bacteroidaceae_Prevotella              101.0\n",
      "Bacteroidaceae_Unassigned               21.0\n",
      "Porphyromonadaceae_Porphyromonas         9.0\n",
      "Rikenellaceae_Alistipes                 22.0\n",
      "Tannerellaceae_Parabacteroides          24.0\n",
      "Tannerellaceae_Unassigned               10.0\n",
      "Unassigned_Unassigned                    7.0\n",
      "WCHB1-69_Unassigned                     10.0\n",
      "Bacteroidaceae_Bacteroides_combined    131.0\n",
      "dtype: float64\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "359.0"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Table2 = clades_df.groupby(['cladeTax','HR_type']).size().reset_index(name=\"count\")\n",
    "Table2 = Table2.pivot(index='cladeTax', columns='HR_type')['count'].fillna(0)\n",
    "\n",
    "Table2['MX'] = Table2[['MX_2_wild_apes', 'MX_3_wild_apes',\n",
    "                 'MX_human_2_wild_apes', 'MX_human_single_wild_ape']].sum(axis=1)\n",
    "Table2 = Table2[['HR_human', 'HR_wild_bonobo', 'HR_wild_chimp', \n",
    "        'HR_wild_gorilla','Unique_CP', 'MX']]\n",
    "\n",
    "Bacteroides_genera = ['Bacteroidaceae_Bacteroides','Bacteroidaceae_Bacteroides_A','Bacteroidaceae_Bacteroides_B']\n",
    "Bacteroides_row = pd.Series(Table2.loc[Bacteroides_genera,:].sum(),name='Bacteroidaceae_Bacteroides_combined')\n",
    "Table2 = Table2.append(Bacteroides_row).drop(Bacteroides_genera)\n",
    "\n",
    "print(Table2.sum(axis=1).sum())\n",
    "\n",
    "taxa_over_5 = Table2[Table2.sum(axis=1)>5].sum(axis=1)\n",
    "print(taxa_over_5)\n",
    "sel_genera = ['Bacteroidaceae_Bacteroides_combined',\n",
    "              'Bacteroidaceae_Prevotella',\n",
    "              'Rikenellaceae_Alistipes',\n",
    "              'Tannerellaceae_Parabacteroides',\n",
    "              'Bacteroidaceae_Unassigned',\n",
    "              'WCHB1-69_Unassigned']\n",
    "Table2_sel_genera = Table2.loc[sel_genera,:]\n",
    "Table2_other_genera = pd.Series(Table2.drop(sel_genera).sum(),name='Other_genera')\n",
    "Table2_sel_genera = Table2_sel_genera.append(Table2_other_genera)\n",
    "Table2_sel_genera.sum(axis=1).sum()\n",
    "#Table2_sel_genera.sum(axis=1).sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7148\n",
      "  cladeName                                        sampleNames  sampleNum  \\\n",
      "0  clade_41  [NielsenHB_2014__O2_UC60_2, ZellerG_2014__CCIS...        788   \n",
      "1  clade_41  [NielsenHB_2014__O2_UC60_2, ZellerG_2014__CCIS...        788   \n",
      "2  clade_41  [NielsenHB_2014__O2_UC60_2, ZellerG_2014__CCIS...        788   \n",
      "3  clade_41  [NielsenHB_2014__O2_UC60_2, ZellerG_2014__CCIS...        788   \n",
      "4  clade_41  [NielsenHB_2014__O2_UC60_2, ZellerG_2014__CCIS...        788   \n",
      "\n",
      "   ASVsNum HR_sampleTypes  HR_sampleNum HR_cat   HR_type  CP_pres  \\\n",
      "0      366        [human]           788     HR  HR_human    False   \n",
      "1      366        [human]           788     HR  HR_human    False   \n",
      "2      366        [human]           788     HR  HR_human    False   \n",
      "3      366        [human]           788     HR  HR_human    False   \n",
      "4      366        [human]           788     HR  HR_human    False   \n",
      "\n",
      "   CP_sampleNum CP_sampleTypes captiveNames      ASVs  \n",
      "0             0             []           []  ASV_2623  \n",
      "1             0             []           []  ASV_1788  \n",
      "2             0             []           []  ASV_7289  \n",
      "3             0             []           []  ASV_7288  \n",
      "4             0             []           []  ASV_7176  \n",
      "6438\n",
      "      ASV                codiv_clade_seq       codiv_clade lineage\n",
      "0   ASV_1  Human6700665 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "1  ASV_13  Human7859459 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "2  ASV_34  Human9825266 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "3  ASV_37  Human8829943 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "4  ASV_73  Human9825266 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "314\n",
      "         ASV   cladeName                                        sampleNames  \\\n",
      "0  ASV_10387  clade_8751                           [SchirmerM_2016__G88949]   \n",
      "1    ASV_408  clade_8751  [CosteaPI_2017__SID713A037-11-0-0, CosteaPI_20...   \n",
      "2   ASV_3299  clade_8751  [LeChatelierE_2013__MH0085, NielsenHB_2014__MH...   \n",
      "3  ASV_10369  clade_8751                           [SchirmerM_2016__G89027]   \n",
      "4  ASV_10428  clade_8751                           [SchirmerM_2016__G89097]   \n",
      "\n",
      "   sampleNum  ASVsNum HR_sampleTypes  HR_sampleNum HR_cat   HR_type CP_pres  \\\n",
      "0     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "1     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "2     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "3     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "4     1751.0    191.0        [human]        1732.0     HR  HR_human    True   \n",
      "\n",
      "   ...  lineage             Family          Genus captive_chimp_HOUZ  \\\n",
      "0  ...      NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "1  ...      NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "2  ...      NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "3  ...      NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "4  ...      NaN  f__Bacteroidaceae  g__Prevotella                0.0   \n",
      "\n",
      "  captive_gorilla_HOUZ captive_orangutan_HOUZ captive_bonobo_COLZ  \\\n",
      "0                  0.0                    0.0                 0.0   \n",
      "1                  0.0                    0.0                 0.0   \n",
      "2                  0.0                    0.0                 0.0   \n",
      "3                  0.0                    0.0                 0.0   \n",
      "4                  0.0                    0.0                 0.0   \n",
      "\n",
      "  captive_gorilla_COLZ captive_orangutan_COLZ  captive_chimp_PC  \n",
      "0                  0.0                    0.0               0.0  \n",
      "1                  0.0                    0.0               0.0  \n",
      "2                  0.0                    0.0               0.0  \n",
      "3                  0.0                    0.0               0.0  \n",
      "4                  0.0                    0.0               0.0  \n",
      "\n",
      "[5 rows x 26 columns]\n"
     ]
    }
   ],
   "source": [
    "#taxonomic info, family and genus\n",
    "tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "tax_fam_dict = dict(zip(tax_table['ASV'], tax_table['Family']))\n",
    "tax_gen_dict = dict(zip(tax_table['ASV'], tax_table['Genus']))\n",
    "\n",
    "#sample to sample type category\n",
    "metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "sample_type_dict = dict(zip(metadata['X.SampleID'], metadata['Description']))\n",
    "\n",
    "asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "\n",
    "sampleNames = asv_table.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "ASV_sampleName_dict = dict(zip(sampleNames.index,sampleNames))\n",
    "\n",
    "allASVs = [leaf.name for leaf in full_tree.get_leaves() if 'ASV' in leaf.name]\n",
    "allASVs = pd.DataFrame(allASVs,columns=['ASV'])\n",
    "print(len(allASVs))\n",
    "\n",
    "#Host restricted clades \n",
    "print(clades_ASVs_df.head())\n",
    "print(len(clades_ASVs_df))\n",
    "#Moeller codiv clades\n",
    "print(blast_res_pass.head())\n",
    "print(len(blast_res_pass))\n",
    "\n",
    "#combine\n",
    "allASVs_HRclade = allASVs.merge(clades_ASVs_df, how='left', left_on='ASV',right_on='ASVs')\n",
    "allASVs_HRclade_codiv = allASVs_HRclade.merge(blast_res_pass, how='left', on='ASV')\n",
    "allASVs_HRclade_codiv.head()\n",
    "\n",
    "#add taxonomy\n",
    "allASVs_HRclade_codiv['Family'] = allASVs_HRclade_codiv['ASV'].apply(lambda ASV: tax_fam_dict[ASV])\n",
    "allASVs_HRclade_codiv['Genus'] = allASVs_HRclade_codiv['ASV'].apply(lambda ASV: tax_gen_dict[ASV])\n",
    "\n",
    "#add captive sp and site info\n",
    "metadata['Description_site'] = metadata['Description']+'_' +metadata['site_code']\n",
    "sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "\n",
    "allASVs_HRclade_codiv['sampleNames'] = allASVs_HRclade_codiv['ASV'].apply(lambda x: ASV_sampleName_dict[x])\n",
    "\n",
    "description_df = allASVs_HRclade_codiv['sampleNames'].apply(lambda l: pd.Series(\n",
    "    [sample_type_site_dict[name] for name in l]).value_counts())\n",
    "description_df = description_df.fillna(0) \n",
    "description_cp_df= description_df[['captive_chimp_HOUZ','captive_gorilla_HOUZ','captive_orangutan_HOUZ',\n",
    "'captive_bonobo_COLZ','captive_gorilla_COLZ','captive_orangutan_COLZ','captive_chimp_PC']]\n",
    "allASVs_HRclade_codiv_cp = allASVs_HRclade_codiv.merge(description_cp_df, left_index=True,right_index=True)\n",
    "\n",
    "allASVs_HRclade_codiv_cp.to_csv('analyses/codiv_moeller_ASVs/full_tree_ASV_table.txt',sep='\\t',index=False)\n",
    "print(allASVs_HRclade_codiv_cp.head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>HR_type</th>\n",
       "      <th>HR_human</th>\n",
       "      <th>HR_wild_bonobo</th>\n",
       "      <th>HR_wild_chimp</th>\n",
       "      <th>HR_wild_gorilla</th>\n",
       "      <th>MX_2_wild_apes</th>\n",
       "      <th>MX_3_wild_apes</th>\n",
       "      <th>MX_human_2_wild_apes</th>\n",
       "      <th>MX_human_single_wild_ape</th>\n",
       "      <th>Unique_CP</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>cladeTax</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Bacteroides</th>\n",
       "      <td>75.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>20.0</td>\n",
       "      <td>4.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Bacteroides_B</th>\n",
       "      <td>15.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Prevotella</th>\n",
       "      <td>75.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>9.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Bacteroidaceae_Unassigned</th>\n",
       "      <td>13.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Rikenellaceae_Alistipes</th>\n",
       "      <td>17.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>5.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tannerellaceae_Parabacteroides</th>\n",
       "      <td>18.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Tannerellaceae_Unassigned</th>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>WCHB1-69_Unassigned</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "HR_type                         HR_human  HR_wild_bonobo  HR_wild_chimp  \\\n",
       "cladeTax                                                                  \n",
       "Bacteroidaceae_Bacteroides          75.0             0.0            1.0   \n",
       "Bacteroidaceae_Bacteroides_B        15.0             0.0            0.0   \n",
       "Bacteroidaceae_Prevotella           75.0             8.0            9.0   \n",
       "Bacteroidaceae_Unassigned           13.0             3.0            1.0   \n",
       "Rikenellaceae_Alistipes             17.0             0.0            0.0   \n",
       "Tannerellaceae_Parabacteroides      18.0             1.0            2.0   \n",
       "Tannerellaceae_Unassigned            3.0             0.0            4.0   \n",
       "WCHB1-69_Unassigned                  0.0             0.0            0.0   \n",
       "\n",
       "HR_type                         HR_wild_gorilla  MX_2_wild_apes  \\\n",
       "cladeTax                                                          \n",
       "Bacteroidaceae_Bacteroides                  0.0             0.0   \n",
       "Bacteroidaceae_Bacteroides_B                0.0             0.0   \n",
       "Bacteroidaceae_Prevotella                   4.0             3.0   \n",
       "Bacteroidaceae_Unassigned                   3.0             0.0   \n",
       "Rikenellaceae_Alistipes                     0.0             0.0   \n",
       "Tannerellaceae_Parabacteroides              0.0             0.0   \n",
       "Tannerellaceae_Unassigned                   1.0             0.0   \n",
       "WCHB1-69_Unassigned                         7.0             3.0   \n",
       "\n",
       "HR_type                         MX_3_wild_apes  MX_human_2_wild_apes  \\\n",
       "cladeTax                                                               \n",
       "Bacteroidaceae_Bacteroides                 0.0                   5.0   \n",
       "Bacteroidaceae_Bacteroides_B               0.0                   0.0   \n",
       "Bacteroidaceae_Prevotella                  0.0                   1.0   \n",
       "Bacteroidaceae_Unassigned                  0.0                   0.0   \n",
       "Rikenellaceae_Alistipes                    0.0                   0.0   \n",
       "Tannerellaceae_Parabacteroides             0.0                   0.0   \n",
       "Tannerellaceae_Unassigned                  0.0                   0.0   \n",
       "WCHB1-69_Unassigned                        0.0                   0.0   \n",
       "\n",
       "HR_type                         MX_human_single_wild_ape  Unique_CP  \n",
       "cladeTax                                                             \n",
       "Bacteroidaceae_Bacteroides                          20.0        4.0  \n",
       "Bacteroidaceae_Bacteroides_B                         3.0        0.0  \n",
       "Bacteroidaceae_Prevotella                            1.0        0.0  \n",
       "Bacteroidaceae_Unassigned                            1.0        0.0  \n",
       "Rikenellaceae_Alistipes                              5.0        0.0  \n",
       "Tannerellaceae_Parabacteroides                       3.0        0.0  \n",
       "Tannerellaceae_Unassigned                            2.0        0.0  \n",
       "WCHB1-69_Unassigned                                  0.0        0.0  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "colsum = Table2.sum(axis=1)\n",
    "Bacteroidaceae_Bacteroides\n",
    "Table2[Table2.sum(axis=1) >= 10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
