Re-pair reads in disordered fastq (repair.sh from bbmap v38.70)
Demultiplex fastqs with barcode-splitter(v0.18.6) or demultiplex(v1.0.1)
Remove primers with cutadapt (v2.5)
Read trimming, error correction, chimera removal and generate ASV table with DADA2(v1.16.0)
Align sequences with mafft(v7.309)
Generate ASV phylogenies with FastTree(v2.1.9)
General data processing in R with tidyverse(v1.3.0), stringr(v1.4.0), reshape2(v1.4.4), rstatix(v0.6.0)
Processing microbiome datasets in R with phyloseq(v1.32.0), DECIPHER(v2.16.1), genefilter(v1.70.0), seqinr(v3.6-1), ape(v5.4-1), phytools(v0.7-70), zoo(v1.8-8), picante(v1.8.2)
Sequence analysis with prodigal(v2.6.3), hmmer(v3.3), transeq and transalign from EMBOSS(v6.6.0.0), BLAST(v2.9.0)
General data processing in python with pandas, numpy
Sequence analysis and defining HR clades in python with Biopython(v1.77), ete3(v3.1.1)
Statistical analysis in R with PMCMRplus(1.5.1), RVAideMemoire(v0.9-78), broom(0.7.1)
Data visualization in R with ggplot2(v3.3.3), ggtree(v2.2.4), cowplot(v1.1.0), RColorBrewer(v1.1-2)

