## Captivity and the Co-diversification of Great Ape Microbiomes
### Authors: Alex H. Nishida & Howard Ochman

#### This manuscript is currently is under-consideration.

#### About this repository
This repository contains scripts for processing raw fastq data as well as scripts for analyzing data to generate figures and tables for the manuscript.

#### Original fastq files:
The original Fastq files were submitted to the NCBI sequence read archive under BioProject accession numbers PRJNA692991, PRJNA693013.

#### Published fastq data:
Published fastq metadata are listed in Table S1

#### File structure
This study analyzes two seperate datasets - 16S and gyrb amplicon datasets.

Scripts are divided into two folders processing and analyses

Dependencies can be found in the dependencies.md file

Processing scripts convert raw fastq to ASV table with associated taxonomy, metadata and phylogeny located in the gyrb and 16s inputs folders

Intermediate files generated by processing scripts were not uploaded due to size

Analyses scripts utilize files in the gyrb and 16s inputs folders to generate all figures and run all statistical tests

### To reproduce all figures and statistics reported in the manuscript 
[30 minutes from install to analysis completion]

Check the dependencies.md file to obtain python modules and R libraries necessary for running analyses scripts

##### commands
###### git clone https://github.com/ahnishida/captive_ape_microbiome.git
###### cd captive_ape_microbiome
###### gunzip results/gyrb/inputs/physeq_Bacteroidales_asv_tab.txt.gz
###### ./scripts/analyses/run_all_analyses.sh