{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script finds the co-diversifying clades within Moeller et al. supplemental fasta by creates a phylogeny and selecting the subtrees that match the topology of figure 1. \n",
    "\n",
    "Inputs: fasta of all Bacteroidaceae seqs, ref_gyrb_gtdbtk Bacteroidales reference seqs (fna and faa)\n",
    "\n",
    "Outputs: fasta of just co-div seqs with lineage and clade labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "from ete3 import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome')\n",
    "outdir = 'results/gyrb/processing/moeller_sup'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate phylogeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 2.5 with Python 3.7.7\n",
      "Command line parameters: -g CGGAGGTAARTTCGAYAAAGG --overlap 21 -e .15 --discard-untrimmed -o results/gyrb/processing/moeller_sup/Bacteroidaceae.fna data/gyrb/moeller_sup/Bacteroidaceae.fna\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.46 s (638 us/read; 0.09 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                     724\n",
      "Reads with adapters:                       724 (100.0%)\n",
      "Reads written (passing filters):           724 (100.0%)\n",
      "\n",
      "Total basepairs processed:       217,200 bp\n",
      "Total written (filtered):        200,991 bp (92.5%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CGGAGGTAARTTCGAYAAAGG; Type: regular 5'; Length: 21; Trimmed: 724 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-5 bp: 0; 6-12 bp: 1; 13-19 bp: 2; 20-21 bp: 3\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "21\t194\t0.0\t3\t194\n",
      "22\t192\t0.0\t3\t192\n",
      "23\t201\t0.0\t3\t201\n",
      "24\t137\t0.0\t3\t137\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Translate nucleic acid sequences\n",
      "Generate an alignment of nucleic coding regions from aligned proteins\n",
      "Error: Guide protein sequence Gorilla3374716_2 not found in nucleic sequence Gorilla3374716\n",
      "Error: Guide protein sequence Human3058374_2 not found in nucleic sequence Human3058374\n",
      "Error: Guide protein sequence Human9403092_2 not found in nucleic sequence Human9403092\n",
      "Error: Guide protein sequence Chimp10413648_2 not found in nucleic sequence Chimp10413648\n",
      "Error: Guide protein sequence Chimp10757854_2 not found in nucleic sequence Chimp10757854\n",
      "Error: Guide protein sequence Chimp10796222_2 not found in nucleic sequence Chimp10796222\n",
      "Error: Guide protein sequence Chimp11562025_2 not found in nucleic sequence Chimp11562025\n",
      "Error: Guide protein sequence Chimp3146215_2 not found in nucleic sequence Chimp3146215\n",
      "Error: Guide protein sequence Chimp8175258_2 not found in nucleic sequence Chimp8175258\n",
      "Error: Guide protein sequence Bonobo2919026_2 not found in nucleic sequence Bonobo2919026\n",
      "Error: Guide protein sequence Bonobo4229344_2 not found in nucleic sequence Bonobo4229344\n",
      "Error: Guide protein sequence Bonobo6282208_2 not found in nucleic sequence Bonobo6282208\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "outdir=results/gyrb/processing/moeller_sup\n",
    "ref_faa=results/gyrb/processing/ref_gyrb_gtdbtk/gyrb_fastas/gtdbtk_gyrb_Bt.faa\n",
    "ref_fna=results/gyrb/processing/ref_gyrb_gtdbtk/gyrb_fastas/gtdbtk_gyrb_Bt.fasta\n",
    "\n",
    "mkdir -pv $outdir\n",
    "\n",
    "#trim primer sequences\n",
    "cutadapt  -g CGGAGGTAARTTCGAYAAAGG  --overlap 21 -e .15 --discard-untrimmed  -o  $outdir/Bacteroidaceae.fna  data/gyrb/moeller_sup/Bacteroidaceae.fna\n",
    "\n",
    "#translate gyrb amplicon seqs\n",
    "transeq -frame 2 -sequence $outdir/Bacteroidaceae.fna  -outseq $outdir/Bacteroidaceae.faa\n",
    "\n",
    "#combine gyrb amplicon and ref seqs\n",
    "cat $ref_faa $outdir/Bacteroidaceae.faa > $outdir/Bacteroidaceae_ref.faa\n",
    "cat $ref_fna $outdir/Bacteroidaceae.fna > $outdir/Bacteroidaceae_ref.fna\n",
    "\n",
    "#align seqs\n",
    "mafft --auto --quiet $outdir/Bacteroidaceae_ref.faa > $outdir/Bacteroidaceae_ref.faa.aln\n",
    "tranalign -asequence  $outdir/Bacteroidaceae_ref.fna -bsequence $outdir/Bacteroidaceae_ref.faa.aln -outseq $outdir/Bacteroidaceae_ref.fna.aln\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#trim start of gyrB gene\n",
    "def trim_aln(original_fasta,trim_fasta,start_pos,end_pos):\n",
    "    with open(original_fasta) as original_fasta:\n",
    "        with open(trim_fasta, 'w') as trim_fasta:\n",
    "            records = SeqIO.parse(original_fasta, 'fasta')\n",
    "            for record in records:\n",
    "                record.seq = record.seq[start_pos:end_pos] \n",
    "                SeqIO.write(record, trim_fasta, 'fasta')\n",
    "trim_aln(f'{outdir}/Bacteroidaceae_ref.fna.aln',f'{outdir}/Bacteroidaceae_ref.fna.aln.trim',429,2952)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FastTree Version 2.1.10 Double precision (No SSE3)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: SH-like 1000\n",
      "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
      "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "      0.17 seconds: Top hits for    121 of   1529 seqs (at seed    100)\n",
      "      0.30 seconds: Top hits for    298 of   1529 seqs (at seed    200)\n",
      "      0.45 seconds: Top hits for    460 of   1529 seqs (at seed    300)\n",
      "      0.59 seconds: Top hits for    637 of   1529 seqs (at seed    400)\n",
      "      0.70 seconds: Top hits for    798 of   1529 seqs (at seed    500)\n",
      "      0.81 seconds: Top hits for    939 of   1529 seqs (at seed    600)\n",
      "      0.92 seconds: Top hits for   1091 of   1529 seqs (at seed    700)\n",
      "      1.06 seconds: Top hits for   1203 of   1529 seqs (at seed    900)\n",
      "      1.46 seconds: Joined    100 of   1526\n",
      "      1.81 seconds: Joined    200 of   1526\n",
      "      2.18 seconds: Joined    300 of   1526\n",
      "      2.53 seconds: Joined    400 of   1526\n",
      "      2.76 seconds: Joined    500 of   1526\n",
      "      3.08 seconds: Joined    600 of   1526\n",
      "      3.54 seconds: Joined    700 of   1526\n",
      "      4.13 seconds: Joined    800 of   1526\n",
      "      4.72 seconds: Joined    900 of   1526\n",
      "      5.27 seconds: Joined   1000 of   1526\n",
      "      5.77 seconds: Joined   1100 of   1526\n",
      "      6.36 seconds: Joined   1200 of   1526\n",
      "      7.01 seconds: Joined   1300 of   1526\n",
      "      7.43 seconds: Joined   1400 of   1526\n",
      "      8.00 seconds: Joined   1500 of   1526\n",
      "Initial topology in 8.05 seconds\n",
      "Refining topology: 42 rounds ME-NNIs, 2 rounds ME-SPRs, 21 rounds ML-NNIs\n",
      "      8.10 seconds: ME NNI round 1 of 42, 301 of 1527 splits, 79 changes (max delta 0.020)\n",
      "      8.21 seconds: ME NNI round 1 of 42, 1001 of 1527 splits, 312 changes (max delta 0.020)\n",
      "      8.33 seconds: ME NNI round 1 of 42, 1501 of 1527 splits, 405 changes (max delta 0.066)\n",
      "      8.44 seconds: ME NNI round 2 of 42, 501 of 1527 splits, 56 changes (max delta 0.021)\n",
      "      8.55 seconds: ME NNI round 2 of 42, 1101 of 1527 splits, 184 changes (max delta 0.100)\n",
      "      8.65 seconds: ME NNI round 3 of 42, 101 of 1527 splits, 1 changes (max delta 0.002)\n",
      "      8.76 seconds: ME NNI round 3 of 42, 801 of 1527 splits, 118 changes (max delta 0.023)\n",
      "      8.88 seconds: ME NNI round 4 of 42, 201 of 1527 splits, 17 changes (max delta 0.018)\n",
      "      8.98 seconds: ME NNI round 5 of 42, 1 of 1527 splits\n",
      "      9.10 seconds: ME NNI round 6 of 42, 1 of 1527 splits\n",
      "      9.21 seconds: ME NNI round 7 of 42, 101 of 1527 splits, 3 changes (max delta 0.003)\n",
      "      9.31 seconds: ME NNI round 10 of 42, 1 of 1527 splits\n",
      "      9.74 seconds: SPR round   1 of   2, 101 of 3056 nodes\n",
      "     10.13 seconds: SPR round   1 of   2, 201 of 3056 nodes\n",
      "     10.43 seconds: SPR round   1 of   2, 301 of 3056 nodes\n",
      "     10.77 seconds: SPR round   1 of   2, 401 of 3056 nodes\n",
      "     11.00 seconds: SPR round   1 of   2, 501 of 3056 nodes\n",
      "     11.38 seconds: SPR round   1 of   2, 601 of 3056 nodes\n",
      "     11.71 seconds: SPR round   1 of   2, 701 of 3056 nodes\n",
      "     11.89 seconds: SPR round   1 of   2, 801 of 3056 nodes\n",
      "     12.14 seconds: SPR round   1 of   2, 901 of 3056 nodes\n",
      "     12.49 seconds: SPR round   1 of   2, 1001 of 3056 nodes\n",
      "     12.73 seconds: SPR round   1 of   2, 1101 of 3056 nodes\n",
      "     13.04 seconds: SPR round   1 of   2, 1201 of 3056 nodes\n",
      "     13.30 seconds: SPR round   1 of   2, 1301 of 3056 nodes\n",
      "     13.60 seconds: SPR round   1 of   2, 1401 of 3056 nodes\n",
      "     13.84 seconds: SPR round   1 of   2, 1501 of 3056 nodes\n",
      "     14.03 seconds: SPR round   1 of   2, 1601 of 3056 nodes\n",
      "     14.43 seconds: SPR round   1 of   2, 1701 of 3056 nodes\n",
      "     14.70 seconds: SPR round   1 of   2, 1801 of 3056 nodes\n",
      "     14.91 seconds: SPR round   1 of   2, 1901 of 3056 nodes\n",
      "     15.22 seconds: SPR round   1 of   2, 2001 of 3056 nodes\n",
      "     15.61 seconds: SPR round   1 of   2, 2101 of 3056 nodes\n",
      "     15.97 seconds: SPR round   1 of   2, 2201 of 3056 nodes\n",
      "     16.39 seconds: SPR round   1 of   2, 2301 of 3056 nodes\n",
      "     16.77 seconds: SPR round   1 of   2, 2401 of 3056 nodes\n",
      "     16.93 seconds: SPR round   1 of   2, 2501 of 3056 nodes\n",
      "     17.08 seconds: SPR round   1 of   2, 2601 of 3056 nodes\n",
      "     17.39 seconds: SPR round   1 of   2, 2701 of 3056 nodes\n",
      "     17.75 seconds: SPR round   1 of   2, 2801 of 3056 nodes\n",
      "     18.12 seconds: SPR round   1 of   2, 2901 of 3056 nodes\n",
      "     18.48 seconds: SPR round   1 of   2, 3001 of 3056 nodes\n",
      "     18.67 seconds: ME NNI round 15 of 42, 1 of 1527 splits\n",
      "     18.77 seconds: ME NNI round 15 of 42, 601 of 1527 splits, 12 changes (max delta 0.006)\n",
      "     18.87 seconds: ME NNI round 15 of 42, 1201 of 1527 splits, 37 changes (max delta 0.006)\n",
      "     18.99 seconds: ME NNI round 16 of 42, 301 of 1527 splits, 4 changes (max delta 0.003)\n",
      "     19.10 seconds: ME NNI round 16 of 42, 1001 of 1527 splits, 17 changes (max delta 0.003)\n",
      "     19.22 seconds: ME NNI round 17 of 42, 101 of 1527 splits, 7 changes (max delta 0.006)\n",
      "     19.71 seconds: SPR round   2 of   2, 101 of 3056 nodes\n",
      "     20.07 seconds: SPR round   2 of   2, 201 of 3056 nodes\n",
      "     20.34 seconds: SPR round   2 of   2, 301 of 3056 nodes\n",
      "     20.74 seconds: SPR round   2 of   2, 401 of 3056 nodes\n",
      "     20.97 seconds: SPR round   2 of   2, 501 of 3056 nodes\n",
      "     21.29 seconds: SPR round   2 of   2, 601 of 3056 nodes\n",
      "     21.62 seconds: SPR round   2 of   2, 701 of 3056 nodes\n",
      "     21.77 seconds: SPR round   2 of   2, 801 of 3056 nodes\n",
      "     22.02 seconds: SPR round   2 of   2, 901 of 3056 nodes\n",
      "     22.34 seconds: SPR round   2 of   2, 1001 of 3056 nodes\n",
      "     22.56 seconds: SPR round   2 of   2, 1101 of 3056 nodes\n",
      "     22.86 seconds: SPR round   2 of   2, 1201 of 3056 nodes\n",
      "     23.10 seconds: SPR round   2 of   2, 1301 of 3056 nodes\n",
      "     23.37 seconds: SPR round   2 of   2, 1401 of 3056 nodes\n",
      "     23.59 seconds: SPR round   2 of   2, 1501 of 3056 nodes\n",
      "     23.76 seconds: SPR round   2 of   2, 1601 of 3056 nodes\n",
      "     24.16 seconds: SPR round   2 of   2, 1701 of 3056 nodes\n",
      "     24.42 seconds: SPR round   2 of   2, 1801 of 3056 nodes\n",
      "     24.62 seconds: SPR round   2 of   2, 1901 of 3056 nodes\n",
      "     24.89 seconds: SPR round   2 of   2, 2001 of 3056 nodes\n",
      "     25.36 seconds: SPR round   2 of   2, 2101 of 3056 nodes\n",
      "     25.85 seconds: SPR round   2 of   2, 2201 of 3056 nodes\n",
      "     26.24 seconds: SPR round   2 of   2, 2301 of 3056 nodes\n",
      "     26.61 seconds: SPR round   2 of   2, 2401 of 3056 nodes\n",
      "     26.77 seconds: SPR round   2 of   2, 2501 of 3056 nodes\n",
      "     26.91 seconds: SPR round   2 of   2, 2601 of 3056 nodes\n",
      "     27.21 seconds: SPR round   2 of   2, 2701 of 3056 nodes\n",
      "     27.59 seconds: SPR round   2 of   2, 2801 of 3056 nodes\n",
      "     27.95 seconds: SPR round   2 of   2, 2901 of 3056 nodes\n",
      "     28.28 seconds: SPR round   2 of   2, 3001 of 3056 nodes\n",
      "     28.49 seconds: ME NNI round 29 of 42, 1 of 1527 splits\n",
      "     28.60 seconds: ME NNI round 29 of 42, 601 of 1527 splits, 23 changes (max delta 0.009)\n",
      "     28.70 seconds: ME NNI round 29 of 42, 1201 of 1527 splits, 43 changes (max delta 0.014)\n",
      "     28.81 seconds: ME NNI round 30 of 42, 301 of 1527 splits, 11 changes (max delta 0.003)\n",
      "     28.92 seconds: ME NNI round 30 of 42, 1001 of 1527 splits, 31 changes (max delta 0.003)\n",
      "     29.04 seconds: ME NNI round 31 of 42, 101 of 1527 splits, 11 changes (max delta 0.011)\n",
      "     29.14 seconds: ME NNI round 35 of 42, 1 of 1527 splits\n",
      "Total branch-length 107.477 after 29.54 sec\n",
      "     29.58 seconds: ML Lengths 1 of 1527 splits\n",
      "     29.76 seconds: ML Lengths 101 of 1527 splits\n",
      "     29.94 seconds: ML Lengths 201 of 1527 splits\n",
      "     30.09 seconds: ML Lengths 301 of 1527 splits\n",
      "     30.25 seconds: ML Lengths 401 of 1527 splits\n",
      "     30.41 seconds: ML Lengths 501 of 1527 splits\n",
      "     30.59 seconds: ML Lengths 601 of 1527 splits\n",
      "     30.74 seconds: ML Lengths 701 of 1527 splits\n",
      "     30.90 seconds: ML Lengths 801 of 1527 splits\n",
      "     31.06 seconds: ML Lengths 901 of 1527 splits\n",
      "     31.22 seconds: ML Lengths 1001 of 1527 splits\n",
      "     31.39 seconds: ML Lengths 1101 of 1527 splits\n",
      "     31.56 seconds: ML Lengths 1201 of 1527 splits\n",
      "     31.71 seconds: ML Lengths 1301 of 1527 splits\n",
      "     31.87 seconds: ML Lengths 1401 of 1527 splits\n",
      "     32.04 seconds: ML Lengths 1501 of 1527 splits\n",
      "     32.62 seconds: ML NNI round 1 of 21, 101 of 1527 splits, 13 changes (max delta 11.412)\n",
      "     33.11 seconds: ML NNI round 1 of 21, 201 of 1527 splits, 48 changes (max delta 38.055)\n",
      "     33.64 seconds: ML NNI round 1 of 21, 301 of 1527 splits, 84 changes (max delta 74.275)\n",
      "     34.16 seconds: ML NNI round 1 of 21, 401 of 1527 splits, 106 changes (max delta 74.275)\n",
      "     34.69 seconds: ML NNI round 1 of 21, 501 of 1527 splits, 131 changes (max delta 74.275)\n",
      "     35.17 seconds: ML NNI round 1 of 21, 601 of 1527 splits, 151 changes (max delta 74.275)\n",
      "     35.73 seconds: ML NNI round 1 of 21, 701 of 1527 splits, 185 changes (max delta 74.275)\n",
      "     36.25 seconds: ML NNI round 1 of 21, 801 of 1527 splits, 229 changes (max delta 74.275)\n",
      "     36.79 seconds: ML NNI round 1 of 21, 901 of 1527 splits, 264 changes (max delta 74.275)\n",
      "     37.29 seconds: ML NNI round 1 of 21, 1001 of 1527 splits, 295 changes (max delta 74.275)\n",
      "     37.80 seconds: ML NNI round 1 of 21, 1101 of 1527 splits, 318 changes (max delta 74.275)\n",
      "     38.30 seconds: ML NNI round 1 of 21, 1201 of 1527 splits, 339 changes (max delta 74.275)\n",
      "     38.78 seconds: ML NNI round 1 of 21, 1301 of 1527 splits, 375 changes (max delta 74.275)\n",
      "     39.31 seconds: ML NNI round 1 of 21, 1401 of 1527 splits, 392 changes (max delta 74.275)\n",
      "     39.85 seconds: ML NNI round 1 of 21, 1501 of 1527 splits, 407 changes (max delta 74.275)\n",
      "ML-NNI round 1: LogLk = -964945.516 NNIs 414 max delta 74.27 Time 40.06\n",
      "     40.06 seconds: Optimizing GTR model, step 1 of 12\n",
      "     41.41 seconds: Optimizing GTR model, step 2 of 12\n",
      "     44.05 seconds: Optimizing GTR model, step 4 of 12\n",
      "     45.60 seconds: Optimizing GTR model, step 5 of 12\n",
      "     47.18 seconds: Optimizing GTR model, step 6 of 12\n",
      "     48.98 seconds: Optimizing GTR model, step 7 of 12\n",
      "     49.93 seconds: Optimizing GTR model, step 8 of 12\n",
      "     52.08 seconds: Optimizing GTR model, step 10 of 12\n",
      "     54.33 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2797 0.2408 0.2599 0.2196\n",
      "GTR rates(ac ag at cg ct gt) 1.8719 2.6568 1.7433 1.7273 5.5579 1.0000\n",
      "     55.51 seconds: ML Lengths 1 of 1527 splits\n",
      "     55.72 seconds: ML Lengths 101 of 1527 splits\n",
      "     55.91 seconds: ML Lengths 201 of 1527 splits\n",
      "     56.09 seconds: ML Lengths 301 of 1527 splits\n",
      "     56.24 seconds: ML Lengths 401 of 1527 splits\n",
      "     56.41 seconds: ML Lengths 501 of 1527 splits\n",
      "     56.57 seconds: ML Lengths 601 of 1527 splits\n",
      "     56.76 seconds: ML Lengths 701 of 1527 splits\n",
      "     56.91 seconds: ML Lengths 801 of 1527 splits\n",
      "     57.08 seconds: ML Lengths 901 of 1527 splits\n",
      "     57.26 seconds: ML Lengths 1001 of 1527 splits\n",
      "     57.42 seconds: ML Lengths 1101 of 1527 splits\n",
      "     57.60 seconds: ML Lengths 1201 of 1527 splits\n",
      "     57.77 seconds: ML Lengths 1301 of 1527 splits\n",
      "     57.94 seconds: ML Lengths 1401 of 1527 splits\n",
      "     58.12 seconds: ML Lengths 1501 of 1527 splits\n",
      "     58.30 seconds: Site likelihoods with rate category 1 of 20\n",
      "     58.43 seconds: Site likelihoods with rate category 2 of 20\n",
      "     58.55 seconds: Site likelihoods with rate category 3 of 20\n",
      "     58.68 seconds: Site likelihoods with rate category 4 of 20\n",
      "     58.80 seconds: Site likelihoods with rate category 5 of 20\n",
      "     58.92 seconds: Site likelihoods with rate category 6 of 20\n",
      "     59.05 seconds: Site likelihoods with rate category 7 of 20\n",
      "     59.17 seconds: Site likelihoods with rate category 8 of 20\n",
      "     59.29 seconds: Site likelihoods with rate category 9 of 20\n",
      "     59.41 seconds: Site likelihoods with rate category 10 of 20\n",
      "     59.54 seconds: Site likelihoods with rate category 11 of 20\n",
      "     59.66 seconds: Site likelihoods with rate category 12 of 20\n",
      "     59.78 seconds: Site likelihoods with rate category 13 of 20\n",
      "     59.90 seconds: Site likelihoods with rate category 14 of 20\n",
      "     60.02 seconds: Site likelihoods with rate category 15 of 20\n",
      "     60.14 seconds: Site likelihoods with rate category 16 of 20\n",
      "     60.27 seconds: Site likelihoods with rate category 17 of 20\n",
      "     60.39 seconds: Site likelihoods with rate category 18 of 20\n",
      "     60.52 seconds: Site likelihoods with rate category 19 of 20\n",
      "     60.65 seconds: Site likelihoods with rate category 20 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 1.143 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "Use -gamma for approximate but comparable Gamma(20) log-likelihoods\n",
      "     60.84 seconds: ML NNI round 2 of 21, 1 of 1527 splits\n",
      "     61.18 seconds: ML NNI round 2 of 21, 101 of 1527 splits, 11 changes (max delta 10.005)\n",
      "     61.57 seconds: ML NNI round 2 of 21, 201 of 1527 splits, 20 changes (max delta 24.550)\n",
      "     61.99 seconds: ML NNI round 2 of 21, 301 of 1527 splits, 39 changes (max delta 55.164)\n",
      "     62.61 seconds: ML NNI round 2 of 21, 401 of 1527 splits, 81 changes (max delta 55.164)\n",
      "     63.06 seconds: ML NNI round 2 of 21, 501 of 1527 splits, 102 changes (max delta 55.164)\n",
      "     63.61 seconds: ML NNI round 2 of 21, 601 of 1527 splits, 130 changes (max delta 55.164)\n",
      "     64.09 seconds: ML NNI round 2 of 21, 701 of 1527 splits, 145 changes (max delta 55.164)\n",
      "     64.66 seconds: ML NNI round 2 of 21, 801 of 1527 splits, 178 changes (max delta 55.164)\n",
      "     65.13 seconds: ML NNI round 2 of 21, 901 of 1527 splits, 201 changes (max delta 55.164)\n",
      "     65.52 seconds: ML NNI round 2 of 21, 1001 of 1527 splits, 212 changes (max delta 55.164)\n",
      "     66.15 seconds: ML NNI round 2 of 21, 1101 of 1527 splits, 251 changes (max delta 55.164)\n",
      "     66.64 seconds: ML NNI round 2 of 21, 1201 of 1527 splits, 283 changes (max delta 55.164)\n",
      "     67.04 seconds: ML NNI round 2 of 21, 1301 of 1527 splits, 300 changes (max delta 55.164)\n",
      "     67.62 seconds: ML NNI round 2 of 21, 1401 of 1527 splits, 335 changes (max delta 55.164)\n",
      "     68.09 seconds: ML NNI round 2 of 21, 1501 of 1527 splits, 350 changes (max delta 55.164)\n",
      "ML-NNI round 2: LogLk = -790966.537 NNIs 354 max delta 55.16 Time 68.20\n",
      "     68.20 seconds: ML NNI round 3 of 21, 1 of 1527 splits\n",
      "     68.70 seconds: ML NNI round 3 of 21, 101 of 1527 splits, 17 changes (max delta 2.033)\n",
      "     69.18 seconds: ML NNI round 3 of 21, 201 of 1527 splits, 40 changes (max delta 3.868)\n",
      "     69.71 seconds: ML NNI round 3 of 21, 301 of 1527 splits, 65 changes (max delta 6.541)\n",
      "     70.19 seconds: ML NNI round 3 of 21, 401 of 1527 splits, 79 changes (max delta 67.568)\n",
      "     70.73 seconds: ML NNI round 3 of 21, 501 of 1527 splits, 104 changes (max delta 67.568)\n",
      "     71.17 seconds: ML NNI round 3 of 21, 601 of 1527 splits, 116 changes (max delta 67.568)\n",
      "     71.78 seconds: ML NNI round 3 of 21, 701 of 1527 splits, 150 changes (max delta 67.568)\n",
      "     72.27 seconds: ML NNI round 3 of 21, 801 of 1527 splits, 179 changes (max delta 67.568)\n",
      "     72.74 seconds: ML NNI round 3 of 21, 901 of 1527 splits, 210 changes (max delta 67.568)\n",
      "     73.09 seconds: ML NNI round 3 of 21, 1001 of 1527 splits, 218 changes (max delta 67.568)\n",
      "     73.39 seconds: ML NNI round 3 of 21, 1101 of 1527 splits, 222 changes (max delta 67.568)\n",
      "ML-NNI round 3: LogLk = -790403.666 NNIs 223 max delta 67.57 Time 73.70\n",
      "     73.70 seconds: ML NNI round 4 of 21, 1 of 1527 splits\n",
      "     73.86 seconds: ML NNI round 4 of 21, 101 of 1527 splits, 1 changes (max delta 3.377)\n",
      "     74.36 seconds: ML NNI round 4 of 21, 201 of 1527 splits, 23 changes (max delta 50.905)\n",
      "     74.94 seconds: ML NNI round 4 of 21, 301 of 1527 splits, 51 changes (max delta 50.905)\n",
      "     75.38 seconds: ML NNI round 4 of 21, 401 of 1527 splits, 68 changes (max delta 50.905)\n",
      "     75.89 seconds: ML NNI round 4 of 21, 501 of 1527 splits, 87 changes (max delta 50.905)\n",
      "     76.26 seconds: ML NNI round 4 of 21, 601 of 1527 splits, 104 changes (max delta 50.905)\n",
      "     76.62 seconds: ML NNI round 4 of 21, 701 of 1527 splits, 133 changes (max delta 50.905)\n",
      "     76.91 seconds: ML NNI round 4 of 21, 801 of 1527 splits, 146 changes (max delta 50.905)\n",
      "ML-NNI round 4: LogLk = -790098.861 NNIs 154 max delta 50.90 Time 77.19\n",
      "     77.18 seconds: ML NNI round 5 of 21, 1 of 1527 splits\n",
      "     77.58 seconds: ML NNI round 5 of 21, 101 of 1527 splits, 14 changes (max delta 0.000)\n",
      "     78.09 seconds: ML NNI round 5 of 21, 201 of 1527 splits, 32 changes (max delta 24.697)\n",
      "     78.56 seconds: ML NNI round 5 of 21, 301 of 1527 splits, 53 changes (max delta 24.697)\n",
      "     78.86 seconds: ML NNI round 5 of 21, 401 of 1527 splits, 64 changes (max delta 24.697)\n",
      "ML-NNI round 5: LogLk = -790018.625 NNIs 66 max delta 24.70 Time 78.92\n",
      "     79.30 seconds: ML NNI round 6 of 21, 101 of 1527 splits, 16 changes (max delta 5.599)\n",
      "     79.54 seconds: ML NNI round 6 of 21, 201 of 1527 splits, 20 changes (max delta 5.599)\n",
      "ML-NNI round 6: LogLk = -789984.442 NNIs 21 max delta 6.03 Time 79.62\n",
      "     79.96 seconds: ML NNI round 7 of 21, 101 of 1527 splits, 12 changes (max delta 23.462)\n",
      "ML-NNI round 7: LogLk = -789939.288 NNIs 12 max delta 23.46 Time 80.08\n",
      "     80.08 seconds: ML NNI round 8 of 21, 1 of 1527 splits\n",
      "ML-NNI round 8: LogLk = -789905.985 NNIs 11 max delta 17.52 Time 80.33\n",
      "     80.33 seconds: ML NNI round 9 of 21, 1 of 1527 splits\n",
      "ML-NNI round 9: LogLk = -789903.460 NNIs 0 max delta 0.00 Time 80.45\n",
      "Turning off heuristics for final round of ML NNIs (converged)\n",
      "     80.44 seconds: ML NNI round 10 of 21, 1 of 1527 splits\n",
      "     81.27 seconds: ML NNI round 10 of 21, 101 of 1527 splits, 4 changes (max delta 5.016)\n",
      "     82.02 seconds: ML NNI round 10 of 21, 201 of 1527 splits, 13 changes (max delta 11.803)\n",
      "     82.74 seconds: ML NNI round 10 of 21, 301 of 1527 splits, 24 changes (max delta 11.803)\n",
      "     83.36 seconds: ML NNI round 10 of 21, 401 of 1527 splits, 41 changes (max delta 11.803)\n",
      "     84.11 seconds: ML NNI round 10 of 21, 501 of 1527 splits, 47 changes (max delta 11.803)\n",
      "     84.82 seconds: ML NNI round 10 of 21, 601 of 1527 splits, 73 changes (max delta 11.803)\n",
      "     85.52 seconds: ML NNI round 10 of 21, 701 of 1527 splits, 86 changes (max delta 11.803)\n",
      "     86.32 seconds: ML NNI round 10 of 21, 801 of 1527 splits, 109 changes (max delta 11.803)\n",
      "     87.07 seconds: ML NNI round 10 of 21, 901 of 1527 splits, 121 changes (max delta 11.803)\n",
      "     87.76 seconds: ML NNI round 10 of 21, 1001 of 1527 splits, 133 changes (max delta 11.803)\n",
      "     88.42 seconds: ML NNI round 10 of 21, 1101 of 1527 splits, 140 changes (max delta 11.803)\n",
      "     89.19 seconds: ML NNI round 10 of 21, 1201 of 1527 splits, 153 changes (max delta 11.803)\n",
      "     89.96 seconds: ML NNI round 10 of 21, 1301 of 1527 splits, 169 changes (max delta 11.803)\n",
      "     90.65 seconds: ML NNI round 10 of 21, 1401 of 1527 splits, 188 changes (max delta 11.803)\n",
      "     91.40 seconds: ML NNI round 10 of 21, 1501 of 1527 splits, 190 changes (max delta 11.803)\n",
      "ML-NNI round 10: LogLk = -788902.432 NNIs 190 max delta 11.80 Time 91.64 (final)\n",
      "     91.64 seconds: ML Lengths 1 of 1527 splits\n",
      "     91.84 seconds: ML Lengths 101 of 1527 splits\n",
      "     92.03 seconds: ML Lengths 201 of 1527 splits\n",
      "     92.22 seconds: ML Lengths 301 of 1527 splits\n",
      "     92.37 seconds: ML Lengths 401 of 1527 splits\n",
      "     92.55 seconds: ML Lengths 501 of 1527 splits\n",
      "     92.71 seconds: ML Lengths 601 of 1527 splits\n",
      "     92.88 seconds: ML Lengths 701 of 1527 splits\n",
      "     93.05 seconds: ML Lengths 801 of 1527 splits\n",
      "     93.23 seconds: ML Lengths 901 of 1527 splits\n",
      "     93.39 seconds: ML Lengths 1001 of 1527 splits\n",
      "     93.56 seconds: ML Lengths 1101 of 1527 splits\n",
      "     93.75 seconds: ML Lengths 1201 of 1527 splits\n",
      "     93.91 seconds: ML Lengths 1301 of 1527 splits\n",
      "     94.08 seconds: ML Lengths 1401 of 1527 splits\n",
      "     94.27 seconds: ML Lengths 1501 of 1527 splits\n",
      "Optimize all lengths: LogLk = -788880.884 Time 94.36\n",
      "     95.26 seconds: ML split tests for    100 of   1526 internal splits\n",
      "     96.17 seconds: ML split tests for    200 of   1526 internal splits\n",
      "     97.07 seconds: ML split tests for    300 of   1526 internal splits\n",
      "     97.94 seconds: ML split tests for    400 of   1526 internal splits\n",
      "     98.80 seconds: ML split tests for    500 of   1526 internal splits\n",
      "     99.67 seconds: ML split tests for    600 of   1526 internal splits\n",
      "    100.55 seconds: ML split tests for    700 of   1526 internal splits\n",
      "    101.44 seconds: ML split tests for    800 of   1526 internal splits\n",
      "    102.32 seconds: ML split tests for    900 of   1526 internal splits\n",
      "    103.23 seconds: ML split tests for   1000 of   1526 internal splits\n",
      "    104.09 seconds: ML split tests for   1100 of   1526 internal splits\n",
      "    104.95 seconds: ML split tests for   1200 of   1526 internal splits\n",
      "    105.94 seconds: ML split tests for   1300 of   1526 internal splits\n",
      "    106.85 seconds: ML split tests for   1400 of   1526 internal splits\n",
      "    107.73 seconds: ML split tests for   1500 of   1526 internal splits\n",
      "Total time: 107.98 seconds Unique: 1529/1607 Bad splits: 9/1526 Worst delta-LogLk 26.994\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "fasttree -nt -gtr < results/gyrb/processing/moeller_sup/Bacteroidaceae_ref.fna.aln.trim > results/gyrb/processing/moeller_sup/Bacteroidaceae_ref.fna.aln.trim.tre\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#view tree in figtree,  \n",
    "#root based on outgroups \n",
    "#select clades with topology matching Figure 1 in Moeller et al. \n",
    "#save subtrees separately\n",
    "#same sequence names separately, annotate with clade designations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_tree = 'results/gyrb/processing/moeller_sup/Bacteroidaceae_ref.fna.aln.trim.tre'   \n",
    "rooted_tree = 'results/gyrb/processing/moeller_sup/Bacteroidaceae_ref.fna.aln.trim.rooted.tre'  \n",
    "Bt1_tree_file = 'results/gyrb/processing/moeller_sup/moeller_codiv_lin_Bt1.tree'  \n",
    "Bt2_tree_file = 'results/gyrb/processing/moeller_sup/moeller_codiv_lin_Bt2.tree'  \n",
    "Bt3_tree_file = 'results/gyrb/processing/moeller_sup/moeller_codiv_lin_Bt3.tree'  \n",
    "\n",
    "tree = Tree(in_tree, format=0)\n",
    "outgroup_taxa=[leaf.name for leaf in tree.get_leaves() if \n",
    "               'p__Gemmatimonadota' in leaf.name or \n",
    "               'o__Chlorobiales' in leaf.name or \n",
    "               'o__SJA-28' in leaf.name]\n",
    "outgroup_MRCA = tree.get_common_ancestor(outgroup_taxa)\n",
    "tree.set_outgroup(outgroup_MRCA)\n",
    "tree.write(outfile=rooted_tree,format=2)\n",
    "\n",
    "def get_lineage_ASVs(in_tree,listTaxa):    \n",
    "    #makes sure we don't miss any ASVs that didn't match in the blast search \\n\"\n",
    "    #but are descended from the MRCA of those that did\\n\"\n",
    "    tree = Tree(in_tree, format=0)\n",
    "    lineage_MRCA = tree.get_common_ancestor(listTaxa)\n",
    "    linage_taxa = [x.name for x in lineage_MRCA.get_leaves()] \n",
    "    tree.prune(linage_taxa)\n",
    "    return(tree)\n",
    "\n",
    "def rename_taxa(lin_tree,listTaxa,cladeName):\n",
    "    lineage_MRCA = lin_tree.get_common_ancestor(listTaxa)\n",
    "    for taxa in lineage_MRCA.get_leaves():    \n",
    "        taxa.name = taxa.name + ' ' + cladeName\n",
    "\n",
    "Bt1_tree = get_lineage_ASVs(rooted_tree,['Gorilla5764570','Bonobo8808876'])\n",
    "Bt1_taxa_oldnames = [leaf.name for leaf in Bt1_tree.get_leaves()]        \n",
    "rename_taxa(Bt1_tree,['Gorilla5764570','Gorilla10693345'],'Bt1_clade1_gorilla')\n",
    "rename_taxa(Bt1_tree,['Chimp6685161','Chimp5888147'],'Bt1_clade1_chimp')\n",
    "rename_taxa(Bt1_tree,['Bonobo11821974','Bonobo8808876'],'Bt1_clade1_bonobo')\n",
    "Bt1_taxa = [leaf.name for leaf in Bt1_tree.get_leaves()] \n",
    "Bt1_tree.write(outfile=Bt1_tree_file,format=2) \n",
    "\n",
    "Bt2_tree = get_lineage_ASVs(rooted_tree,['GCF_002933775.1p__Bacteroidotao__Bacteroidalesf__Bacteroidaceaes__Prevotella_sp002933775',\n",
    "                                         'Bonobo5991239'])\n",
    "Bt2_taxa_oldnames = [leaf.name for leaf in Bt2_tree.get_leaves()]        \n",
    "rename_taxa(Bt2_tree,['Gorilla7248598','Gorilla1454159'],'Bt2_clade1_gorilla')\n",
    "rename_taxa(Bt2_tree,['Chimp8269086','Chimp3094078'],'Bt2_clade1_chimp')\n",
    "rename_taxa(Bt2_tree,['Bonobo11725448','Bonobo5991239'],'Bt2_clade1_bonobo')\n",
    "rename_taxa(Bt2_tree,['Chimp1322981','Chimp3450568'],'Bt2_clade2_chimp')\n",
    "rename_taxa(Bt2_tree,['Bonobo11213703','Bonobo6937590'],'Bt2_clade2_bonobo')\n",
    "Bt2_taxa = [leaf.name for leaf in Bt2_tree.get_leaves()] \n",
    "Bt2_tree.write(outfile=Bt2_tree_file,format=2) \n",
    "\n",
    "Bt3_tree = get_lineage_ASVs(rooted_tree,['Human1330495','Bonobo4232320'])\n",
    "Bt3_taxa_oldnames = [leaf.name for leaf in Bt3_tree.get_leaves()]        \n",
    "rename_taxa(Bt3_tree,['Human1330495','Human12108738'],'Bt3_clade1_human')\n",
    "rename_taxa(Bt3_tree,['Chimp2629519','Chimp873638'],'Bt3_clade1_chimp')\n",
    "rename_taxa(Bt3_tree,['Bonobo6212174','Bonobo4232320'],'Bt3_clade1_bonobo')\n",
    "Bt3_taxa = [leaf.name for leaf in Bt3_tree.get_leaves()] \n",
    "Bt3_tree.write(outfile=Bt3_tree_file,format=2) \n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>codiv_taxa</th>\n",
       "      <th>codiv_clade</th>\n",
       "      <th>codiv_lineage</th>\n",
       "      <th>hostSpecies</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Gorilla5764570 Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1</td>\n",
       "      <td>wild_gorilla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Gorilla3748318 Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1</td>\n",
       "      <td>wild_gorilla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Gorilla12208365 Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1</td>\n",
       "      <td>wild_gorilla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Gorilla4447500 Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1</td>\n",
       "      <td>wild_gorilla</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Gorilla8497213 Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1_clade1_gorilla</td>\n",
       "      <td>Bt1</td>\n",
       "      <td>wild_gorilla</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           codiv_taxa         codiv_clade codiv_lineage  \\\n",
       "0   Gorilla5764570 Bt1_clade1_gorilla  Bt1_clade1_gorilla           Bt1   \n",
       "1   Gorilla3748318 Bt1_clade1_gorilla  Bt1_clade1_gorilla           Bt1   \n",
       "2  Gorilla12208365 Bt1_clade1_gorilla  Bt1_clade1_gorilla           Bt1   \n",
       "3   Gorilla4447500 Bt1_clade1_gorilla  Bt1_clade1_gorilla           Bt1   \n",
       "4   Gorilla8497213 Bt1_clade1_gorilla  Bt1_clade1_gorilla           Bt1   \n",
       "\n",
       "    hostSpecies  \n",
       "0  wild_gorilla  \n",
       "1  wild_gorilla  \n",
       "2  wild_gorilla  \n",
       "3  wild_gorilla  \n",
       "4  wild_gorilla  "
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "codiv_taxa_oldnames = Bt1_taxa_oldnames + Bt2_taxa_oldnames + Bt3_taxa_oldnames\n",
    "codiv_taxa = Bt1_taxa + Bt2_taxa + Bt3_taxa\n",
    "codivDict = dict(zip(codiv_taxa_oldnames, codiv_taxa))\n",
    "codivDF = pd.DataFrame(list(zip(codiv_taxa,codiv_taxa_oldnames)),columns=['codiv_taxa','codiv_taxa_oldnames'])\n",
    "codivDF['moeller_seq'] = codivDF['codiv_taxa'].apply(lambda x: 'moeller' if 'p__Bacteroidota' not in x else 'ref')\n",
    "codivDF = codivDF[codivDF['moeller_seq'] == 'moeller']\n",
    "codivDF['codiv_clade'] = codivDF['codiv_taxa'].apply(lambda x: x.split(' ')[1])\n",
    "codivDF['codiv_lineage'] = codivDF['codiv_clade'].apply(lambda x: x.split('_')[0])\n",
    "codivDF['hostSpecies'] = codivDF['codiv_clade'].apply(lambda x: 'wild_'+x.split('_')[-1])\n",
    "codivDF = codivDF[['codiv_taxa','codiv_clade','codiv_lineage','hostSpecies']]\n",
    "codivDF.to_csv(f'{outdir}/moeller_codiv_HRclades.txt',index=False,sep='\\t')\n",
    "codivDF.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gorilla10693345\n",
      "Gorilla10893151\n",
      "Gorilla12208365\n",
      "Gorilla1454159\n",
      "Gorilla3280123\n",
      "Gorilla3748318\n",
      "Gorilla4351822\n",
      "Gorilla4447500\n",
      "Gorilla5764570\n",
      "Gorilla6306341\n",
      "Gorilla7248598\n",
      "Gorilla7611566\n",
      "Gorilla7655534\n",
      "Gorilla8443136\n",
      "Gorilla8497213\n",
      "Gorilla8907336\n",
      "Gorilla9193943\n",
      "Gorilla9805772\n",
      "Human10045430\n",
      "Human10283488\n",
      "Human10889793\n",
      "Human10912985\n",
      "Human10912992\n",
      "Human12108738\n",
      "Human1330489\n",
      "Human1330495\n",
      "Human1781917\n",
      "Human237218\n",
      "Human2572122\n",
      "Human2935124\n",
      "Human3034958\n",
      "Human3367149\n",
      "Human3418197\n",
      "Human3961325\n",
      "Human4630891\n",
      "Human4639158\n",
      "Human4639166\n",
      "Human5110045\n",
      "Human5110051\n",
      "Human5465528\n",
      "Human5838232\n",
      "Human6390749\n",
      "Human6405675\n",
      "Human663970\n",
      "Human663977\n",
      "Human6649477\n",
      "Human6700665\n",
      "Human6700667\n",
      "Human7250453\n",
      "Human7563063\n",
      "Human7563070\n",
      "Human7859459\n",
      "Human8146958\n",
      "Human8295127\n",
      "Human8829943\n",
      "Human8842535\n",
      "Human9177678\n",
      "Human9518949\n",
      "Human9825266\n",
      "Chimp10218980\n",
      "Chimp1067447\n",
      "Chimp11049685\n",
      "Chimp11943611\n",
      "Chimp1322981\n",
      "Chimp1324508\n",
      "Chimp1493190\n",
      "Chimp1528180\n",
      "Chimp168543\n",
      "Chimp1862511\n",
      "Chimp2164238\n",
      "Chimp2506335\n",
      "Chimp2629519\n",
      "Chimp2757936\n",
      "Chimp2936652\n",
      "Chimp304514\n",
      "Chimp3094078\n",
      "Chimp3094087\n",
      "Chimp3162311\n",
      "Chimp3450568\n",
      "Chimp3450573\n",
      "Chimp3781815\n",
      "Chimp3970093\n",
      "Chimp4678997\n",
      "Chimp5183434\n",
      "Chimp5211526\n",
      "Chimp5888147\n",
      "Chimp665501\n",
      "Chimp6685161\n",
      "Chimp6973326\n",
      "Chimp6973327\n",
      "Chimp7736944\n",
      "Chimp7736952\n",
      "Chimp8269086\n",
      "Chimp828594\n",
      "Chimp873638\n",
      "Chimp8740623\n",
      "Chimp8780480\n",
      "Chimp9057405\n",
      "Chimp9057407\n",
      "Chimp9228565\n",
      "Chimp934253\n",
      "Chimp985651\n",
      "Chimp9859517\n",
      "Chimp9931019\n",
      "Chimp9931021\n",
      "Bonobo10034489\n",
      "Bonobo10057151\n",
      "Bonobo10352189\n",
      "Bonobo10513316\n",
      "Bonobo10520847\n",
      "Bonobo10609639\n",
      "Bonobo10623458\n",
      "Bonobo1078\n",
      "Bonobo11021063\n",
      "Bonobo11105611\n",
      "Bonobo11213703\n",
      "Bonobo11242356\n",
      "Bonobo11367003\n",
      "Bonobo11646549\n",
      "Bonobo11725448\n",
      "Bonobo11725461\n",
      "Bonobo11768907\n",
      "Bonobo11821974\n",
      "Bonobo12102095\n",
      "Bonobo12157978\n",
      "Bonobo12387329\n",
      "Bonobo1306358\n",
      "Bonobo1624513\n",
      "Bonobo1787673\n",
      "Bonobo1873599\n",
      "Bonobo1946453\n",
      "Bonobo2078905\n",
      "Bonobo2166737\n",
      "Bonobo2193524\n",
      "Bonobo229236\n",
      "Bonobo229241\n",
      "Bonobo23228\n",
      "Bonobo2366405\n",
      "Bonobo2450173\n",
      "Bonobo2645793\n",
      "Bonobo2919016\n",
      "Bonobo3176696\n",
      "Bonobo3311273\n",
      "Bonobo3371043\n",
      "Bonobo3681147\n",
      "Bonobo3793226\n",
      "Bonobo3971783\n",
      "Bonobo3994923\n",
      "Bonobo4076472\n",
      "Bonobo4186256\n",
      "Bonobo4208740\n",
      "Bonobo4208742\n",
      "Bonobo4214590\n",
      "Bonobo4214599\n",
      "Bonobo4232320\n",
      "Bonobo4255176\n",
      "Bonobo4255178\n",
      "Bonobo4400446\n",
      "Bonobo4469078\n",
      "Bonobo447600\n",
      "Bonobo4568548\n",
      "Bonobo4632650\n",
      "Bonobo4632659\n",
      "Bonobo4668782\n",
      "Bonobo4886707\n",
      "Bonobo4935167\n",
      "Bonobo4950863\n",
      "Bonobo5054462\n",
      "Bonobo5348599\n",
      "Bonobo5507217\n",
      "Bonobo5555405\n",
      "Bonobo5596313\n",
      "Bonobo5703402\n",
      "Bonobo5731561\n",
      "Bonobo5805938\n",
      "Bonobo5856903\n",
      "Bonobo5991239\n",
      "Bonobo6156032\n",
      "Bonobo6212174\n",
      "Bonobo6331470\n",
      "Bonobo6396279\n",
      "Bonobo6757245\n",
      "Bonobo6757253\n",
      "Bonobo6828899\n",
      "Bonobo6932003\n",
      "Bonobo6937590\n",
      "Bonobo7084162\n",
      "Bonobo7228450\n",
      "Bonobo742599\n",
      "Bonobo759069\n",
      "Bonobo7657683\n",
      "Bonobo781247\n",
      "Bonobo7895675\n",
      "Bonobo7932810\n",
      "Bonobo8148512\n",
      "Bonobo8213484\n",
      "Bonobo8808876\n",
      "Bonobo8808878\n",
      "Bonobo8821370\n",
      "Bonobo8827876\n",
      "Bonobo8964982\n",
      "Bonobo9176466\n",
      "Bonobo9375067\n",
      "Bonobo9719\n",
      "Bonobo972742\n",
      "Bonobo9728\n",
      "Bonobo9832770\n",
      "Bonobo9846630\n"
     ]
    }
   ],
   "source": [
    "with open(f'{outdir}/Bacteroidaceae.fna') as original: \n",
    "    with open(f'{outdir}/moeller_codiv_Bacteroidaceae.fna', 'w') as covid:\n",
    "        records = SeqIO.parse(original, 'fasta')\n",
    "        for record in records:\n",
    "            if record.id in codiv_taxa_oldnames:\n",
    "                print(record.id)\n",
    "                record.id = codivDict[record.id]\n",
    "                record.description = ''\n",
    "                SeqIO.write(record, covid, 'fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
