{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#This downloads metagenome assemblies from Segata lab data repository and filters out gyrB sequences \n",
    "#using a blast db created from ASVs from this study and GTDB-Tk reference db"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>https</th>\n",
       "      <th>filename</th>\n",
       "      <th>folder</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>https://www.dropbox.com/s/njbwzpazz4a1dy1/Asni...</td>\n",
       "      <td>AsnicarF_2017.tar.bz2</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>https://www.dropbox.com/s/n6867hhmo12xnwf/Back...</td>\n",
       "      <td>BackhedF_2015.tar.bz2</td>\n",
       "      <td>BackhedF_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>https://www.dropbox.com/s/ac6cpq50pt5mqj6/Beng...</td>\n",
       "      <td>Bengtsson-PalmeJ_2015.tar.bz2</td>\n",
       "      <td>Bengtsson-PalmeJ_2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>https://www.dropbox.com/s/uj9cy9914l0cq3x/Brit...</td>\n",
       "      <td>BritoIL_2016.tar.bz2</td>\n",
       "      <td>BritoIL_2016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>https://www.dropbox.com/s/dtxmz1apdgtf3d1/Cast...</td>\n",
       "      <td>Castro-NallarE_2015.tar.bz2</td>\n",
       "      <td>Castro-NallarE_2015</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                               https  \\\n",
       "0  https://www.dropbox.com/s/njbwzpazz4a1dy1/Asni...   \n",
       "1  https://www.dropbox.com/s/n6867hhmo12xnwf/Back...   \n",
       "2  https://www.dropbox.com/s/ac6cpq50pt5mqj6/Beng...   \n",
       "3  https://www.dropbox.com/s/uj9cy9914l0cq3x/Brit...   \n",
       "4  https://www.dropbox.com/s/dtxmz1apdgtf3d1/Cast...   \n",
       "\n",
       "                        filename                 folder  \n",
       "0          AsnicarF_2017.tar.bz2          AsnicarF_2017  \n",
       "1          BackhedF_2015.tar.bz2          BackhedF_2015  \n",
       "2  Bengtsson-PalmeJ_2015.tar.bz2  Bengtsson-PalmeJ_2015  \n",
       "3           BritoIL_2016.tar.bz2           BritoIL_2016  \n",
       "4    Castro-NallarE_2015.tar.bz2    Castro-NallarE_2015  "
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metagenomic_datasets = pd.read_csv('metadata/metagenome_http_to_dnload.txt',header=None)\n",
    "metagenomic_datasets.columns = ['https']\n",
    "metagenomic_datasets['filename'] = metagenomic_datasets['https'].apply(lambda x: x.split('/')[-1])\n",
    "metagenomic_datasets['folder'] = metagenomic_datasets['filename'].apply(lambda x: x.split('.')[0])\n",
    "metagenomic_datasets.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create blastdb to filter metagenomic assembly reads\n",
    "#contains Bacteroidales seqs from gtdbtk taxonomy giving wide taxonomic breadth and\n",
    "#also Bt gyrB ASVs from present study. It is necessary to include wild ape seqs because\n",
    "#they are quite divergent.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: results/gyrb_metagenomic: File exists\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir results/gyrb_metagenomic\n",
    "cat results/gyrb_bt_gtdbtk_ref/blast_db/gtdbtk_gyrb_amplicon.fasta \\\n",
    "results/gyrb_all/Bt/ASVs_filtered.fasta > results/gyrb_metagenomic/Bt_ASV_gtdbtk.fasta\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AsnicarF_2017__MV_FEI1_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEI2_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEI3_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEI4_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEI4_t2Q15.fa\n",
      "AsnicarF_2017__MV_FEI5_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEI5_t2Q14.fa\n",
      "AsnicarF_2017__MV_FEI5_t3Q15.fa\n",
      "AsnicarF_2017__MV_FEM1_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEM2_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEM3_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEM4_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEM4_t2Q15.fa\n",
      "AsnicarF_2017__MV_FEM5_t1Q14.fa\n",
      "AsnicarF_2017__MV_FEM5_t2Q14.fa\n",
      "AsnicarF_2017__MV_FEM5_t3Q15.fa\n",
      "AsnicarF_2017__MV_MIM2_t1M14.fa\n",
      "AsnicarF_2017__MV_MIM3_t1M14.fa\n",
      "AsnicarF_2017__MV_MIM4_t2F15.fa\n",
      "AsnicarF_2017__MV_MIM5_t2M14.fa\n",
      "AsnicarF_2017__MV_MIM5_t3F15.fa\n",
      "AsnicarF_2017__MV_FEI1_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEI2_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEI3_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEI4_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEI4_t2Q15.fa no hits\n",
      "AsnicarF_2017__MV_FEI5_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEI5_t2Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEI5_t3Q15.fa no hits\n",
      "AsnicarF_2017__MV_FEM1_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEM2_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEM3_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEM4_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEM4_t2Q15.fa no hits\n",
      "AsnicarF_2017__MV_FEM5_t1Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEM5_t2Q14.fa no hits\n",
      "AsnicarF_2017__MV_FEM5_t3Q15.fa no hits\n",
      "AsnicarF_2017__MV_MIM2_t1M14.fa no hits\n",
      "AsnicarF_2017__MV_MIM3_t1M14.fa no hits\n",
      "AsnicarF_2017__MV_MIM4_t2F15.fa no hits\n",
      "AsnicarF_2017__MV_MIM5_t2M14.fa no hits\n",
      "AsnicarF_2017__MV_MIM5_t3F15.fa no hits\n"
     ]
    }
   ],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/')\n",
    "folder = 'AsnicarF_2017'\n",
    "fastas = [f for f in os.listdir(f'results/gyrb_metagenomic/{folder}/') if f.endswith('.fa')]\n",
    "os.system(f'mkdir -pv results/gyrb_metagenomic/hmmsearch/{folder}')\n",
    "for fasta in fastas:\n",
    "    print(fasta)\n",
    "    \n",
    "def filter_gyrb_reads(fasta):\n",
    "    #os.system(f'prodigal -i results/gyrb_metagenomic/{folder}/{fasta} \\\n",
    "    #                     -a results/gyrb_metagenomic/{folder}/{fasta}_genes \\\n",
    "    #                     -d results/gyrb_metagenomic/{folder}/{fasta}_prot')\n",
    "    os.system(f'hmmsearch --tblout results/gyrb_metagenomic/hmmsearch/{folder}/{fasta}_hmmout.txt \\\n",
    "                ref_seqs/GTDBTK_db/markers/tigrfam/individual_hmms/TIGR01059.HMM \\\n",
    "                results/gyrb_metagenomic/{folder}/{fasta}_prot')\n",
    "    try:\n",
    "        hmm_hits = pd.read_csv(f'results/gyrb_metagenomic/hmmsearch/{folder}/{fasta}_hmmout.txt',delim_whitespace=True,header=None,comment='#')\n",
    "    except:\n",
    "        print(fasta,'no hits')\n",
    "        return\n",
    "    \n",
    "    hmm_hits = hmm_hits[0] \n",
    "    hmm_hits_eval = hmm_hits[0][hmm_hits[4]<float(1e250)]\n",
    "\n",
    "    #extract genes\n",
    "    records = SeqIO.parse(f'results/gyrb_metagenomic/{folder}/{fasta}_genes', 'fasta')\n",
    "    records_hits = [rec for rec in records if rec.id in hmm_hits]\n",
    "    SeqIO.write_fasta(records_hits,f'results/gyrb_metagenomic/hmmsearch/{folder}/{fasta}_genes','fasta')\n",
    "\n",
    "    os.system(f'blastx -query results/gyrb_metagenomic/hmmsearch/{folder}/{fasta}_genes \\\n",
    "                       -db results/gyrb_bt_gtdbtk_ref/blast_db/gtdbtk_gyrb_amplicon.faa \\\n",
    "                       -outfmt \"7 qseqid sseqid pident length qlen evalue sseq\" \\\n",
    "                       -out results/gyrb_metagenomic/blastout/{fasta}_blastout.txt')\n",
    "    blast_res = pd.read_csv(f'results/gyrb_metagenomic/blastout/{fasta}_blastout.txt',sep='\\t',comment='#')\n",
    "\n",
    "for fasta in fastas:\n",
    "    filter_gyrb_reads(fasta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "       "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "https://www.dropbox.com/s/njbwzpazz4a1dy1/AsnicarF_2017.tar.bz2 AsnicarF_2017.tar.bz2 AsnicarF_2017\n",
      "AsnicarF_2017\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI1_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI1_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:04\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI1_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEI1_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI1_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 41591 sequences in 1.36189 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI2_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI2_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:06\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI2_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEI2_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI2_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 17137 sequences in 0.573672 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI3_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI3_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:07\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI3_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEI3_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI3_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 24404 sequences in 0.849889 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI4_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI4_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:08\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI4_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEI4_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI4_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 6911 sequences in 0.335948 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI4_t2Q15.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI4_t2Q15\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:09\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI4_t2Q15.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEI4_t2Q15.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI4_t2Q15.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 50190 sequences in 1.65784 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI5_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI5_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:11\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI5_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEI5_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI5_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 36428 sequences in 1.2387 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI5_t2Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI5_t2Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:13\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI5_t2Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEI5_t2Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI5_t2Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 50532 sequences in 1.75123 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI5_t3Q15.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEI5_t3Q15\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:15\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI5_t3Q15.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEI5_t3Q15.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEI5_t3Q15.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 110453 sequences in 3.5046 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM1_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM1_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:19\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM1_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEM1_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM1_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 143664 sequences in 4.26767 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM2_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM2_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:24\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM2_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEM2_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM2_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 249869 sequences in 7.41829 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM3_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM3_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:33\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM3_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEM3_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM3_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 217567 sequences in 6.32837 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM4_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM4_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:40\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM4_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEM4_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM4_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 89431 sequences in 3.00109 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM4_t2Q15.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM4_t2Q15\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:44\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM4_t2Q15.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEM4_t2Q15.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM4_t2Q15.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 299562 sequences in 8.85689 seconds.\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AsnicarF_2017/AsnicarF_2017__MV_FEM5_t1Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM5_t1Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:08:54\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM5_t1Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEM5_t1Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM5_t1Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 260457 sequences in 7.78004 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM5_t2Q14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM5_t2Q14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:09:02\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM5_t2Q14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEM5_t2Q14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM5_t2Q14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 168190 sequences in 4.89836 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM5_t3Q15.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_FEM5_t3Q15\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:09:08\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM5_t3Q15.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_FEM5_t3Q15.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_FEM5_t3Q15.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 418522 sequences in 11.4788 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM2_t1M14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM2_t1M14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:09:21\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM2_t1M14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_MIM2_t1M14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM2_t1M14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1539 sequences in 0.0371642 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM3_t1M14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM3_t1M14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:09:21\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM3_t1M14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_MIM3_t1M14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM3_t1M14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 2081 sequences in 0.0493381 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM4_t2F15.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM4_t2F15\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:09:22\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM4_t2F15.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_MIM4_t2F15.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM4_t2F15.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 2642 sequences in 0.0654712 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM5_t2M14.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM5_t2M14\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:09:22\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM5_t2M14.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_MIM5_t2M14.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM5_t2M14.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 1005 sequences in 0.024236 seconds.\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM5_t3F15.fa\n",
      "AsnicarF_2017/AsnicarF_2017__MV_MIM5_t3F15\n",
      "\n",
      "\n",
      "Building a new DB, current time: 05/19/2020 16:09:22\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM5_t3F15.fa\n",
      "New DB title:  AsnicarF_2017/AsnicarF_2017__MV_MIM5_t3F15.fa\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb_metagenomic/AsnicarF_2017/AsnicarF_2017__MV_MIM5_t3F15.fa\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 36212 sequences in 0.888296 seconds.\n"
     ]
    }
   ],
   "source": [
    "for index, row in metagenomic_datasets.iterrows():\n",
    "    http = row['https']\n",
    "    filename = row['filename']\n",
    "    folder = row['folder']\n",
    "    print(http,filename,folder)\n",
    "    !scripts/metagenomic_samples/dnload_blast.sh $http $filename $folder\n",
    "    break\n",
    "    #if not os.path.exists(f'blastout/{folder}'):\n",
    "    #    download_metagenomic_dataset_blast_gyrb_seqs(http,filename,folder)\n",
    "    #else:\n",
    "    #    print(folder,'already downloaded')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\t#prep output folders\n",
    "\tmkdir blastout/$folder/\n",
    "\tmkdir fastaout/$folder/\n",
    "\t\n",
    "\t#loop through sample files\n",
    "\tfor file in $(cat sample_files/${folder}_files.txt)\n",
    "\tdo \t\n",
    "\t\t#blast search of sample file for gyrb seqs\n",
    "\t\techo $file\n",
    "\t\tmakeblastdb -in $file -dbtype nucl\n",
    "\t\tblastn -query Bif_rep_set_passing.fasta -db $file -outfmt"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.4"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
