{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Generates a metadata sheet with 9000 plus individual metagenomic samples. Reformats the metadata from Passoli et al to be compatible with this study and then double checks that no samples in the asv sequence table are missing from metadata."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "amplicon_metadata_file = 'metadata/metadata_Bt_samples_gyrb.txt'\n",
    "metagenomic_metadata_file = 'metadata/passoli_metadata.txt'\n",
    "metagenomic_metadata_file = 'metadata/passoli_metadata_formatted.txt'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.SampleID</th>\n",
       "      <th>BarcodeSequence</th>\n",
       "      <th>LinkerPrimerSequence</th>\n",
       "      <th>Old_SampleID</th>\n",
       "      <th>dataset</th>\n",
       "      <th>genus_sp</th>\n",
       "      <th>subspecies</th>\n",
       "      <th>common_name</th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>site</th>\n",
       "      <th>site_code</th>\n",
       "      <th>country</th>\n",
       "      <th>captivity_status</th>\n",
       "      <th>amplicon</th>\n",
       "      <th>sample</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>wd.bon.IK.KSG2876.Bt</td>\n",
       "      <td>TCACCTCATACGGCTA</td>\n",
       "      <td>CAAGCAGAAGACGGCATACGAGAT</td>\n",
       "      <td>83</td>\n",
       "      <td>gyrb_moeller_wild</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>bonobo</td>\n",
       "      <td>KSG2876</td>\n",
       "      <td>DRC.IK</td>\n",
       "      <td>IK</td>\n",
       "      <td>DRC</td>\n",
       "      <td>wild</td>\n",
       "      <td>Bt</td>\n",
       "      <td>wd.bon.IK.KSG2876</td>\n",
       "      <td>wild_bonobo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>wd.bon.IK.KSG2883.Bt</td>\n",
       "      <td>GCTTGCTTCGTTGCAA</td>\n",
       "      <td>CAAGCAGAAGACGGCATACGAGAT</td>\n",
       "      <td>97</td>\n",
       "      <td>gyrb_moeller_wild</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>bonobo</td>\n",
       "      <td>KSG2883</td>\n",
       "      <td>DRC.IK</td>\n",
       "      <td>IK</td>\n",
       "      <td>DRC</td>\n",
       "      <td>wild</td>\n",
       "      <td>Bt</td>\n",
       "      <td>wd.bon.IK.KSG2883</td>\n",
       "      <td>wild_bonobo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>wd.bon.IK.KSG2877.Bt</td>\n",
       "      <td>TCACCTCACGTTGCAA</td>\n",
       "      <td>CAAGCAGAAGACGGCATACGAGAT</td>\n",
       "      <td>98</td>\n",
       "      <td>gyrb_moeller_wild</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>bonobo</td>\n",
       "      <td>KSG2877</td>\n",
       "      <td>DRC.IK</td>\n",
       "      <td>IK</td>\n",
       "      <td>DRC</td>\n",
       "      <td>wild</td>\n",
       "      <td>Bt</td>\n",
       "      <td>wd.bon.IK.KSG2877</td>\n",
       "      <td>wild_bonobo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>wd.bon.IK.KSG2879.Bt</td>\n",
       "      <td>GCTTGCTTCTGAAGTC</td>\n",
       "      <td>CAAGCAGAAGACGGCATACGAGAT</td>\n",
       "      <td>112</td>\n",
       "      <td>gyrb_moeller_wild</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>bonobo</td>\n",
       "      <td>KSG2879</td>\n",
       "      <td>DRC.IK</td>\n",
       "      <td>IK</td>\n",
       "      <td>DRC</td>\n",
       "      <td>wild</td>\n",
       "      <td>Bt</td>\n",
       "      <td>wd.bon.IK.KSG2879</td>\n",
       "      <td>wild_bonobo</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>wd.bon.IK.KSG2882.Bt</td>\n",
       "      <td>TCACCTCACTGAAGTC</td>\n",
       "      <td>CAAGCAGAAGACGGCATACGAGAT</td>\n",
       "      <td>113</td>\n",
       "      <td>gyrb_moeller_wild</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>Pan_paniscus</td>\n",
       "      <td>bonobo</td>\n",
       "      <td>KSG2882</td>\n",
       "      <td>DRC.IK</td>\n",
       "      <td>IK</td>\n",
       "      <td>DRC</td>\n",
       "      <td>wild</td>\n",
       "      <td>Bt</td>\n",
       "      <td>wd.bon.IK.KSG2882</td>\n",
       "      <td>wild_bonobo</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             X.SampleID   BarcodeSequence      LinkerPrimerSequence  \\\n",
       "0  wd.bon.IK.KSG2876.Bt  TCACCTCATACGGCTA  CAAGCAGAAGACGGCATACGAGAT   \n",
       "1  wd.bon.IK.KSG2883.Bt  GCTTGCTTCGTTGCAA  CAAGCAGAAGACGGCATACGAGAT   \n",
       "2  wd.bon.IK.KSG2877.Bt  TCACCTCACGTTGCAA  CAAGCAGAAGACGGCATACGAGAT   \n",
       "3  wd.bon.IK.KSG2879.Bt  GCTTGCTTCTGAAGTC  CAAGCAGAAGACGGCATACGAGAT   \n",
       "4  wd.bon.IK.KSG2882.Bt  TCACCTCACTGAAGTC  CAAGCAGAAGACGGCATACGAGAT   \n",
       "\n",
       "  Old_SampleID            dataset      genus_sp    subspecies common_name  \\\n",
       "0           83  gyrb_moeller_wild  Pan_paniscus  Pan_paniscus      bonobo   \n",
       "1           97  gyrb_moeller_wild  Pan_paniscus  Pan_paniscus      bonobo   \n",
       "2           98  gyrb_moeller_wild  Pan_paniscus  Pan_paniscus      bonobo   \n",
       "3          112  gyrb_moeller_wild  Pan_paniscus  Pan_paniscus      bonobo   \n",
       "4          113  gyrb_moeller_wild  Pan_paniscus  Pan_paniscus      bonobo   \n",
       "\n",
       "  uniqueid    site site_code country captivity_status amplicon  \\\n",
       "0  KSG2876  DRC.IK        IK     DRC             wild       Bt   \n",
       "1  KSG2883  DRC.IK        IK     DRC             wild       Bt   \n",
       "2  KSG2877  DRC.IK        IK     DRC             wild       Bt   \n",
       "3  KSG2879  DRC.IK        IK     DRC             wild       Bt   \n",
       "4  KSG2882  DRC.IK        IK     DRC             wild       Bt   \n",
       "\n",
       "              sample  Description  \n",
       "0  wd.bon.IK.KSG2876  wild_bonobo  \n",
       "1  wd.bon.IK.KSG2883  wild_bonobo  \n",
       "2  wd.bon.IK.KSG2877  wild_bonobo  \n",
       "3  wd.bon.IK.KSG2879  wild_bonobo  \n",
       "4  wd.bon.IK.KSG2882  wild_bonobo  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome')\n",
    "amplicon_metadata = pd.read_csv(amplicon_metadata_file,sep='\\t')\n",
    "amplicon_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th># Reads</th>\n",
       "      <th># Bases</th>\n",
       "      <th>Median Read Length</th>\n",
       "      <th>Body Site</th>\n",
       "      <th>Body Subsite</th>\n",
       "      <th>Age (years)</th>\n",
       "      <th>Age Category</th>\n",
       "      <th>Infant Age (days)</th>\n",
       "      <th>Westernized</th>\n",
       "      <th>Country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI1_t1Q14</td>\n",
       "      <td>27553455</td>\n",
       "      <td>2748531713</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>90.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI2_t1Q14</td>\n",
       "      <td>32197545</td>\n",
       "      <td>3210606434</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>90.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI3_t1Q14</td>\n",
       "      <td>32310216</td>\n",
       "      <td>3222955818</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>90.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI4_t1Q14</td>\n",
       "      <td>14270590</td>\n",
       "      <td>1423299652</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI4_t2Q15</td>\n",
       "      <td>37677351</td>\n",
       "      <td>3735487055</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Study      Sample ID   # Reads     # Bases  Median Read Length  \\\n",
       "0  AsnicarF_2017  MV_FEI1_t1Q14  27553455  2748531713               100.0   \n",
       "1  AsnicarF_2017  MV_FEI2_t1Q14  32197545  3210606434               100.0   \n",
       "2  AsnicarF_2017  MV_FEI3_t1Q14  32310216  3222955818               100.0   \n",
       "3  AsnicarF_2017  MV_FEI4_t1Q14  14270590  1423299652               100.0   \n",
       "4  AsnicarF_2017  MV_FEI4_t2Q15  37677351  3735487055               100.0   \n",
       "\n",
       "  Body Site Body Subsite Age (years) Age Category  Infant Age (days)  \\\n",
       "0     Stool          NaN           1      Newborn               90.0   \n",
       "1     Stool          NaN           1      Newborn               90.0   \n",
       "2     Stool          NaN           1      Newborn               90.0   \n",
       "3     Stool          NaN           1      Newborn                NaN   \n",
       "4     Stool          NaN           1      Newborn                NaN   \n",
       "\n",
       "  Westernized Country  \n",
       "0         Yes     ITA  \n",
       "1         Yes     ITA  \n",
       "2         Yes     ITA  \n",
       "3         Yes     ITA  \n",
       "4         Yes     ITA  "
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metagenomic_metadata = pd.read_csv(metagenomic_metadata_file,sep='\\t')\n",
    "metagenomic_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "metagenomic_metadata['X.SampleID'] = metagenomic_metadata.apply(lambda row: row['Study'] +'__' + row['Sample ID'],axis=1)\n",
    "metagenomic_metadata['X.SampleID'] = metagenomic_metadata['X.SampleID'].replace('.','').replace('-','')\n",
    "metagenomic_metadata['BarcodeSequence'] = 'NA'\n",
    "metagenomic_metadata['LinkerPrimerSequence'] = 'NA'\n",
    "metagenomic_metadata['Old_SampleID'] = metagenomic_metadata['Sample ID']\n",
    "metagenomic_metadata['dataset'] = metagenomic_metadata['Study']\n",
    "metagenomic_metadata['genus_sp'] = 'Homo_sapiens'\n",
    "metagenomic_metadata['subspecies']  = 'Homo_sapiens'\n",
    "metagenomic_metadata['common_name'] = 'human'\n",
    "metagenomic_metadata['uniqueid'] = 'NA'\n",
    "metagenomic_metadata['site'] = metagenomic_metadata['Country']\n",
    "metagenomic_metadata['site_code'] = metagenomic_metadata['Country']\n",
    "metagenomic_metadata['country'] = metagenomic_metadata['Country']\n",
    "metagenomic_metadata['captivity_status'] = np.where(metagenomic_metadata['Westernized']=='Yes', 'western_human', 'non_western_human') \n",
    "metagenomic_metadata['amplicon'] = 'META'\n",
    "metagenomic_metadata['sample'] = metagenomic_metadata['Sample ID']\n",
    "metagenomic_metadata['Description'] = metagenomic_metadata['captivity_status']\n",
    "metagenomic_metadata_select = metagenomic_metadata[['X.SampleID','BarcodeSequence','LinkerPrimerSequence','Old_SampleID','dataset',\n",
    "                     'genus_sp','subspecies','common_name','uniqueid','site','site_code','country',\n",
    "                     'captivity_status','amplicon','sample','Description']]\n",
    "metagenomic_metadata_select.head()\n",
    "metagenomic_metadata_select.to_csv(,sep='\\t',index=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.system('cp metadata/metadata_Bt_samples_gyrb.txt metadata/metadata_Bt_samples_gyrb_passoli.txt')\n",
    "with open('metadata/metadata_Bt_samples_gyrb_passoli.txt','a') as g:\n",
    "    with open('metadata/passoli_metadata_formatted.txt','r') as f:\n",
    "        f.readline()\n",
    "        for l in f:\n",
    "            g.write(l)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>wd.chi.GM.9.Bt</th>\n",
       "      <th>wd.gor.CP.CR1426.Bt</th>\n",
       "      <th>wd.gor.CP.CR5756.Bt</th>\n",
       "      <th>wd.gor.CP.CR6097.Bt</th>\n",
       "      <th>wd.gor.CP.CR2138.Bt</th>\n",
       "      <th>we.hum.CT.B11.Bt</th>\n",
       "      <th>we.hum.CT.C12.Bt</th>\n",
       "      <th>wd.chi.GM.341.Bt</th>\n",
       "      <th>wd.chi.GM.76.Bt</th>\n",
       "      <th>...</th>\n",
       "      <th>ZellerG_2014__CCMD88272491ST-21-0</th>\n",
       "      <th>ZellerG_2014__CCMD90311071ST-21-0</th>\n",
       "      <th>ZellerG_2014__CCMD93344354ST-21-0</th>\n",
       "      <th>ZellerG_2014__CCMD95433940ST-21-0</th>\n",
       "      <th>ZellerG_2014__CCMD96553385ST-21-0</th>\n",
       "      <th>ZellerG_2014__MMPU29365221ST</th>\n",
       "      <th>ZellerG_2014__MMPU68403337ST</th>\n",
       "      <th>ZellerG_2014__MMPU72854103ST</th>\n",
       "      <th>ZellerG_2014__MMPU84450604ST</th>\n",
       "      <th>ZellerG_2014__MMPU99077057ST</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASV_1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASV_2</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASV_3</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASV_4</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASV_5</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 7519 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  Unnamed: 0  wd.chi.GM.9.Bt  wd.gor.CP.CR1426.Bt  wd.gor.CP.CR5756.Bt  \\\n",
       "0      ASV_1               0                    0                    0   \n",
       "1      ASV_2               0                    0                    0   \n",
       "2      ASV_3               0                    0                    0   \n",
       "3      ASV_4               0                    0                    0   \n",
       "4      ASV_5               0                    0                    0   \n",
       "\n",
       "   wd.gor.CP.CR6097.Bt  wd.gor.CP.CR2138.Bt  we.hum.CT.B11.Bt  \\\n",
       "0                    0                    0                 0   \n",
       "1                    0                    0                 1   \n",
       "2                    0                    0                 1   \n",
       "3                    0                    0                 0   \n",
       "4                    0                    0                 1   \n",
       "\n",
       "   we.hum.CT.C12.Bt  wd.chi.GM.341.Bt  wd.chi.GM.76.Bt  ...  \\\n",
       "0                 1                 0                0  ...   \n",
       "1                 1                 0                0  ...   \n",
       "2                 1                 0                0  ...   \n",
       "3                 1                 0                0  ...   \n",
       "4                 0                 1                1  ...   \n",
       "\n",
       "   ZellerG_2014__CCMD88272491ST-21-0  ZellerG_2014__CCMD90311071ST-21-0  \\\n",
       "0                                  1                                  1   \n",
       "1                                  1                                  1   \n",
       "2                                  0                                  1   \n",
       "3                                  1                                  0   \n",
       "4                                  0                                  0   \n",
       "\n",
       "   ZellerG_2014__CCMD93344354ST-21-0  ZellerG_2014__CCMD95433940ST-21-0  \\\n",
       "0                                  0                                  0   \n",
       "1                                  1                                  0   \n",
       "2                                  1                                  0   \n",
       "3                                  1                                  1   \n",
       "4                                  1                                  0   \n",
       "\n",
       "   ZellerG_2014__CCMD96553385ST-21-0  ZellerG_2014__MMPU29365221ST  \\\n",
       "0                                  1                             1   \n",
       "1                                  1                             1   \n",
       "2                                  0                             1   \n",
       "3                                  0                             0   \n",
       "4                                  0                             0   \n",
       "\n",
       "   ZellerG_2014__MMPU68403337ST  ZellerG_2014__MMPU72854103ST  \\\n",
       "0                             0                             0   \n",
       "1                             0                             1   \n",
       "2                             1                             1   \n",
       "3                             1                             0   \n",
       "4                             0                             0   \n",
       "\n",
       "   ZellerG_2014__MMPU84450604ST  ZellerG_2014__MMPU99077057ST  \n",
       "0                             1                             1  \n",
       "1                             0                             1  \n",
       "2                             1                             1  \n",
       "3                             1                             0  \n",
       "4                             0                             0  \n",
       "\n",
       "[5 rows x 7519 columns]"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "all_samples_metadata = pd.read_csv('metadata/metadata_Bt_samples_gyrb_passoli.txt',sep='\\t')\n",
    "all_samples_metadata.tail()\n",
    "asv_tab = pd.read_csv('results/gyrb/inputs/ASVs_filtered_counts.tsv',sep='\\t')\n",
    "asv_tab.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'Unnamed: 0'}"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "set(list(asv_tab.columns)) - set(list(all_samples_metadata['X.SampleID']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_samples = asv_tab[['Unnamed: 0','wd.chi.GM.1203dup.Bt','cp.chi.HOUZ.strains.Bt',\n",
    "'cp.ora.HOUZ.strains.Bt',\n",
    "'cp.gor.COLZ.strains.Bt',\n",
    "'cp.gor.HOUZ.strains.Bt']]\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "ASVs = strain_samples[strain_samples['cp.chi.HOUZ.strains.Bt']+\n",
    "               strain_samples['cp.ora.HOUZ.strains.Bt']+\n",
    "               strain_samples['cp.gor.COLZ.strains.Bt']+\n",
    "               strain_samples['cp.gor.HOUZ.strains.Bt']!=0]['Unnamed: 0']\n",
    "\n",
    "asvs_in_strain_samples = asv_tab[asv_tab['Unnamed: 0'].isin(list(ASVs))]\n",
    "asvs_in_strain_samples.to_csv('asvs_in_strain_samples.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "wd.chi.GM.1203dup.Bt 17\n",
      "cp.chi.HOUZ.strains.Bt 13\n",
      "cp.ora.HOUZ.strains.Bt 4\n",
      "cp.gor.COLZ.strains.Bt 6\n",
      "cp.gor.HOUZ.strains.Bt 3\n"
     ]
    }
   ],
   "source": [
    "for i in ['wd.chi.GM.1203dup.Bt','cp.chi.HOUZ.strains.Bt',\n",
    "'cp.ora.HOUZ.strains.Bt',\n",
    "'cp.gor.COLZ.strains.Bt',\n",
    "'cp.gor.HOUZ.strains.Bt']:\n",
    "    print(i,sum(strain_samples[i]))\n",
    "    print()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [],
   "source": [
    "strain_samples.to_csv('test.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
