{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This scripts generates a metadata sheet with 9000 plus individual metagenomic samples by reformatting the metadata from Passoli et al to be compatible with this study."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "import os "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "root_dir = '/Volumes/AHN/captive_ape_microbiome/'\n",
    "os.chdir(root_dir)\n",
    "\n",
    "analysis_dir = 'results/gyrb/processing/metagenomic_samples/'\n",
    "os.chdir(analysis_dir)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Study</th>\n",
       "      <th>Sample ID</th>\n",
       "      <th># Reads</th>\n",
       "      <th># Bases</th>\n",
       "      <th>Median Read Length</th>\n",
       "      <th>Body Site</th>\n",
       "      <th>Body Subsite</th>\n",
       "      <th>Age (years)</th>\n",
       "      <th>Age Category</th>\n",
       "      <th>Infant Age (days)</th>\n",
       "      <th>Westernized</th>\n",
       "      <th>Country</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI1_t1Q14</td>\n",
       "      <td>27553455</td>\n",
       "      <td>2748531713</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>90.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI2_t1Q14</td>\n",
       "      <td>32197545</td>\n",
       "      <td>3210606434</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>90.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI3_t1Q14</td>\n",
       "      <td>32310216</td>\n",
       "      <td>3222955818</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>90.0</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI4_t1Q14</td>\n",
       "      <td>14270590</td>\n",
       "      <td>1423299652</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>MV_FEI4_t2Q15</td>\n",
       "      <td>37677351</td>\n",
       "      <td>3735487055</td>\n",
       "      <td>100.0</td>\n",
       "      <td>Stool</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1</td>\n",
       "      <td>Newborn</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Yes</td>\n",
       "      <td>ITA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           Study      Sample ID   # Reads     # Bases  Median Read Length  \\\n",
       "0  AsnicarF_2017  MV_FEI1_t1Q14  27553455  2748531713               100.0   \n",
       "1  AsnicarF_2017  MV_FEI2_t1Q14  32197545  3210606434               100.0   \n",
       "2  AsnicarF_2017  MV_FEI3_t1Q14  32310216  3222955818               100.0   \n",
       "3  AsnicarF_2017  MV_FEI4_t1Q14  14270590  1423299652               100.0   \n",
       "4  AsnicarF_2017  MV_FEI4_t2Q15  37677351  3735487055               100.0   \n",
       "\n",
       "  Body Site Body Subsite Age (years) Age Category  Infant Age (days)  \\\n",
       "0     Stool          NaN           1      Newborn               90.0   \n",
       "1     Stool          NaN           1      Newborn               90.0   \n",
       "2     Stool          NaN           1      Newborn               90.0   \n",
       "3     Stool          NaN           1      Newborn                NaN   \n",
       "4     Stool          NaN           1      Newborn                NaN   \n",
       "\n",
       "  Westernized Country  \n",
       "0         Yes     ITA  \n",
       "1         Yes     ITA  \n",
       "2         Yes     ITA  \n",
       "3         Yes     ITA  \n",
       "4         Yes     ITA  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metagenomic_metadata_file = 'metagenomic_metadata_all_samples.txt'\n",
    "metagenomic_metadata = pd.read_csv(metagenomic_metadata_file,sep='\\t')\n",
    "metagenomic_metadata.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>X.SampleID</th>\n",
       "      <th>Old_SampleID</th>\n",
       "      <th>dataset</th>\n",
       "      <th>genus_sp</th>\n",
       "      <th>subspecies</th>\n",
       "      <th>common_name</th>\n",
       "      <th>uniqueid</th>\n",
       "      <th>site</th>\n",
       "      <th>site_code</th>\n",
       "      <th>country</th>\n",
       "      <th>captivity_status</th>\n",
       "      <th>amplicon</th>\n",
       "      <th>sample</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AsnicarF_2017__MV_FEI1_t1Q14</td>\n",
       "      <td>MV_FEI1_t1Q14</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>human</td>\n",
       "      <td>NA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>industrialized_human</td>\n",
       "      <td>META</td>\n",
       "      <td>MV_FEI1_t1Q14</td>\n",
       "      <td>industrialized_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AsnicarF_2017__MV_FEI2_t1Q14</td>\n",
       "      <td>MV_FEI2_t1Q14</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>human</td>\n",
       "      <td>NA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>industrialized_human</td>\n",
       "      <td>META</td>\n",
       "      <td>MV_FEI2_t1Q14</td>\n",
       "      <td>industrialized_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AsnicarF_2017__MV_FEI3_t1Q14</td>\n",
       "      <td>MV_FEI3_t1Q14</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>human</td>\n",
       "      <td>NA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>industrialized_human</td>\n",
       "      <td>META</td>\n",
       "      <td>MV_FEI3_t1Q14</td>\n",
       "      <td>industrialized_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AsnicarF_2017__MV_FEI4_t1Q14</td>\n",
       "      <td>MV_FEI4_t1Q14</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>human</td>\n",
       "      <td>NA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>industrialized_human</td>\n",
       "      <td>META</td>\n",
       "      <td>MV_FEI4_t1Q14</td>\n",
       "      <td>industrialized_human</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AsnicarF_2017__MV_FEI4_t2Q15</td>\n",
       "      <td>MV_FEI4_t2Q15</td>\n",
       "      <td>AsnicarF_2017</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>Homo_sapiens</td>\n",
       "      <td>human</td>\n",
       "      <td>NA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>ITA</td>\n",
       "      <td>industrialized_human</td>\n",
       "      <td>META</td>\n",
       "      <td>MV_FEI4_t2Q15</td>\n",
       "      <td>industrialized_human</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                     X.SampleID   Old_SampleID        dataset      genus_sp  \\\n",
       "0  AsnicarF_2017__MV_FEI1_t1Q14  MV_FEI1_t1Q14  AsnicarF_2017  Homo_sapiens   \n",
       "1  AsnicarF_2017__MV_FEI2_t1Q14  MV_FEI2_t1Q14  AsnicarF_2017  Homo_sapiens   \n",
       "2  AsnicarF_2017__MV_FEI3_t1Q14  MV_FEI3_t1Q14  AsnicarF_2017  Homo_sapiens   \n",
       "3  AsnicarF_2017__MV_FEI4_t1Q14  MV_FEI4_t1Q14  AsnicarF_2017  Homo_sapiens   \n",
       "4  AsnicarF_2017__MV_FEI4_t2Q15  MV_FEI4_t2Q15  AsnicarF_2017  Homo_sapiens   \n",
       "\n",
       "     subspecies common_name uniqueid site site_code country  \\\n",
       "0  Homo_sapiens       human       NA  ITA       ITA     ITA   \n",
       "1  Homo_sapiens       human       NA  ITA       ITA     ITA   \n",
       "2  Homo_sapiens       human       NA  ITA       ITA     ITA   \n",
       "3  Homo_sapiens       human       NA  ITA       ITA     ITA   \n",
       "4  Homo_sapiens       human       NA  ITA       ITA     ITA   \n",
       "\n",
       "       captivity_status amplicon         sample           Description  \n",
       "0  industrialized_human     META  MV_FEI1_t1Q14  industrialized_human  \n",
       "1  industrialized_human     META  MV_FEI2_t1Q14  industrialized_human  \n",
       "2  industrialized_human     META  MV_FEI3_t1Q14  industrialized_human  \n",
       "3  industrialized_human     META  MV_FEI4_t1Q14  industrialized_human  \n",
       "4  industrialized_human     META  MV_FEI4_t2Q15  industrialized_human  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "metagenomic_metadata['X.SampleID'] = metagenomic_metadata.apply(lambda row: row['Study'] +'__' + row['Sample ID'],axis=1)\n",
    "metagenomic_metadata['X.SampleID'] = metagenomic_metadata['X.SampleID'].replace('.','').replace('-','')\n",
    "metagenomic_metadata['Old_SampleID'] = metagenomic_metadata['Sample ID']\n",
    "metagenomic_metadata['dataset'] = metagenomic_metadata['Study']\n",
    "metagenomic_metadata['genus_sp'] = 'Homo_sapiens'\n",
    "metagenomic_metadata['subspecies']  = 'Homo_sapiens'\n",
    "metagenomic_metadata['common_name'] = 'human'\n",
    "metagenomic_metadata['uniqueid'] = 'NA'\n",
    "metagenomic_metadata['site'] = metagenomic_metadata['Country']\n",
    "metagenomic_metadata['site_code'] = metagenomic_metadata['Country']\n",
    "metagenomic_metadata['country'] = metagenomic_metadata['Country']\n",
    "metagenomic_metadata['captivity_status'] = np.where(metagenomic_metadata['Westernized']=='Yes', 'industrialized_human', 'non_industrialized_human') \n",
    "metagenomic_metadata['amplicon'] = 'META'\n",
    "metagenomic_metadata['sample'] = metagenomic_metadata['Sample ID']\n",
    "metagenomic_metadata['Description'] = metagenomic_metadata['captivity_status']\n",
    "metagenomic_metadata_select = metagenomic_metadata[['X.SampleID','Old_SampleID','dataset',\n",
    "                     'genus_sp','subspecies','common_name','uniqueid','site','site_code','country',\n",
    "                     'captivity_status','amplicon','sample','Description']]\n",
    "\n",
    "metagenomic_metadata_select.to_csv('metagenomic_metadata_all_samples_formatted.txt',sep='\\t',index=None,header=None)\n",
    "metagenomic_metadata_select.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.chdir(root_dir)\n",
    "metagenomic_metadata = analysis_dir + 'metagenomic_metadata_all_samples_formatted.txt'\n",
    "metagenomic_metadata\n",
    "amplicon_metadata = 'metadata/metadata_gyrb_amp.txt'\n",
    "amplicon_metagenomic_metadata = 'metadata/metadata_gyrb_amp_meta.txt'\n",
    "os.system(\"AWK '{print $0}' \"+amplicon_metadata+\" \"+metagenomic_metadata+\" > \"+amplicon_metagenomic_metadata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
