{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script finds the co-diversifying clades within Moeller et al. supplemental fasta by creates a phylogeny and selecting the subtrees that match the topology of figure 1. \n",
    "\n",
    "Inputs: fasta of all Bacteroidaceae seqs, ref_gyrb_gtdbtk Bacteroidales reference seqs (fna and faa)\n",
    "\n",
    "Outputs: fasta of just co-div seqs with lineage and clade labels"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq\n",
    "from ete3 import Tree"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome')\n",
    "outdir = 'results/gyrb/processing/moeller_sup'\n",
    "final_outdir = 'results/gyrb/inputs'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system(f'mkdir -pv {outdir}')\n",
    "os.system(f'mkdir -pv {final_outdir}')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### subset ref sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "genus_count = {}\n",
    "with open('ref_seqs/gtdbtk_gyrb.fasta') as original_fasta:\n",
    "    with open(outdir+'/gtdbtk_gyrb.fasta', 'w') as subset_fasta:\n",
    "        records = SeqIO.parse(original_fasta, 'fasta')\n",
    "        for record in records:\n",
    "            if 'o__Bacteroidales' in record.description or 'c__Chlorobia' in record.description or 'c__Ignavibacteria' in record.description:\n",
    "                record.id = record.description.replace(' ','').replace(';','')\n",
    "                record.description = ''\n",
    "                SeqIO.write(record, subset_fasta, 'fasta')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate phylogeny of moeller sup and gtdbtk ref seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 2.5 with Python 3.7.7\n",
      "Command line parameters: -g CGGAGGTAARTTCGAYAAAGG --overlap 21 -e .15 --discard-untrimmed -o results/gyrb/processing/moeller_sup/Bacteroidaceae.fna data/gyrb/moeller_sup/Bacteroidaceae.fna\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.02 s (21 us/read; 2.84 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                     724\n",
      "Reads with adapters:                       724 (100.0%)\n",
      "Reads written (passing filters):           724 (100.0%)\n",
      "\n",
      "Total basepairs processed:       217,200 bp\n",
      "Total written (filtered):        200,991 bp (92.5%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CGGAGGTAARTTCGAYAAAGG; Type: regular 5'; Length: 21; Trimmed: 724 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-5 bp: 0; 6-12 bp: 1; 13-19 bp: 2; 20-21 bp: 3\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "21\t194\t0.0\t3\t194\n",
      "22\t192\t0.0\t3\t192\n",
      "23\t201\t0.0\t3\t201\n",
      "24\t137\t0.0\t3\t137\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Translate nucleic acid sequences\n",
      "Translate nucleic acid sequences\n",
      "Generate an alignment of nucleic coding regions from aligned proteins\n",
      "Error: Guide protein sequence Gorilla3374716_2 not found in nucleic sequence Gorilla3374716\n",
      "Error: Guide protein sequence Human3058374_2 not found in nucleic sequence Human3058374\n",
      "Error: Guide protein sequence Human9403092_2 not found in nucleic sequence Human9403092\n",
      "Error: Guide protein sequence Chimp10413648_2 not found in nucleic sequence Chimp10413648\n",
      "Error: Guide protein sequence Chimp10757854_2 not found in nucleic sequence Chimp10757854\n",
      "Error: Guide protein sequence Chimp10796222_2 not found in nucleic sequence Chimp10796222\n",
      "Error: Guide protein sequence Chimp11562025_2 not found in nucleic sequence Chimp11562025\n",
      "Error: Guide protein sequence Chimp3146215_2 not found in nucleic sequence Chimp3146215\n",
      "Error: Guide protein sequence Chimp8175258_2 not found in nucleic sequence Chimp8175258\n",
      "Error: Guide protein sequence Bonobo2919026_2 not found in nucleic sequence Bonobo2919026\n",
      "Error: Guide protein sequence Bonobo4229344_2 not found in nucleic sequence Bonobo4229344\n",
      "Error: Guide protein sequence Bonobo6282208_2 not found in nucleic sequence Bonobo6282208\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "outdir=results/gyrb/processing/moeller_sup\n",
    "\n",
    "#trim primer sequences\n",
    "cutadapt  -g CGGAGGTAARTTCGAYAAAGG  --overlap 21 -e .15 --discard-untrimmed  -o  $outdir/Bacteroidaceae.fna  data/gyrb/moeller_sup/Bacteroidaceae.fna\n",
    "\n",
    "#translate gyrb amplicon seqs\n",
    "transeq -frame 2 -sequence $outdir/Bacteroidaceae.fna  -outseq $outdir/Bacteroidaceae.faa\n",
    "\n",
    "#translate ref seqs\n",
    "transeq -frame 1 -sequence $outdir/gtdbtk_gyrb.fasta  -outseq $outdir/gtdbtk_gyrb.faa\n",
    "\n",
    "#combine gyrb amplicon and ref seqs\n",
    "cat $outdir/gtdbtk_gyrb.faa $outdir/Bacteroidaceae.faa > $outdir/Bacteroidaceae_ref.faa\n",
    "cat $outdir/gtdbtk_gyrb.fasta  $outdir/Bacteroidaceae.fna > $outdir/Bacteroidaceae_ref.fna\n",
    "\n",
    "#align seqs\n",
    "mafft --auto --quiet $outdir/Bacteroidaceae_ref.faa > $outdir/Bacteroidaceae_ref.faa.aln\n",
    "tranalign -asequence  $outdir/Bacteroidaceae_ref.fna -bsequence $outdir/Bacteroidaceae_ref.faa.aln -outseq $outdir/Bacteroidaceae_ref.fna.aln\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "#trim start of gyrB gene\n",
    "def trim_aln(original_fasta,trim_fasta,start_pos,end_pos):\n",
    "    with open(original_fasta) as original_fasta:\n",
    "        with open(trim_fasta, 'w') as trim_fasta:\n",
    "            records = SeqIO.parse(original_fasta, 'fasta')\n",
    "            for record in records:\n",
    "                record.seq = record.seq[start_pos:end_pos] \n",
    "                SeqIO.write(record, trim_fasta, 'fasta')\n",
    "trim_aln(f'{outdir}/Bacteroidaceae_ref.fna.aln',f'{outdir}/Bacteroidaceae_ref.fna.aln.trim',469,3039)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "FastTree Version 2.1.10 Double precision (No SSE3)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: SH-like 1000\n",
      "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
      "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "      0.15 seconds: Top hits for    149 of   1579 seqs (at seed    100)\n",
      "      0.29 seconds: Top hits for    353 of   1579 seqs (at seed    200)\n",
      "      0.39 seconds: Top hits for    450 of   1579 seqs (at seed    300)\n",
      "      0.54 seconds: Top hits for    613 of   1579 seqs (at seed    400)\n",
      "      0.65 seconds: Top hits for    714 of   1579 seqs (at seed    500)\n",
      "      0.77 seconds: Top hits for    923 of   1579 seqs (at seed    600)\n",
      "      0.88 seconds: Top hits for   1090 of   1579 seqs (at seed    700)\n",
      "      0.99 seconds: Top hits for   1155 of   1579 seqs (at seed    900)\n",
      "      1.09 seconds: Top hits for   1578 of   1579 seqs (at seed   1200)\n",
      "      1.41 seconds: Joined    100 of   1576\n",
      "      1.76 seconds: Joined    200 of   1576\n",
      "      2.05 seconds: Joined    300 of   1576\n",
      "      2.40 seconds: Joined    400 of   1576\n",
      "      2.66 seconds: Joined    500 of   1576\n",
      "      2.98 seconds: Joined    600 of   1576\n",
      "      3.45 seconds: Joined    700 of   1576\n",
      "      3.98 seconds: Joined    800 of   1576\n",
      "      4.54 seconds: Joined    900 of   1576\n",
      "      5.07 seconds: Joined   1000 of   1576\n",
      "      5.57 seconds: Joined   1100 of   1576\n",
      "      6.16 seconds: Joined   1200 of   1576\n",
      "      6.72 seconds: Joined   1300 of   1576\n",
      "      7.28 seconds: Joined   1400 of   1576\n",
      "      7.75 seconds: Joined   1500 of   1576\n",
      "Initial topology in 8.06 seconds\n",
      "Refining topology: 42 rounds ME-NNIs, 2 rounds ME-SPRs, 21 rounds ML-NNIs\n",
      "      8.05 seconds: ME NNI round 1 of 42, 1 of 1577 splits\n",
      "      8.16 seconds: ME NNI round 1 of 42, 601 of 1577 splits, 185 changes (max delta 0.035)\n",
      "      8.27 seconds: ME NNI round 1 of 42, 1301 of 1577 splits, 388 changes (max delta 0.053)\n",
      "      8.38 seconds: ME NNI round 2 of 42, 301 of 1577 splits, 43 changes (max delta 0.011)\n",
      "      8.50 seconds: ME NNI round 2 of 42, 901 of 1577 splits, 148 changes (max delta 0.100)\n",
      "      8.62 seconds: ME NNI round 2 of 42, 1501 of 1577 splits, 256 changes (max delta 0.100)\n",
      "      8.73 seconds: ME NNI round 3 of 42, 501 of 1577 splits, 65 changes (max delta 0.011)\n",
      "      8.83 seconds: ME NNI round 3 of 42, 1101 of 1577 splits, 162 changes (max delta 0.100)\n",
      "      8.94 seconds: ME NNI round 4 of 42, 501 of 1577 splits, 52 changes (max delta 0.017)\n",
      "      9.05 seconds: ME NNI round 5 of 42, 201 of 1577 splits, 12 changes (max delta 0.012)\n",
      "      9.16 seconds: ME NNI round 6 of 42, 201 of 1577 splits, 14 changes (max delta 0.015)\n",
      "      9.27 seconds: ME NNI round 8 of 42, 101 of 1577 splits, 8 changes (max delta 0.004)\n",
      "      9.73 seconds: SPR round   1 of   2, 101 of 3156 nodes\n",
      "     10.09 seconds: SPR round   1 of   2, 201 of 3156 nodes\n",
      "     10.38 seconds: SPR round   1 of   2, 301 of 3156 nodes\n",
      "     10.55 seconds: SPR round   1 of   2, 401 of 3156 nodes\n",
      "     10.69 seconds: SPR round   1 of   2, 501 of 3156 nodes\n",
      "     11.05 seconds: SPR round   1 of   2, 601 of 3156 nodes\n",
      "     11.41 seconds: SPR round   1 of   2, 701 of 3156 nodes\n",
      "     11.73 seconds: SPR round   1 of   2, 801 of 3156 nodes\n",
      "     12.11 seconds: SPR round   1 of   2, 901 of 3156 nodes\n",
      "     12.38 seconds: SPR round   1 of   2, 1001 of 3156 nodes\n",
      "     12.66 seconds: SPR round   1 of   2, 1101 of 3156 nodes\n",
      "     12.97 seconds: SPR round   1 of   2, 1201 of 3156 nodes\n",
      "     13.14 seconds: SPR round   1 of   2, 1301 of 3156 nodes\n",
      "     13.38 seconds: SPR round   1 of   2, 1401 of 3156 nodes\n",
      "     13.61 seconds: SPR round   1 of   2, 1501 of 3156 nodes\n",
      "     13.97 seconds: SPR round   1 of   2, 1601 of 3156 nodes\n",
      "     14.30 seconds: SPR round   1 of   2, 1701 of 3156 nodes\n",
      "     14.52 seconds: SPR round   1 of   2, 1801 of 3156 nodes\n",
      "     14.70 seconds: SPR round   1 of   2, 1901 of 3156 nodes\n",
      "     15.05 seconds: SPR round   1 of   2, 2001 of 3156 nodes\n",
      "     15.31 seconds: SPR round   1 of   2, 2101 of 3156 nodes\n",
      "     15.55 seconds: SPR round   1 of   2, 2201 of 3156 nodes\n",
      "     15.88 seconds: SPR round   1 of   2, 2301 of 3156 nodes\n",
      "     16.24 seconds: SPR round   1 of   2, 2401 of 3156 nodes\n",
      "     16.42 seconds: SPR round   1 of   2, 2501 of 3156 nodes\n",
      "     16.64 seconds: SPR round   1 of   2, 2601 of 3156 nodes\n",
      "     17.00 seconds: SPR round   1 of   2, 2701 of 3156 nodes\n",
      "     17.37 seconds: SPR round   1 of   2, 2801 of 3156 nodes\n",
      "     17.76 seconds: SPR round   1 of   2, 2901 of 3156 nodes\n",
      "     18.13 seconds: SPR round   1 of   2, 3001 of 3156 nodes\n",
      "     18.47 seconds: SPR round   1 of   2, 3101 of 3156 nodes\n",
      "     18.68 seconds: ME NNI round 15 of 42, 1 of 1577 splits\n",
      "     18.78 seconds: ME NNI round 15 of 42, 601 of 1577 splits, 19 changes (max delta 0.006)\n",
      "     18.89 seconds: ME NNI round 15 of 42, 1301 of 1577 splits, 49 changes (max delta 0.020)\n",
      "     19.00 seconds: ME NNI round 16 of 42, 301 of 1577 splits, 0 changes\n",
      "     19.11 seconds: ME NNI round 16 of 42, 1001 of 1577 splits, 22 changes (max delta 0.006)\n",
      "     19.22 seconds: ME NNI round 17 of 42, 1 of 1577 splits\n",
      "     19.33 seconds: SPR round   2 of   2, 1 of 3156 nodes\n",
      "     19.69 seconds: SPR round   2 of   2, 101 of 3156 nodes\n",
      "     20.04 seconds: SPR round   2 of   2, 201 of 3156 nodes\n",
      "     20.32 seconds: SPR round   2 of   2, 301 of 3156 nodes\n",
      "     20.49 seconds: SPR round   2 of   2, 401 of 3156 nodes\n",
      "     20.62 seconds: SPR round   2 of   2, 501 of 3156 nodes\n",
      "     20.96 seconds: SPR round   2 of   2, 601 of 3156 nodes\n",
      "     21.29 seconds: SPR round   2 of   2, 701 of 3156 nodes\n",
      "     21.62 seconds: SPR round   2 of   2, 801 of 3156 nodes\n",
      "     21.96 seconds: SPR round   2 of   2, 901 of 3156 nodes\n",
      "     22.23 seconds: SPR round   2 of   2, 1001 of 3156 nodes\n",
      "     22.49 seconds: SPR round   2 of   2, 1101 of 3156 nodes\n",
      "     22.78 seconds: SPR round   2 of   2, 1201 of 3156 nodes\n",
      "     22.96 seconds: SPR round   2 of   2, 1301 of 3156 nodes\n",
      "     23.17 seconds: SPR round   2 of   2, 1401 of 3156 nodes\n",
      "     23.37 seconds: SPR round   2 of   2, 1501 of 3156 nodes\n",
      "     23.52 seconds: SPR round   2 of   2, 1601 of 3156 nodes\n",
      "     23.76 seconds: SPR round   2 of   2, 1701 of 3156 nodes\n",
      "     24.10 seconds: SPR round   2 of   2, 1801 of 3156 nodes\n",
      "     24.40 seconds: SPR round   2 of   2, 1901 of 3156 nodes\n",
      "     24.73 seconds: SPR round   2 of   2, 2001 of 3156 nodes\n",
      "     24.98 seconds: SPR round   2 of   2, 2101 of 3156 nodes\n",
      "     25.32 seconds: SPR round   2 of   2, 2201 of 3156 nodes\n",
      "     25.54 seconds: SPR round   2 of   2, 2301 of 3156 nodes\n",
      "     25.88 seconds: SPR round   2 of   2, 2401 of 3156 nodes\n",
      "     26.06 seconds: SPR round   2 of   2, 2501 of 3156 nodes\n",
      "     26.30 seconds: SPR round   2 of   2, 2601 of 3156 nodes\n",
      "     26.68 seconds: SPR round   2 of   2, 2701 of 3156 nodes\n",
      "     27.06 seconds: SPR round   2 of   2, 2801 of 3156 nodes\n",
      "     27.45 seconds: SPR round   2 of   2, 2901 of 3156 nodes\n",
      "     27.83 seconds: SPR round   2 of   2, 3001 of 3156 nodes\n",
      "     28.19 seconds: SPR round   2 of   2, 3101 of 3156 nodes\n",
      "     28.42 seconds: ME NNI round 29 of 42, 1 of 1577 splits\n",
      "     28.53 seconds: ME NNI round 29 of 42, 601 of 1577 splits, 17 changes (max delta 0.014)\n",
      "     28.63 seconds: ME NNI round 29 of 42, 1201 of 1577 splits, 22 changes (max delta 0.014)\n",
      "     28.74 seconds: ME NNI round 30 of 42, 201 of 1577 splits, 0 changes\n",
      "     28.85 seconds: ME NNI round 30 of 42, 901 of 1577 splits, 13 changes (max delta 0.003)\n",
      "     28.95 seconds: ME NNI round 30 of 42, 1501 of 1577 splits, 22 changes (max delta 0.003)\n",
      "     29.06 seconds: ME NNI round 35 of 42, 1 of 1577 splits\n",
      "Total branch-length 113.625 after 29.41 sec\n",
      "     29.45 seconds: ML Lengths 1 of 1577 splits\n",
      "     29.62 seconds: ML Lengths 101 of 1577 splits\n",
      "     29.78 seconds: ML Lengths 201 of 1577 splits\n",
      "     29.93 seconds: ML Lengths 301 of 1577 splits\n",
      "     30.10 seconds: ML Lengths 401 of 1577 splits\n",
      "     30.26 seconds: ML Lengths 501 of 1577 splits\n",
      "     30.42 seconds: ML Lengths 601 of 1577 splits\n",
      "     30.56 seconds: ML Lengths 701 of 1577 splits\n",
      "     30.72 seconds: ML Lengths 801 of 1577 splits\n",
      "     30.87 seconds: ML Lengths 901 of 1577 splits\n",
      "     31.02 seconds: ML Lengths 1001 of 1577 splits\n",
      "     31.18 seconds: ML Lengths 1101 of 1577 splits\n",
      "     31.33 seconds: ML Lengths 1201 of 1577 splits\n",
      "     31.49 seconds: ML Lengths 1301 of 1577 splits\n",
      "     31.65 seconds: ML Lengths 1401 of 1577 splits\n",
      "     31.82 seconds: ML Lengths 1501 of 1577 splits\n",
      "     31.95 seconds: ML NNI round 1 of 21, 1 of 1577 splits\n",
      "     32.47 seconds: ML NNI round 1 of 21, 101 of 1577 splits, 16 changes (max delta 17.201)\n",
      "     32.94 seconds: ML NNI round 1 of 21, 201 of 1577 splits, 44 changes (max delta 17.201)\n",
      "     33.46 seconds: ML NNI round 1 of 21, 301 of 1577 splits, 80 changes (max delta 39.383)\n",
      "     33.98 seconds: ML NNI round 1 of 21, 401 of 1577 splits, 107 changes (max delta 39.383)\n",
      "     34.50 seconds: ML NNI round 1 of 21, 501 of 1577 splits, 146 changes (max delta 39.383)\n",
      "     35.00 seconds: ML NNI round 1 of 21, 601 of 1577 splits, 177 changes (max delta 39.383)\n",
      "     35.52 seconds: ML NNI round 1 of 21, 701 of 1577 splits, 218 changes (max delta 39.383)\n",
      "     36.03 seconds: ML NNI round 1 of 21, 801 of 1577 splits, 237 changes (max delta 39.383)\n",
      "     36.52 seconds: ML NNI round 1 of 21, 901 of 1577 splits, 269 changes (max delta 39.383)\n",
      "     37.03 seconds: ML NNI round 1 of 21, 1001 of 1577 splits, 289 changes (max delta 39.383)\n",
      "     37.57 seconds: ML NNI round 1 of 21, 1101 of 1577 splits, 315 changes (max delta 74.398)\n",
      "     38.10 seconds: ML NNI round 1 of 21, 1201 of 1577 splits, 353 changes (max delta 74.398)\n",
      "     38.59 seconds: ML NNI round 1 of 21, 1301 of 1577 splits, 394 changes (max delta 74.398)\n",
      "     39.12 seconds: ML NNI round 1 of 21, 1401 of 1577 splits, 404 changes (max delta 74.398)\n",
      "     39.72 seconds: ML NNI round 1 of 21, 1501 of 1577 splits, 418 changes (max delta 74.398)\n",
      "ML-NNI round 1: LogLk = -1024100.914 NNIs 426 max delta 74.40 Time 40.19\n",
      "     40.19 seconds: Optimizing GTR model, step 1 of 12\n",
      "     41.54 seconds: Optimizing GTR model, step 2 of 12\n",
      "     44.14 seconds: Optimizing GTR model, step 4 of 12\n",
      "     45.69 seconds: Optimizing GTR model, step 5 of 12\n",
      "     47.24 seconds: Optimizing GTR model, step 6 of 12\n",
      "     49.03 seconds: Optimizing GTR model, step 7 of 12\n",
      "     49.97 seconds: Optimizing GTR model, step 8 of 12\n",
      "     52.14 seconds: Optimizing GTR model, step 10 of 12\n",
      "     54.42 seconds: Optimizing GTR model, step 12 of 12\n",
      "GTR Frequencies: 0.2824 0.2388 0.2590 0.2198\n",
      "GTR rates(ac ag at cg ct gt) 1.8393 2.6022 1.7376 1.6691 5.4926 1.0000\n",
      "     55.57 seconds: ML Lengths 1 of 1577 splits\n",
      "     55.75 seconds: ML Lengths 101 of 1577 splits\n",
      "     55.94 seconds: ML Lengths 201 of 1577 splits\n",
      "     56.12 seconds: ML Lengths 301 of 1577 splits\n",
      "     56.27 seconds: ML Lengths 401 of 1577 splits\n",
      "     56.44 seconds: ML Lengths 501 of 1577 splits\n",
      "     56.60 seconds: ML Lengths 601 of 1577 splits\n",
      "     56.77 seconds: ML Lengths 701 of 1577 splits\n",
      "     56.94 seconds: ML Lengths 801 of 1577 splits\n",
      "     57.11 seconds: ML Lengths 901 of 1577 splits\n",
      "     57.27 seconds: ML Lengths 1001 of 1577 splits\n",
      "     57.43 seconds: ML Lengths 1101 of 1577 splits\n",
      "     57.61 seconds: ML Lengths 1201 of 1577 splits\n",
      "     57.79 seconds: ML Lengths 1301 of 1577 splits\n",
      "     57.97 seconds: ML Lengths 1401 of 1577 splits\n",
      "     58.13 seconds: ML Lengths 1501 of 1577 splits\n",
      "     58.36 seconds: Site likelihoods with rate category 1 of 20\n",
      "     58.48 seconds: Site likelihoods with rate category 2 of 20\n",
      "     58.60 seconds: Site likelihoods with rate category 3 of 20\n",
      "     58.72 seconds: Site likelihoods with rate category 4 of 20\n",
      "     58.84 seconds: Site likelihoods with rate category 5 of 20\n",
      "     58.96 seconds: Site likelihoods with rate category 6 of 20\n",
      "     59.09 seconds: Site likelihoods with rate category 7 of 20\n",
      "     59.20 seconds: Site likelihoods with rate category 8 of 20\n",
      "     59.32 seconds: Site likelihoods with rate category 9 of 20\n",
      "     59.43 seconds: Site likelihoods with rate category 10 of 20\n",
      "     59.55 seconds: Site likelihoods with rate category 11 of 20\n",
      "     59.67 seconds: Site likelihoods with rate category 12 of 20\n",
      "     59.79 seconds: Site likelihoods with rate category 13 of 20\n",
      "     59.91 seconds: Site likelihoods with rate category 14 of 20\n",
      "     60.03 seconds: Site likelihoods with rate category 15 of 20\n",
      "     60.15 seconds: Site likelihoods with rate category 16 of 20\n",
      "     60.28 seconds: Site likelihoods with rate category 17 of 20\n",
      "     60.40 seconds: Site likelihoods with rate category 18 of 20\n",
      "     60.53 seconds: Site likelihoods with rate category 19 of 20\n",
      "     60.65 seconds: Site likelihoods with rate category 20 of 20\n",
      "Switched to using 20 rate categories (CAT approximation)\n",
      "Rate categories were divided by 1.167 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "Use -gamma for approximate but comparable Gamma(20) log-likelihoods\n",
      "     60.83 seconds: ML NNI round 2 of 21, 1 of 1577 splits\n",
      "     61.18 seconds: ML NNI round 2 of 21, 101 of 1577 splits, 13 changes (max delta 21.021)\n",
      "     61.48 seconds: ML NNI round 2 of 21, 201 of 1577 splits, 21 changes (max delta 27.903)\n",
      "     61.81 seconds: ML NNI round 2 of 21, 301 of 1577 splits, 29 changes (max delta 27.903)\n",
      "     62.33 seconds: ML NNI round 2 of 21, 401 of 1577 splits, 65 changes (max delta 27.903)\n",
      "     62.93 seconds: ML NNI round 2 of 21, 501 of 1577 splits, 103 changes (max delta 27.903)\n",
      "     63.53 seconds: ML NNI round 2 of 21, 601 of 1577 splits, 146 changes (max delta 27.903)\n",
      "     63.92 seconds: ML NNI round 2 of 21, 701 of 1577 splits, 163 changes (max delta 27.903)\n",
      "     64.40 seconds: ML NNI round 2 of 21, 801 of 1577 splits, 179 changes (max delta 68.525)\n",
      "     64.86 seconds: ML NNI round 2 of 21, 901 of 1577 splits, 199 changes (max delta 68.525)\n",
      "     65.32 seconds: ML NNI round 2 of 21, 1001 of 1577 splits, 216 changes (max delta 68.525)\n",
      "     65.78 seconds: ML NNI round 2 of 21, 1101 of 1577 splits, 240 changes (max delta 68.525)\n",
      "     66.33 seconds: ML NNI round 2 of 21, 1201 of 1577 splits, 265 changes (max delta 68.525)\n",
      "     66.67 seconds: ML NNI round 2 of 21, 1301 of 1577 splits, 274 changes (max delta 68.525)\n",
      "     66.91 seconds: ML NNI round 2 of 21, 1401 of 1577 splits, 277 changes (max delta 68.525)\n",
      "     67.37 seconds: ML NNI round 2 of 21, 1501 of 1577 splits, 302 changes (max delta 68.525)\n",
      "ML-NNI round 2: LogLk = -840644.892 NNIs 332 max delta 68.52 Time 67.86\n",
      "     67.85 seconds: ML NNI round 3 of 21, 1 of 1577 splits\n",
      "     68.20 seconds: ML NNI round 3 of 21, 101 of 1577 splits, 0 changes\n",
      "     68.54 seconds: ML NNI round 3 of 21, 201 of 1577 splits, 21 changes (max delta 0.000)\n",
      "     68.89 seconds: ML NNI round 3 of 21, 301 of 1577 splits, 39 changes (max delta 5.875)\n",
      "     69.25 seconds: ML NNI round 3 of 21, 401 of 1577 splits, 49 changes (max delta 28.156)\n",
      "     69.76 seconds: ML NNI round 3 of 21, 501 of 1577 splits, 90 changes (max delta 28.156)\n",
      "     70.34 seconds: ML NNI round 3 of 21, 601 of 1577 splits, 125 changes (max delta 28.156)\n",
      "     70.83 seconds: ML NNI round 3 of 21, 701 of 1577 splits, 149 changes (max delta 28.156)\n",
      "     71.10 seconds: ML NNI round 3 of 21, 801 of 1577 splits, 153 changes (max delta 28.156)\n",
      "     71.54 seconds: ML NNI round 3 of 21, 901 of 1577 splits, 168 changes (max delta 28.156)\n",
      "     71.97 seconds: ML NNI round 3 of 21, 1001 of 1577 splits, 180 changes (max delta 28.156)\n",
      "     72.34 seconds: ML NNI round 3 of 21, 1101 of 1577 splits, 200 changes (max delta 28.156)\n",
      "ML-NNI round 3: LogLk = -840278.613 NNIs 225 max delta 28.16 Time 72.73\n",
      "     72.73 seconds: ML NNI round 4 of 21, 1 of 1577 splits\n",
      "     72.85 seconds: ML NNI round 4 of 21, 101 of 1577 splits, 5 changes (max delta 0.000)\n",
      "     73.12 seconds: ML NNI round 4 of 21, 201 of 1577 splits, 20 changes (max delta 5.619)\n",
      "     73.57 seconds: ML NNI round 4 of 21, 301 of 1577 splits, 51 changes (max delta 5.619)\n",
      "     74.08 seconds: ML NNI round 4 of 21, 401 of 1577 splits, 85 changes (max delta 5.619)\n",
      "     74.43 seconds: ML NNI round 4 of 21, 501 of 1577 splits, 96 changes (max delta 5.619)\n",
      "     74.81 seconds: ML NNI round 4 of 21, 601 of 1577 splits, 111 changes (max delta 5.619)\n",
      "     75.15 seconds: ML NNI round 4 of 21, 701 of 1577 splits, 124 changes (max delta 39.477)\n",
      "     75.55 seconds: ML NNI round 4 of 21, 801 of 1577 splits, 148 changes (max delta 39.477)\n",
      "ML-NNI round 4: LogLk = -840176.500 NNIs 153 max delta 39.48 Time 75.72\n",
      "     75.72 seconds: ML NNI round 5 of 21, 1 of 1577 splits\n",
      "     76.05 seconds: ML NNI round 5 of 21, 101 of 1577 splits, 20 changes (max delta 0.000)\n",
      "     76.49 seconds: ML NNI round 5 of 21, 201 of 1577 splits, 47 changes (max delta 0.000)\n",
      "     76.95 seconds: ML NNI round 5 of 21, 301 of 1577 splits, 76 changes (max delta 10.605)\n",
      "     77.29 seconds: ML NNI round 5 of 21, 401 of 1577 splits, 93 changes (max delta 24.351)\n",
      "ML-NNI round 5: LogLk = -840127.373 NNIs 93 max delta 24.35 Time 77.33\n",
      "     77.63 seconds: ML NNI round 6 of 21, 101 of 1577 splits, 22 changes (max delta 3.219)\n",
      "     78.04 seconds: ML NNI round 6 of 21, 201 of 1577 splits, 32 changes (max delta 3.219)\n",
      "ML-NNI round 6: LogLk = -840118.252 NNIs 32 max delta 3.22 Time 78.11\n",
      "     78.48 seconds: ML NNI round 7 of 21, 101 of 1577 splits, 14 changes (max delta 0.000)\n",
      "ML-NNI round 7: LogLk = -840117.529 NNIs 14 max delta 0.00 Time 78.51\n",
      "Turning off heuristics for final round of ML NNIs (converged)\n",
      "     79.25 seconds: ML NNI round 8 of 21, 101 of 1577 splits, 5 changes (max delta 14.047)\n",
      "     79.93 seconds: ML NNI round 8 of 21, 201 of 1577 splits, 10 changes (max delta 14.047)\n",
      "     80.61 seconds: ML NNI round 8 of 21, 301 of 1577 splits, 18 changes (max delta 14.047)\n",
      "     81.26 seconds: ML NNI round 8 of 21, 401 of 1577 splits, 43 changes (max delta 14.047)\n",
      "     81.94 seconds: ML NNI round 8 of 21, 501 of 1577 splits, 51 changes (max delta 14.047)\n",
      "     82.61 seconds: ML NNI round 8 of 21, 601 of 1577 splits, 70 changes (max delta 14.047)\n",
      "     83.24 seconds: ML NNI round 8 of 21, 701 of 1577 splits, 96 changes (max delta 14.047)\n",
      "     83.87 seconds: ML NNI round 8 of 21, 801 of 1577 splits, 119 changes (max delta 14.047)\n",
      "     84.51 seconds: ML NNI round 8 of 21, 901 of 1577 splits, 126 changes (max delta 14.047)\n",
      "     85.11 seconds: ML NNI round 8 of 21, 1001 of 1577 splits, 141 changes (max delta 14.047)\n",
      "     85.79 seconds: ML NNI round 8 of 21, 1101 of 1577 splits, 145 changes (max delta 14.047)\n",
      "     86.54 seconds: ML NNI round 8 of 21, 1201 of 1577 splits, 154 changes (max delta 14.941)\n",
      "     87.21 seconds: ML NNI round 8 of 21, 1301 of 1577 splits, 163 changes (max delta 14.941)\n",
      "     87.92 seconds: ML NNI round 8 of 21, 1401 of 1577 splits, 173 changes (max delta 31.137)\n",
      "     88.53 seconds: ML NNI round 8 of 21, 1501 of 1577 splits, 188 changes (max delta 31.137)\n",
      "ML-NNI round 8: LogLk = -838809.389 NNIs 203 max delta 31.14 Time 89.01 (final)\n",
      "     89.01 seconds: ML Lengths 1 of 1577 splits\n",
      "     89.19 seconds: ML Lengths 101 of 1577 splits\n",
      "     89.37 seconds: ML Lengths 201 of 1577 splits\n",
      "     89.54 seconds: ML Lengths 301 of 1577 splits\n",
      "     89.70 seconds: ML Lengths 401 of 1577 splits\n",
      "     89.88 seconds: ML Lengths 501 of 1577 splits\n",
      "     90.05 seconds: ML Lengths 601 of 1577 splits\n",
      "     90.21 seconds: ML Lengths 701 of 1577 splits\n",
      "     90.38 seconds: ML Lengths 801 of 1577 splits\n",
      "     90.53 seconds: ML Lengths 901 of 1577 splits\n",
      "     90.69 seconds: ML Lengths 1001 of 1577 splits\n",
      "     90.87 seconds: ML Lengths 1101 of 1577 splits\n",
      "     91.04 seconds: ML Lengths 1201 of 1577 splits\n",
      "     91.21 seconds: ML Lengths 1301 of 1577 splits\n",
      "     91.40 seconds: ML Lengths 1401 of 1577 splits\n",
      "     91.55 seconds: ML Lengths 1501 of 1577 splits\n",
      "Optimize all lengths: LogLk = -838778.359 Time 91.70\n",
      "     92.54 seconds: ML split tests for    100 of   1576 internal splits\n",
      "     93.36 seconds: ML split tests for    200 of   1576 internal splits\n",
      "     94.20 seconds: ML split tests for    300 of   1576 internal splits\n",
      "     94.99 seconds: ML split tests for    400 of   1576 internal splits\n",
      "     95.81 seconds: ML split tests for    500 of   1576 internal splits\n",
      "     96.65 seconds: ML split tests for    600 of   1576 internal splits\n",
      "     97.48 seconds: ML split tests for    700 of   1576 internal splits\n",
      "     98.30 seconds: ML split tests for    800 of   1576 internal splits\n",
      "     99.09 seconds: ML split tests for    900 of   1576 internal splits\n",
      "     99.92 seconds: ML split tests for   1000 of   1576 internal splits\n",
      "    100.73 seconds: ML split tests for   1100 of   1576 internal splits\n",
      "    101.57 seconds: ML split tests for   1200 of   1576 internal splits\n",
      "    102.39 seconds: ML split tests for   1300 of   1576 internal splits\n",
      "    103.20 seconds: ML split tests for   1400 of   1576 internal splits\n",
      "    104.00 seconds: ML split tests for   1500 of   1576 internal splits\n",
      "Total time: 104.61 seconds Unique: 1579/1657 Bad splits: 24/1576 Worst delta-LogLk 27.087\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "fasttree -nt -gtr < results/gyrb/processing/moeller_sup/Bacteroidaceae_ref.fna.aln.trim > results/gyrb/processing/moeller_sup/Bacteroidaceae_ref.fna.aln.trim.tre\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### extract codiv clades from tree\n",
    "view tree in figtree,  \n",
    "root based on outgroups \n",
    "select clades with topology matching Figure 1 in Moeller et al. \n",
    "save subtrees separately\n",
    "same sequence names separately, annotate with clade designations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "in_tree = f'{outdir}/Bacteroidaceae_ref.fna.aln.trim.tre'   \n",
    "rooted_tree = f'{outdir}//Bacteroidaceae_ref.fna.aln.trim.rooted.tre'  \n",
    "Bt1_tree_file = f'{outdir}/moeller_codiv_lin_Bt1.tree'  \n",
    "Bt2_tree_file = f'{outdir}/moeller_codiv_lin_Bt2.tree'  \n",
    "Bt3_tree_file = f'{outdir}/moeller_codiv_lin_Bt3.tree'  \n",
    "\n",
    "tree = Tree(in_tree, format=0)\n",
    "outgroup_taxa=[leaf.name for leaf in tree.get_leaves() if \n",
    "               'c__Ignavibacteria' in leaf.name or \n",
    "               'c__Chlorobia' in leaf.name]\n",
    "outgroup_MRCA = tree.get_common_ancestor(outgroup_taxa)\n",
    "tree.set_outgroup(outgroup_MRCA)\n",
    "tree.write(outfile=rooted_tree,format=2)\n",
    "\n",
    "def get_lineage_ASVs(in_tree,listTaxa):    \n",
    "    #makes sure we don't miss any ASVs that didn't match in the blast search \\n\"\n",
    "    #but are descended from the MRCA of those that did\\n\"\n",
    "    tree = Tree(in_tree, format=0)\n",
    "    lineage_MRCA = tree.get_common_ancestor(listTaxa)\n",
    "    linage_taxa = [x.name for x in lineage_MRCA.get_leaves()] \n",
    "    tree.prune(linage_taxa)\n",
    "    return(tree)\n",
    "\n",
    "def rename_taxa(lin_tree,listTaxa,cladeName):\n",
    "    lineage_MRCA = lin_tree.get_common_ancestor(listTaxa)\n",
    "    for taxa in lineage_MRCA.get_leaves():    \n",
    "        taxa.name = taxa.name + ' ' + cladeName\n",
    "\n",
    "Bt1_tree = get_lineage_ASVs(rooted_tree,['Gorilla5764570','Bonobo7657683'])\n",
    "Bt1_taxa_oldnames = [leaf.name for leaf in Bt1_tree.get_leaves()]        \n",
    "rename_taxa(Bt1_tree,['Gorilla5764570','Gorilla10693345'],'Bt1_clade1_gorilla')\n",
    "rename_taxa(Bt1_tree,['Chimp6685161','Chimp5888147'],'Bt1_clade1_chimp')\n",
    "rename_taxa(Bt1_tree,['Bonobo11821974','Bonobo7657683'],'Bt1_clade1_bonobo')\n",
    "Bt1_taxa = [leaf.name for leaf in Bt1_tree.get_leaves()] \n",
    "Bt1_tree.write(outfile=Bt1_tree_file,format=2) \n",
    "\n",
    "Bt2_tree = get_lineage_ASVs(rooted_tree,['Gorilla7248598',\n",
    "                                         'Bonobo6937590'])\n",
    "Bt2_taxa_oldnames = [leaf.name for leaf in Bt2_tree.get_leaves()]        \n",
    "rename_taxa(Bt2_tree,['Gorilla7248598','Gorilla1454159'],'Bt2_clade1_gorilla')\n",
    "rename_taxa(Bt2_tree,['Chimp7736944','Chimp3970093'],'Bt2_clade1_chimp')\n",
    "rename_taxa(Bt2_tree,['Bonobo11725448','Bonobo4214590'],'Bt2_clade1_bonobo')\n",
    "rename_taxa(Bt2_tree,['Chimp3781815','Chimp3450568'],'Bt2_clade2_chimp')\n",
    "rename_taxa(Bt2_tree,['Bonobo11213703','Bonobo6937590'],'Bt2_clade2_bonobo')\n",
    "Bt2_taxa = [leaf.name for leaf in Bt2_tree.get_leaves()] \n",
    "Bt2_tree.write(outfile=Bt2_tree_file,format=2) \n",
    "\n",
    "Bt3_tree = get_lineage_ASVs(rooted_tree,['Human5838232','Bonobo10623458'])\n",
    "Bt3_taxa_oldnames = [leaf.name for leaf in Bt3_tree.get_leaves()]        \n",
    "rename_taxa(Bt3_tree,['Human5838232','Human2935124'],'Bt3_clade1_human')\n",
    "rename_taxa(Bt3_tree,['Chimp1528180','Chimp873638'],'Bt3_clade1_chimp')\n",
    "rename_taxa(Bt3_tree,['Bonobo6212174','Bonobo4232320'],'Bt3_clade1_bonobo')\n",
    "Bt3_taxa = [leaf.name for leaf in Bt3_tree.get_leaves()] \n",
    "Bt3_tree.write(outfile=Bt3_tree_file,format=2) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### create table with all codiv sequences, clade, and lineage info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [],
   "source": [
    "codiv_taxa_oldnames = Bt1_taxa_oldnames + Bt2_taxa_oldnames + Bt3_taxa_oldnames\n",
    "codiv_taxa = Bt1_taxa + Bt2_taxa + Bt3_taxa\n",
    "codivDict = dict(zip(codiv_taxa_oldnames, codiv_taxa))\n",
    "\n",
    "codivDF = pd.DataFrame(list(zip(codiv_taxa,codiv_taxa_oldnames)),columns=['codiv_taxa','codiv_taxa_oldnames'])\n",
    "\n",
    "codivDF['moeller_seq'] = codivDF['codiv_taxa'].apply(lambda x: 'moeller' if 'p__Bacteroidota' not in x else 'ref')\n",
    "codivDF = codivDF[codivDF['moeller_seq'] == 'moeller']\n",
    "codivDF['codiv_clade'] = codivDF['codiv_taxa'].apply(lambda x: x.split(' ')[1])\n",
    "codivDF['codiv_lineage'] = codivDF['codiv_clade'].apply(lambda x: x.split('_')[0])\n",
    "codivDF['hostSpecies'] = codivDF['codiv_clade'].apply(lambda x: x.split('_')[-1])\n",
    "codivDF = codivDF[['codiv_taxa','codiv_clade','codiv_lineage','hostSpecies']]\n",
    "codivDF.to_csv(f'{outdir}/moeller_codiv_HRclades.txt',index=False,sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### subset supplement fasta to only have codiv sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open(f'{outdir}/Bacteroidaceae.fna') as original: \n",
    "    with open(f'{outdir}/moeller_codiv_Bacteroidaceae.fna', 'w') as covid:\n",
    "        records = SeqIO.parse(original, 'fasta')\n",
    "        for record in records:\n",
    "            if record.id in codiv_taxa_oldnames:\n",
    "                record.id = codivDict[record.id]\n",
    "                record.description = ''\n",
    "                SeqIO.write(record, covid, 'fasta')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### copy over output files used by analysis scripts "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.system(f'cp {outdir}/moeller_codiv_Bacteroidaceae.fna {final_outdir}/moeller_codiv_Bacteroidaceae.fna')\n",
    "os.system(f'cp {outdir}/moeller_codiv_HRclades.txt {final_outdir}/moeller_codiv_HRclades.txt')\n",
    "os.system(f'cp {outdir}/moeller_codiv_lin_Bt1.tree {final_outdir}/moeller_codiv_lin_Bt1.tree')\n",
    "os.system(f'cp {outdir}/moeller_codiv_lin_Bt2.tree {final_outdir}/moeller_codiv_lin_Bt2.tree')\n",
    "os.system(f'cp {outdir}/moeller_codiv_lin_Bt3.tree {final_outdir}/moeller_codiv_lin_Bt3.tree')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
