{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "This script finds the co-diversifying clades within Moeller et al. supplemental fasta by creates a phylogeny and selecting the subtrees that match the topology of figure 1. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os\n",
    "from Bio import SeqIO\n",
    "from Bio import Seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Generate phylogeny"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This is cutadapt 2.5 with Python 3.7.7\n",
      "Command line parameters: -g CGGAGGTAARTTCGAYAAAGG --overlap 21 -e .15 --discard-untrimmed -o results/gyrb/processing/moeller_sup/Bacteroidaceae.fna data/gyrb/moeller_sup/Bacteroidaceae.fna\n",
      "Processing reads on 1 core in single-end mode ...\n",
      "Finished in 0.04 s (53 us/read; 1.13 M reads/minute).\n",
      "\n",
      "=== Summary ===\n",
      "\n",
      "Total reads processed:                     724\n",
      "Reads with adapters:                       724 (100.0%)\n",
      "Reads written (passing filters):           724 (100.0%)\n",
      "\n",
      "Total basepairs processed:       217,200 bp\n",
      "Total written (filtered):        200,991 bp (92.5%)\n",
      "\n",
      "=== Adapter 1 ===\n",
      "\n",
      "Sequence: CGGAGGTAARTTCGAYAAAGG; Type: regular 5'; Length: 21; Trimmed: 724 times.\n",
      "\n",
      "No. of allowed errors:\n",
      "0-5 bp: 0; 6-12 bp: 1; 13-19 bp: 2; 20-21 bp: 3\n",
      "\n",
      "Overview of removed sequences\n",
      "length\tcount\texpect\tmax.err\terror counts\n",
      "21\t194\t0.0\t3\t194\n",
      "22\t192\t0.0\t3\t192\n",
      "23\t201\t0.0\t3\t201\n",
      "24\t137\t0.0\t3\t137\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "cp: results/gyrb_bt_gtdbtk_ref/blastdb/gtdbtk_gyrb_Bt.faa: No such file or directory\n",
      "cp: results/gyrb_bt_gtdbtk_ref/blastdb/gtdbtk_gyrb_Bt.fasta: No such file or directory\n",
      "Translate nucleic acid sequences\n",
      "Generate an alignment of nucleic coding regions from aligned proteins\n",
      "Error: Guide protein sequence Gorilla3374716_2 not found in nucleic sequence Gorilla3374716\n",
      "Error: Guide protein sequence Human3058374_2 not found in nucleic sequence Human3058374\n",
      "Error: Guide protein sequence Human9403092_2 not found in nucleic sequence Human9403092\n",
      "Error: Guide protein sequence Chimp10413648_2 not found in nucleic sequence Chimp10413648\n",
      "Error: Guide protein sequence Chimp10757854_2 not found in nucleic sequence Chimp10757854\n",
      "Error: Guide protein sequence Chimp10796222_2 not found in nucleic sequence Chimp10796222\n",
      "Error: Guide protein sequence Chimp11562025_2 not found in nucleic sequence Chimp11562025\n",
      "Error: Guide protein sequence Chimp3146215_2 not found in nucleic sequence Chimp3146215\n",
      "Error: Guide protein sequence Chimp8175258_2 not found in nucleic sequence Chimp8175258\n",
      "Error: Guide protein sequence Bonobo2919026_2 not found in nucleic sequence Bonobo2919026\n",
      "Error: Guide protein sequence Bonobo4229344_2 not found in nucleic sequence Bonobo4229344\n",
      "Error: Guide protein sequence Bonobo6282208_2 not found in nucleic sequence Bonobo6282208\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "mkdir -pv results/gyrb/processing/moeller_sup\n",
    "\n",
    "#copy reference fasta\n",
    "cp results/gyrb_bt_gtdbtk_ref/blastdb/gtdbtk_gyrb_Bt.faa results/gyrb/processing/moeller_sup/gtdbtk_gyrb_Bt.faa\n",
    "cp results/gyrb_bt_gtdbtk_ref/blastdb/gtdbtk_gyrb_Bt.fasta results/gyrb/processing/moeller_sup/gtdbtk_gyrb_Bt.fasta\n",
    "\n",
    "#trim primer sequences\n",
    "cutadapt  -g CGGAGGTAARTTCGAYAAAGG  --overlap 21 -e .15 --discard-untrimmed  -o  results/gyrb/processing/moeller_sup/Bacteroidaceae.fna  data/gyrb/moeller_sup/Bacteroidaceae.fna\n",
    "\n",
    "cd results/gyrb/processing/moeller_sup\n",
    "#translate gyrb amplicon seqs\n",
    "transeq -frame 2 -sequence Bacteroidaceae.fna  -outseq Bacteroidaceae.faa\n",
    "\n",
    "#combine gyrb amplicon and ref seqs\n",
    "cat gtdbtk_gyrb_Bt.faa Bacteroidaceae.faa > Bacteroidaceae_ref.faa\n",
    "cat gtdbtk_gyrb_Bt.fasta Bacteroidaceae.fna > Bacteroidaceae_ref.fna\n",
    "\n",
    "#align seqs\n",
    "mafft --auto --quiet Bacteroidaceae_ref.faa > Bacteroidaceae_ref.faa.aln\n",
    "tranalign -asequence  Bacteroidaceae_ref.fna -bsequence Bacteroidaceae_ref.faa.aln -outseq Bacteroidaceae_ref.fna.aln\n",
    "\n",
    "#select range of sites because some have seqs have different "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "FastTree Version 2.1.10 Double precision (No SSE3)\n",
      "Alignment: standard input\n",
      "Nucleotide distances: Jukes-Cantor Joins: balanced Support: SH-like 1000\n",
      "Search: Normal +NNI +SPR (2 rounds range 10) +ML-NNI opt-each=1\n",
      "TopHits: 1.00*sqrtN close=default refresh=0.80\n",
      "ML Model: Generalized Time-Reversible, CAT approximation with 20 rate categories\n",
      "Initial topology in 7.30 seconds0 of   1526   529 seqs (at seed   1100)   \n",
      "Refining topology: 42 rounds ME-NNIs, 2 rounds ME-SPRs, 21 rounds ML-NNIs\n",
      "Total branch-length 107.195 after 27.43 sec1 of 1527 splits   , 24 changes (max delta 0.003)   \n",
      "ML-NNI round 1: LogLk = -963821.905 NNIs 422 max delta 74.59 Time 37.58ges (max delta 74.595)   \n",
      "GTR Frequencies: 0.2797 0.2408 0.2599 0.2196ep 11 of 12   \n",
      "GTR rates(ac ag at cg ct gt) 1.8762 2.6509 1.7148 1.7302 5.5709 1.0000\n",
      "Switched to using 20 rate categories (CAT approximation)20 of 20   \n",
      "Rate categories were divided by 1.149 so that average rate = 1.0\n",
      "CAT-based log-likelihoods may not be comparable across runs\n",
      "Use -gamma for approximate but comparable Gamma(20) log-likelihoods\n",
      "ML-NNI round 2: LogLk = -789490.762 NNIs 361 max delta 56.83 Time 63.79ges (max delta 56.826)   \n",
      "ML-NNI round 3: LogLk = -789072.921 NNIs 214 max delta 37.20 Time 68.85ges (max delta 37.202)   \n",
      "ML-NNI round 4: LogLk = -788977.814 NNIs 142 max delta 24.21 Time 71.75es (max delta 24.215)   \n",
      "ML-NNI round 5: LogLk = -788933.158 NNIs 63 max delta 33.92 Time 73.24es (max delta 33.923)   \n",
      "ML-NNI round 6: LogLk = -788922.289 NNIs 24 max delta 4.43 Time 73.85ges (max delta 4.430)   \n",
      "ML-NNI round 7: LogLk = -788910.112 NNIs 15 max delta 9.17 Time 74.27ges (max delta 9.166)   \n",
      "ML-NNI round 8: LogLk = -788907.744 NNIs 8 max delta 1.77 Time 74.57\n",
      "ML-NNI round 9: LogLk = -788907.514 NNIs 0 max delta 0.00 Time 74.64\n",
      "Turning off heuristics for final round of ML NNIs (converged)\n",
      "ML-NNI round 10: LogLk = -787645.453 NNIs 201 max delta 50.25 Time 84.59 (final) delta 50.253)   \n",
      "Optimize all lengths: LogLk = -787608.163 Time 87.17 \n",
      "Total time: 99.56 seconds Unique: 1529/1607 Bad splits: 15/1526 Worst delta-LogLk 28.003\n"
     ]
    }
   ],
   "source": [
    "!fasttree -nt -gtr < results/gyrb/processing/moeller_sup/Bacteroidaceae_ref.fna.aln.selection > results/gyrb/processing/moeller_sup/Bacteroidaceae_ref.fna.aln.selection.tre"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#view tree in figtree, manually select clades with topology matching Figure 1 in Moeller et al. \n",
    "#save subtrees separately\n",
    "#same sequence names separately, annotate with clade designations"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "codiv_asvs = pd.read_csv('results/moeller_sup/Bt_all_codiv_lineages_names.txt',sep=' ',index_col=None,header=None)\n",
    "codiv_asvs.columns = ['codiv_clade','header']\n",
    "codivDict = dict(zip(codiv_asvs['header'], codiv_asvs['codiv_clade']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Gorilla10693345\n",
      "Gorilla10893151\n",
      "Gorilla12208365\n",
      "Gorilla1454159\n",
      "Gorilla3280123\n",
      "Gorilla3748318\n",
      "Gorilla4351822\n",
      "Gorilla4447500\n",
      "Gorilla5764570\n",
      "Gorilla6306341\n",
      "Gorilla7248598\n",
      "Gorilla7611566\n",
      "Gorilla7655534\n",
      "Gorilla8443136\n",
      "Gorilla8497213\n",
      "Gorilla8907336\n",
      "Gorilla9193943\n",
      "Gorilla9805772\n",
      "Human10045430\n",
      "Human10283488\n",
      "Human10889793\n",
      "Human10912985\n",
      "Human10912992\n",
      "Human12108738\n",
      "Human1330489\n",
      "Human1330495\n",
      "Human1781917\n",
      "Human237218\n",
      "Human2572122\n",
      "Human2935124\n",
      "Human3034958\n",
      "Human3367149\n",
      "Human3418197\n",
      "Human3961325\n",
      "Human4630891\n",
      "Human4639158\n",
      "Human4639166\n",
      "Human5110045\n",
      "Human5110051\n",
      "Human5465528\n",
      "Human5838232\n",
      "Human6390749\n",
      "Human6405675\n",
      "Human663970\n",
      "Human663977\n",
      "Human6649477\n",
      "Human6700665\n",
      "Human6700667\n",
      "Human7250453\n",
      "Human7563063\n",
      "Human7563070\n",
      "Human7859459\n",
      "Human8146958\n",
      "Human8295127\n",
      "Human8829943\n",
      "Human8842535\n",
      "Human9177678\n",
      "Human9518949\n",
      "Human9825266\n",
      "Chimp10218980\n",
      "Chimp1067447\n",
      "Chimp11049685\n",
      "Chimp11943611\n",
      "Chimp1322981\n",
      "Chimp1324508\n",
      "Chimp1493190\n",
      "Chimp1528180\n",
      "Chimp168543\n",
      "Chimp1862511\n",
      "Chimp2164238\n",
      "Chimp2506335\n",
      "Chimp2629519\n",
      "Chimp2757936\n",
      "Chimp2936652\n",
      "Chimp304514\n",
      "Chimp3094078\n",
      "Chimp3094087\n",
      "Chimp3162311\n",
      "Chimp3450568\n",
      "Chimp3450573\n",
      "Chimp3781815\n",
      "Chimp3970093\n",
      "Chimp4678997\n",
      "Chimp5183434\n",
      "Chimp5211526\n",
      "Chimp5888147\n",
      "Chimp665501\n",
      "Chimp6685161\n",
      "Chimp6973326\n",
      "Chimp6973327\n",
      "Chimp7736944\n",
      "Chimp7736952\n",
      "Chimp8269086\n",
      "Chimp828594\n",
      "Chimp873638\n",
      "Chimp8740623\n",
      "Chimp8780480\n",
      "Chimp9057405\n",
      "Chimp9057407\n",
      "Chimp9228565\n",
      "Chimp934253\n",
      "Chimp985651\n",
      "Chimp9859517\n",
      "Chimp9931019\n",
      "Chimp9931021\n",
      "Bonobo10034489\n",
      "Bonobo10057151\n",
      "Bonobo10352189\n",
      "Bonobo10513316\n",
      "Bonobo10520847\n",
      "Bonobo10609639\n",
      "Bonobo10623458\n",
      "Bonobo1078\n",
      "Bonobo11021063\n",
      "Bonobo11105611\n",
      "Bonobo11213703\n",
      "Bonobo11242356\n",
      "Bonobo11367003\n",
      "Bonobo11646549\n",
      "Bonobo11725448\n",
      "Bonobo11725461\n",
      "Bonobo11768907\n",
      "Bonobo11821974\n",
      "Bonobo12102095\n",
      "Bonobo12157978\n",
      "Bonobo12387329\n",
      "Bonobo1306358\n",
      "Bonobo1624513\n",
      "Bonobo1787673\n",
      "Bonobo1873599\n",
      "Bonobo1946453\n",
      "Bonobo2078905\n",
      "Bonobo2166737\n",
      "Bonobo2193524\n",
      "Bonobo229236\n",
      "Bonobo229241\n",
      "Bonobo23228\n",
      "Bonobo2366405\n",
      "Bonobo2450173\n",
      "Bonobo2645793\n",
      "Bonobo2919016\n",
      "Bonobo3176696\n",
      "Bonobo3311273\n",
      "Bonobo3371043\n",
      "Bonobo3681147\n",
      "Bonobo3793226\n",
      "Bonobo3971783\n",
      "Bonobo3994923\n",
      "Bonobo4076472\n",
      "Bonobo4186256\n",
      "Bonobo4208740\n",
      "Bonobo4208742\n",
      "Bonobo4214590\n",
      "Bonobo4214599\n",
      "Bonobo4232320\n",
      "Bonobo4255176\n",
      "Bonobo4255178\n",
      "Bonobo4400446\n",
      "Bonobo4469078\n",
      "Bonobo447600\n",
      "Bonobo4568548\n",
      "Bonobo4632650\n",
      "Bonobo4632659\n",
      "Bonobo4668782\n",
      "Bonobo4886707\n",
      "Bonobo4935167\n",
      "Bonobo4950863\n",
      "Bonobo5054462\n",
      "Bonobo5348599\n",
      "Bonobo5507217\n",
      "Bonobo5555405\n",
      "Bonobo5596313\n",
      "Bonobo5703402\n",
      "Bonobo5731561\n",
      "Bonobo5805938\n",
      "Bonobo5856903\n",
      "Bonobo5991239\n",
      "Bonobo6156032\n",
      "Bonobo6212174\n",
      "Bonobo6331470\n",
      "Bonobo6396279\n",
      "Bonobo6757245\n",
      "Bonobo6757253\n",
      "Bonobo6828899\n",
      "Bonobo6932003\n",
      "Bonobo6937590\n",
      "Bonobo7084162\n",
      "Bonobo7228450\n",
      "Bonobo742599\n",
      "Bonobo759069\n",
      "Bonobo7657683\n",
      "Bonobo781247\n",
      "Bonobo7895675\n",
      "Bonobo7932810\n",
      "Bonobo8148512\n",
      "Bonobo8213484\n",
      "Bonobo8808876\n",
      "Bonobo8808878\n",
      "Bonobo8821370\n",
      "Bonobo8827876\n",
      "Bonobo8964982\n",
      "Bonobo9176466\n",
      "Bonobo9375067\n",
      "Bonobo9719\n",
      "Bonobo972742\n",
      "Bonobo9728\n",
      "Bonobo9832770\n",
      "Bonobo9846630\n"
     ]
    }
   ],
   "source": [
    "\n",
    "with open('results/moeller_sup/Bacteroidaceae.fna') as original: \n",
    "    with open('results/moeller_sup/Bacteroidaceae_codiv.fna', 'w') as covid:\n",
    "        records = SeqIO.parse(original, 'fasta')\n",
    "        for record in records:\n",
    "            if record.id in codivDict.keys():\n",
    "                print(record.id)\n",
    "                HR_clade = codivDict[record.id]\n",
    "                record.description = HR_clade\n",
    "                SeqIO.write(record, covid, 'fasta')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
