{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from ete3 import Tree\n",
    "from collections import Counter\n",
    "\n",
    "os.getcwd()\n",
    "os.chdir('/Volumes/AHN/captive_ape_microbiome/results/gyrb/')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "#inputs\n",
    "metadata_file = 'inputs/metadata_gyrb_amp_meta_passing_samples.txt'\n",
    "tax_table_file = 'inputs/ASVs_taxonomy.txt'\n",
    "asv_table_file = 'inputs/ASVs_filtered_counts.tsv'\n",
    "asv_fasta_file = 'inputs/ASVs_filtered.fasta'\n",
    "full_tree_file = 'inputs/ASVs_filtered_ref_full.tree'\n",
    "full_tree = Tree(full_tree_file, format=0)\n",
    "moeller_codiv_fasta = 'inputs/codiv_Bacteroidaceae.fna'\n",
    "pident_cutoff = 95"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "\n",
    "sampleNames = asv_table.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "ASV_sampleName_dict = dict(zip(sampleNames.index,sampleNames))\n",
    "\n",
    "#sample to sample type category\n",
    "metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "sample_type_dict = dict(zip(metadata['X.SampleID'], metadata['Description']))\n",
    "\n",
    "#taxonomic info, family and genus\n",
    "tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "tax_fam_dict = dict(zip(tax_table['ASV'], tax_table['Family']))\n",
    "tax_gen_dict = dict(zip(tax_table['ASV'], tax_table['Genus']))\n",
    "    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Determine ASVs that hit to Moeller co-div clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 08/08/2020 23:12:20\n",
      "New DB name:   /Volumes/AHN/captive_ape_microbiome/results/gyrb/analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "New DB title:  analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "Sequence type: Nucleotide\n",
      "Deleted existing Nucleotide BLAST database named /Volumes/AHN/captive_ape_microbiome/results/gyrb/analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 208 sequences in 0.0111239 seconds.\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "mkdir: analyses/codiv_moeller_ASVs: File exists\n"
     ]
    }
   ],
   "source": [
    "%%bash\n",
    "\n",
    "mkdir analyses/codiv_moeller_ASVs\n",
    "cp inputs/moeller_codiv_Bacteroidaceae.fna analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna\n",
    "\n",
    "#make blastdb\n",
    "makeblastdb -in analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna -dbtype nucl\n",
    "\n",
    "#blast moeller co-div seqs\n",
    "blastn -query inputs/ASVs_filtered.fasta -db analyses/codiv_moeller_ASVs/moeller_codiv_Bacteroidaceae.fna \\\n",
    "-outfmt \"7 qseqid salltitles sseqid pident length qlen evalue\" \\\n",
    "-out analyses/codiv_moeller_ASVs/codiv_blastout_ASVs.txt \\\n",
    "-max_target_seqs 5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "314 ASVs matching co-div clades with greater than 95 percent identity\n",
      "      ASV                codiv_clade_seq       codiv_clade lineage\n",
      "0   ASV_1  Human6700665 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "1  ASV_13  Human7859459 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "2  ASV_34  Human9825266 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "3  ASV_37  Human8829943 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "4  ASV_73  Human9825266 Bt3_clade1_human  Bt3_clade1_human     Bt3\n",
      "Bt2_clade1_bonobo     90\n",
      "Bt3_clade1_human      85\n",
      "Bt2_clade2_bonobo     33\n",
      "Bt1_clade1_bonobo     23\n",
      "Bt2_clade1_chimp      20\n",
      "Bt3_clade1_chimp      17\n",
      "Bt3_clade1_bonobo     15\n",
      "Bt1_clade1_gorilla    10\n",
      "Bt1_clade1_chimp       8\n",
      "Bt2_clade1_gorilla     7\n",
      "Bt2_clade2_chimp       6\n",
      "Name: codiv_clade, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#filter blast output to top hit per ASV, remove hits not pass thresholds\n",
    "blast_res = pd.read_csv('analyses/codiv_moeller_ASVs/codiv_blastout_ASVs.txt',sep='\\t',comment='#',header=None)\n",
    "blast_res.columns = ['ASV','codiv_clade_seq','sseqid','pident','length','qlen','evalue']\n",
    "blast_res = blast_res.groupby('ASV').head(1).reset_index(drop=True) #get top hit\n",
    "blast_res_pass = blast_res[blast_res['pident']>pident_cutoff].reset_index(drop=True) #QC filter\n",
    "blast_res_pass = blast_res_pass[blast_res_pass['length']>200].reset_index(drop=True) #QC filter\n",
    "print(len(blast_res_pass),'ASVs matching co-div clades with greater than',pident_cutoff,'percent identity')\n",
    "blast_res_pass['codiv_clade'] = blast_res_pass['codiv_clade_seq'].apply(lambda x: x.split(' ')[1])\n",
    "blast_res_pass['lineage'] = blast_res_pass['codiv_clade'].apply(lambda x: x.split('_')[0]) \n",
    "blast_res_pass = blast_res_pass[['ASV','codiv_clade_seq', 'codiv_clade','lineage']]\n",
    "print(blast_res_pass.head())\n",
    "print(blast_res_pass['codiv_clade'].value_counts())\n",
    "#output hits passing that will be used to determine the mcra of the lineage\n",
    "blast_res_pass.to_csv('analyses/codiv_moeller_ASVs/codiv_clades_ASVs.txt',sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###defining host restricted clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(['human'], 4, 'HR', 'human', False, [], 0)\n",
      "Bacteroidaceae_Unassigned\n"
     ]
    }
   ],
   "source": [
    "def is_HR(sampleNames,sample_type_dict):\n",
    "    \"\"\"given a list of sample names uses sample type dictionary to determine how many sample types are present\n",
    "    designates ASVs as host restricted = 1 sample type or mixed = multiple sample types.\n",
    "    Captive sample types are not considered so some clades will have a 0 sample type length and they can fall within\n",
    "    host-restricted clades or mixed clades or neither\"\"\"\n",
    "    sampleTypes = [sample_type_dict[name] for name in sampleNames]\n",
    "    sampleTypes = [x.replace('non_western_','').replace('western_','') for x in sampleTypes]\n",
    "    neutral_sampleTypes = ['captive_gorilla','captive_bonobo','captive_chimp','captive_orangutan']\n",
    "    HR_sampleTypes = list(set(sampleTypes) - set(neutral_sampleTypes))\n",
    "    HR_sampleNum = len([x for x in sampleTypes if x not in neutral_sampleTypes])\n",
    "    CP_sampleTypes = list(set(sampleTypes) & set(neutral_sampleTypes))\n",
    "    CP_sampleNum = len([x for x in sampleTypes if x in neutral_sampleTypes])\n",
    "    CP_pres = True if len(CP_sampleTypes) > 0 else False\n",
    "    if len(HR_sampleTypes) == 0:\n",
    "        HR_cat,HR_type='CP','CP'\n",
    "    if len(HR_sampleTypes) == 1: #identifies host-restricted clades\n",
    "        HR_cat,HR_type='HR',HR_sampleTypes[0]  \n",
    "    if len(HR_sampleTypes) > 1:    \n",
    "        HR_cat,HR_type ='MX','MX'\n",
    "    return(HR_sampleTypes,HR_sampleNum,HR_cat,HR_type,CP_pres,CP_sampleTypes,CP_sampleNum)\n",
    "           \n",
    "           \n",
    "print(is_HR(['CosteaPI_2017__SID713A076-11-0-0','LiSS_2016__FAT_024-22-42-0', \n",
    "       'LiSS_2016__FAT_024-22-84-0', 'LiSS_2016__FAT_DON_11-22-0-4'],sample_type_dict)) \n",
    "\n",
    "def get_consensus_taxonomy(listASVs,tax_fam_dict,tax_gen_dict):\n",
    "    fam = list(set([tax_fam_dict[ASV] for ASV in listASVs])-set(['Unassigned']))\n",
    "    gen = list(set([tax_gen_dict[ASV] for ASV in listASVs])-set(['Unassigned']))\n",
    "    fam = fam[0].split('__')[1] if len(fam) == 1 else 'Unassigned'\n",
    "    if fam == 'Unassigned':\n",
    "        gen='Unassigned'\n",
    "    else:\n",
    "        gen = gen[0].split('__')[1] if len(gen) == 1 else 'Unassigned'  \n",
    "    return(fam+'_'+gen)   \n",
    "print(get_consensus_taxonomy(['ASV_1','ASV_2','ASV_1078'],tax_fam_dict,tax_gen_dict))\n",
    "\n",
    "def search_clades(tree, samples_cutoff, BS_support,ASV_sampleName_dict):\n",
    "    \"\"\"Finds nodes with at least 50% BS support containing ASVs only found in a single wild ape species\n",
    "    ie wild_gorilla, wild_chimp or wild_bonobo, cpat\"\"\"\n",
    "    clades_prelim = []\n",
    "    counter = 1\n",
    "    for n in tree.traverse():\n",
    "        if n.support > float(BS_support): #makes sure Bootstrap support is over threshold\n",
    "            ASVs = [leaf.name for leaf in n.iter_leaves() if 'ASV' in leaf.name]\n",
    "            sampleNames = list(set(list(flatten([ASV_sampleName_dict[ASV] for ASV in ASVs])))) \n",
    "            cladeName='clade_'+str(counter)\n",
    "            counter+=1\n",
    "            clade = [cladeName,ASVs,sampleNames]\n",
    "            if len(sampleNames)>samples_cutoff:\n",
    "                clades_prelim.append(clade)\n",
    "    clades_prelim = pd.DataFrame(clades_prelim, columns = \n",
    "                          ['cladeName','ASVs','sampleNames'])\n",
    "    clades_prelim['sampleNum'] = clades_prelim['sampleNames'].apply(lambda x: len(x))\n",
    "    clades_prelim['ASVsNum'] = clades_prelim['ASVs'].apply(lambda x: len(x))\n",
    "    return(clades_prelim)\n",
    "\n",
    "def eliminate_redundant_clades(clades_df,offlimits_ASVs):\n",
    "    \"\"\"sorts clades and returns the largest non overlapping clade \"\"\"\n",
    "    df = clades_df.sort_values('ASVsNum',ascending=False) #start with the largest clades first\n",
    "    NRclades = []\n",
    "    for index, row in df.iterrows():\n",
    "        if len(set(row['ASVs']) & set(offlimits_ASVs)) == 0: \n",
    "            offlimits_ASVs = offlimits_ASVs + row['ASVs']\n",
    "            NRclades.append(row['cladeName']) \n",
    "    res = df[df['cladeName'].isin(NRclades)]   \n",
    "    return(res)  \n",
    "\n",
    "def expand_clades(clades_df):\n",
    "    ASV = clades_df.apply(lambda x: pd.Series(x['ASVs']),axis=1).stack().reset_index(level=1, drop=True)\n",
    "    ASV.name = 'ASVs'\n",
    "    clades_ASVs_df = clades_df.drop('ASVs', axis=1).join(ASV)\n",
    "    return(clades_ASVs_df)\n",
    "                                 \n",
    "def host_restricted_clades(asv_table_file,metadata_file,tax_table_file):\n",
    "    asv_table = pd.read_csv(asv_table_file,sep='\\t',index_col=0)\n",
    "    sampleNames = asv_table.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "    ASV_sampleName_dict = dict(zip(sampleNames.index,sampleNames))\n",
    "\n",
    "    #sample to sample type category\n",
    "    metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "    sample_type_dict = dict(zip(metadata['X.SampleID'], metadata['Description']))\n",
    "\n",
    "    #taxonomic info, family and genus\n",
    "    tax_table = pd.read_csv(tax_table_file,sep='\\t',index_col=None)\n",
    "    tax_fam_dict = dict(zip(tax_table['ASV'], tax_table['Family']))\n",
    "    tax_gen_dict = dict(zip(tax_table['ASV'], tax_table['Genus']))\n",
    "    \n",
    "    #search tree for clades\n",
    "    clades_prelim = search_clades(full_tree, 5, .5, ASV_sampleName_dict)\n",
    "    hr =clades_prelim['sampleNames'].apply(lambda x:  pd.Series(is_HR(x,sample_type_dict),\n",
    "                index=['HR_sampleTypes','HR_sampleNum','HR_cat','HR_type','CP_pres','CP_sampleTypes','CP_sampleNum']))\n",
    "    clades_prelim_hr = clades_prelim.merge(hr,right_index=True,left_index=True)\n",
    "\n",
    "    #identify HR clades\n",
    "    HR_clades = clades_prelim_hr[clades_prelim_hr['HR_cat']=='HR']\n",
    "    HR_clades = HR_clades[HR_clades['HR_sampleNum']>5]\n",
    "    HR_clades = eliminate_redundant_clades(HR_clades,[])\n",
    "    HR_clades_ASVs = expand_clades(HR_clades)\n",
    "\n",
    "    #identify MX clades that don't contain any HR clades\n",
    "    MX_clades = clades_prelim_hr[clades_prelim_hr['HR_cat']=='MX']\n",
    "    MX_clades = MX_clades[MX_clades['sampleNum']>5]\n",
    "    MX_clades = eliminate_redundant_clades(MX_clades,offlimits_ASVs=list(HR_clades_ASVs['ASVs']))\n",
    "    MX_clades_ASVs = expand_clades(MX_clades)\n",
    "    \n",
    "    #identify MX clades that don't contain any HR clades\n",
    "    CP_clades = clades_prelim_hr[clades_prelim_hr['HR_cat']=='CP']\n",
    "    CP_clades = CP_clades[CP_clades['sampleNum']>5]\n",
    "    HR_MX_ASVs = list(HR_clades_ASVs['ASVs']) + list(MX_clades_ASVs['ASVs'])\n",
    "    CP_clades = eliminate_redundant_clades(CP_clades,offlimits_ASVs=HR_MX_ASVs)\n",
    "    CP_clades_ASVs = expand_clades(CP_clades)\n",
    "    \n",
    "    #merge dataframes\n",
    "    clades = pd.concat([HR_clades,MX_clades,CP_clades])\n",
    "    clades.reset_index(drop=True, inplace=True)\n",
    "    clades_ASVs = pd.concat([HR_clades_ASVs,MX_clades_ASVs,CP_clades_ASVs])\n",
    "    clades_ASVs.reset_index(drop=True, inplace=True)\n",
    "    \n",
    "    \n",
    "    clades['cladeTax']=clades['ASVs'].apply(lambda x: \n",
    "        get_consensus_taxonomy(x,tax_fam_dict,tax_gen_dict))\n",
    "    \n",
    "    sample_type_counts =  metadata['Description'].value_counts()\n",
    "    description_df = clades['sampleNames'].apply(lambda l: pd.Series(\n",
    "    [sample_type_dict[name] for name in l]).value_counts())\n",
    "    description_df = description_df.fillna(0)  \n",
    "    sample_type_percent = description_df/sample_type_counts\n",
    "    clades=clades.merge(sample_type_percent,left_index=True,right_index=True)  \n",
    "    \n",
    "    return(clades,clades_ASVs)\n",
    "    \n",
    "clades_df,clades_ASVs_df = host_restricted_clades(asv_table_file,metadata_file,tax_table_file)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cladeName</th>\n",
       "      <th>ASVs</th>\n",
       "      <th>sampleNames</th>\n",
       "      <th>sampleNum</th>\n",
       "      <th>ASVsNum</th>\n",
       "      <th>HR_sampleTypes</th>\n",
       "      <th>HR_sampleNum</th>\n",
       "      <th>HR_cat</th>\n",
       "      <th>HR_type</th>\n",
       "      <th>CP_pres</th>\n",
       "      <th>...</th>\n",
       "      <th>cladeTax</th>\n",
       "      <th>captive_bonobo</th>\n",
       "      <th>captive_chimp</th>\n",
       "      <th>captive_gorilla</th>\n",
       "      <th>captive_orangutan</th>\n",
       "      <th>non_western_human</th>\n",
       "      <th>western_human</th>\n",
       "      <th>wild_bonobo</th>\n",
       "      <th>wild_chimp</th>\n",
       "      <th>wild_gorilla</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>clade_41</td>\n",
       "      <td>[ASV_2623, ASV_1788, ASV_7289, ASV_7288, ASV_7...</td>\n",
       "      <td>[HMP_2012__SRS017701, LeChatelierE_2013__MH043...</td>\n",
       "      <td>788</td>\n",
       "      <td>366</td>\n",
       "      <td>[human]</td>\n",
       "      <td>788</td>\n",
       "      <td>HR</td>\n",
       "      <td>human</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>UBA932_Unassigned</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.493151</td>\n",
       "      <td>0.078766</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>clade_8751</td>\n",
       "      <td>[ASV_10387, ASV_408, ASV_3299, ASV_10369, ASV_...</td>\n",
       "      <td>[HeQ_2017__SZAXPI029466-158, SchirmerM_2016__G...</td>\n",
       "      <td>1751</td>\n",
       "      <td>191</td>\n",
       "      <td>[human]</td>\n",
       "      <td>1732</td>\n",
       "      <td>HR</td>\n",
       "      <td>human</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>Bacteroidaceae_Prevotella</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.500000</td>\n",
       "      <td>0.181818</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.616438</td>\n",
       "      <td>0.208229</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>clade_2710</td>\n",
       "      <td>[ASV_1720, ASV_2738, ASV_6484, ASV_8324, ASV_8...</td>\n",
       "      <td>[LiuW_2016__SRR3993059, QinN_2014__LV-11, LiSS...</td>\n",
       "      <td>487</td>\n",
       "      <td>190</td>\n",
       "      <td>[human]</td>\n",
       "      <td>485</td>\n",
       "      <td>HR</td>\n",
       "      <td>human</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>Bacteroidaceae_Unassigned</td>\n",
       "      <td>0.153846</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.465753</td>\n",
       "      <td>0.036297</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>clade_1342</td>\n",
       "      <td>[ASV_8718, ASV_838, ASV_1875, ASV_8091, ASV_28...</td>\n",
       "      <td>[LeChatelierE_2013__MH0430, ZeeviD_2015__PNP_M...</td>\n",
       "      <td>744</td>\n",
       "      <td>182</td>\n",
       "      <td>[human]</td>\n",
       "      <td>744</td>\n",
       "      <td>HR</td>\n",
       "      <td>human</td>\n",
       "      <td>False</td>\n",
       "      <td>...</td>\n",
       "      <td>Bacteroidaceae_Prevotella</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.371820</td>\n",
       "      <td>0.081411</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>clade_4172</td>\n",
       "      <td>[ASV_1694, ASV_6113, ASV_6131, ASV_1340, ASV_6...</td>\n",
       "      <td>[CosteaPI_2017__SID713A055-11-0-0, QinN_2014__...</td>\n",
       "      <td>700</td>\n",
       "      <td>147</td>\n",
       "      <td>[human]</td>\n",
       "      <td>639</td>\n",
       "      <td>HR</td>\n",
       "      <td>human</td>\n",
       "      <td>True</td>\n",
       "      <td>...</td>\n",
       "      <td>Bacteroidaceae_Unassigned</td>\n",
       "      <td>0.923077</td>\n",
       "      <td>0.961538</td>\n",
       "      <td>0.863636</td>\n",
       "      <td>0.454545</td>\n",
       "      <td>0.526419</td>\n",
       "      <td>0.054372</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 22 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "    cladeName                                               ASVs  \\\n",
       "0    clade_41  [ASV_2623, ASV_1788, ASV_7289, ASV_7288, ASV_7...   \n",
       "1  clade_8751  [ASV_10387, ASV_408, ASV_3299, ASV_10369, ASV_...   \n",
       "2  clade_2710  [ASV_1720, ASV_2738, ASV_6484, ASV_8324, ASV_8...   \n",
       "3  clade_1342  [ASV_8718, ASV_838, ASV_1875, ASV_8091, ASV_28...   \n",
       "4  clade_4172  [ASV_1694, ASV_6113, ASV_6131, ASV_1340, ASV_6...   \n",
       "\n",
       "                                         sampleNames  sampleNum  ASVsNum  \\\n",
       "0  [HMP_2012__SRS017701, LeChatelierE_2013__MH043...        788      366   \n",
       "1  [HeQ_2017__SZAXPI029466-158, SchirmerM_2016__G...       1751      191   \n",
       "2  [LiuW_2016__SRR3993059, QinN_2014__LV-11, LiSS...        487      190   \n",
       "3  [LeChatelierE_2013__MH0430, ZeeviD_2015__PNP_M...        744      182   \n",
       "4  [CosteaPI_2017__SID713A055-11-0-0, QinN_2014__...        700      147   \n",
       "\n",
       "  HR_sampleTypes  HR_sampleNum HR_cat HR_type  CP_pres  ...  \\\n",
       "0        [human]           788     HR   human    False  ...   \n",
       "1        [human]          1732     HR   human     True  ...   \n",
       "2        [human]           485     HR   human     True  ...   \n",
       "3        [human]           744     HR   human    False  ...   \n",
       "4        [human]           639     HR   human     True  ...   \n",
       "\n",
       "                    cladeTax  captive_bonobo captive_chimp  captive_gorilla  \\\n",
       "0          UBA932_Unassigned        0.000000      0.000000         0.000000   \n",
       "1  Bacteroidaceae_Prevotella        0.153846      0.500000         0.181818   \n",
       "2  Bacteroidaceae_Unassigned        0.153846      0.000000         0.000000   \n",
       "3  Bacteroidaceae_Prevotella        0.000000      0.000000         0.000000   \n",
       "4  Bacteroidaceae_Unassigned        0.923077      0.961538         0.863636   \n",
       "\n",
       "   captive_orangutan  non_western_human  western_human  wild_bonobo  \\\n",
       "0           0.000000           0.493151       0.078766          0.0   \n",
       "1           0.000000           0.616438       0.208229          0.0   \n",
       "2           0.000000           0.465753       0.036297          0.0   \n",
       "3           0.000000           0.371820       0.081411          0.0   \n",
       "4           0.454545           0.526419       0.054372          0.0   \n",
       "\n",
       "   wild_chimp  wild_gorilla  \n",
       "0         0.0           0.0  \n",
       "1         0.0           0.0  \n",
       "2         0.0           0.0  \n",
       "3         0.0           0.0  \n",
       "4         0.0           0.0  \n",
       "\n",
       "[5 rows x 22 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clades_df_sh = clades_df.drop(['ASVs','sampleNames'], axis=1)\n",
    "clades_df_sh.to_csv('analyses/codiv_moeller_ASVs/table_full_tree_clades_collapsed.txt',sep='\\t',index=False)\n",
    "clades_df.head()\n",
    "clades_df_cp = clades_df.loc[(clades_df.captive_bonobo > 0.15) | \n",
    "              (clades_df.captive_chimp > 0.15)  |\n",
    "              (clades_df.captive_gorilla > 0.15)|\n",
    "              (clades_df.captive_orangutan > 0.15)] \n",
    "clades_df_sh.to_csv('analyses/codiv_moeller_ASVs/table_prominent_captive_clades.txt',sep='\\t',index=False)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### search full tree for host-restricted clades"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "collasped_tree = Tree(full_tree_file, format=0)\n",
    "\n",
    "def collapse_node(tree,clade_ASVs,cladeName):\n",
    "    if len(clade_ASVs) > 1:\n",
    "        node = tree.get_common_ancestor(clade_ASVs)\n",
    "        node.name = cladeName\n",
    "        children = node.get_children()\n",
    "        node.remove_child(children[1])\n",
    "        node.remove_child(children[0])\n",
    "    else:\n",
    "        ASV = clade_ASVs[0]\n",
    "        leaf = tree.get_leaves_by_name(name=ASV)[0]\n",
    "        leaf.name = cladeName\n",
    "\n",
    "for clade,row in clades_df.iterrows():\n",
    "    collapse_node(collasped_tree,row['ASVs'],row['cladeName'])\n",
    "\n",
    "clade_leaves = [leaf for leaf in collasped_tree.get_leaves() if 'clade' in leaf.name] \n",
    "collasped_tree.prune(clade_leaves) #eliminate leftover ASVs and ref taxa\n",
    "collasped_tree.write(format=2, outfile='analyses/codiv_moeller_ASVs/full_tree_clades_collapsed.tre')\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## ASV info "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#AllASVs\n",
    "allASVs = [leaf.name for leaf in full_tree.get_leaves() if 'ASV' in leaf.name]\n",
    "allASVs = pd.DataFrame(allASVs,columns=['ASV'])\n",
    "\n",
    "#Host restricted clades \n",
    "clades_ASV = pd.concat([HRclades_ASVs,MXclades_ASVs,CPclades_ASVs])\n",
    "clades_ASV['ASV'].value_counts() #double check each \n",
    "HRclades_ASV = clades_ASV[clades_ASV['HR_cat']=='HR']\n",
    "HRclades_ASV = HRclades_ASV[['ASV','cladeName','HR_clade']]\n",
    "HRclades_ASV.shape\n",
    "\n",
    "#Moeller codiv clades\n",
    "print(blast_res_pass.head())\n",
    "\n",
    "#combine\n",
    "allASVs_HRclade = allASVs.merge(HRclades_ASV, how='left', on='ASV')\n",
    "allASVs_HRclade_codiv = allASVs_HRclade.merge(blast_res_pass, how='left', on='ASV')\n",
    "\n",
    "#add taxonomy\n",
    "allASVs_HRclade_codiv['Family'] = allASVs_HRclade_codiv['ASV'].apply(lambda ASV: tax_fam_dict[ASV])\n",
    "allASVs_HRclade_codiv['Genus'] = allASVs_HRclade_codiv['ASV'].apply(lambda ASV: tax_gen_dict[ASV])\n",
    "\n",
    "#add captive ape data\n",
    "metadata['Description_site'] = metadata['Description']+'_' +metadata['site_code']\n",
    "sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "\n",
    "def get_captive_sample(ASV):\n",
    "    captive_desc = ['captive_chimp_HOUZ','captive_gorilla_HOUZ','captive_orangutan_HOUZ',\n",
    "                    'captive_bonobo_COLZ','captive_gorilla_COLZ','captive_orangutan_COLZ','captive_chimp_PC']\n",
    "    res = pd.Series([0 for nil in captive_desc],index=captive_desc)\n",
    "    sampleNames = ASV_sampleName_dict[ASV]\n",
    "    \n",
    "    desc = [sample_type_site_dict[sample] for sample in sampleNames]\n",
    "    desc = [d for d in desc if d in captive_desc]\n",
    "    desc = pd.Series(desc).value_counts()\n",
    "    desc_res = res.add(desc,fill_value=0)\n",
    "    \n",
    "    HR_res = pd.Series(is_HR(sampleNames),index=['sampleTypes','ASV_cat','HR_ASV','ASV_sampleNum'])\n",
    "    res = pd.concat([HR_res,desc_res])\n",
    "    return(res)\n",
    "\n",
    "capt_desc = allASVs_HRclade_codiv['ASV'].apply(lambda ASV: get_captive_sample(ASV))\n",
    "capt_desc['captive_all']= capt_desc.loc[:, 'captive_chimp_HOUZ':'captive_chimp_PC'].sum(axis=1)\n",
    "\n",
    "allASVs_HRclade_codiv_capt = allASVs_HRclade_codiv.merge(capt_desc,left_index=True, right_index=True)\n",
    "\n",
    "#output to tsv\n",
    "allASVs_HRclade_codiv_capt.to_csv('analyses/codiv_moeller_ASVs/full_tree_ASV_table.txt',sep='\\t',index=False)\n",
    "allASVs_HRclade_codiv_capt.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
