{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "from collections import Counter"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "os.chdir('/Volumes/AHN/captive_ape_microbiome')\n",
    "%run scripts/analyses/functions.ipynb"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "INDIR = 'results/16s/inputs'\n",
    "metadata_file = f'{INDIR}/16S_metadata.txt'\n",
    "tax_table_file = f'{INDIR}/ASVs_taxonomy.txt'\n",
    "asv_table_file = f'{INDIR}/ASV_tab.txt'\n",
    "OUTDIR = 'results/16s/analyses'\n",
    "!mkdir -pv $OUTDIR/tables\n",
    "!mkdir -pv $OUTDIR/figures"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### designate 16-ASVs as host-restricted, mixed host, or unique to captive"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2042 total 16S ASVs\n",
      "872 16S ASVs found in captive apes\n",
      "MX_human_wild_apes    486\n",
      "HR_human              230\n",
      "Unique_CP             104\n",
      "MX_wild_apes           36\n",
      "HR_wild_chimp           9\n",
      "HR_wild_gorilla         7\n",
      "Name: HR_type, dtype: int64\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>sampleNames</th>\n",
       "      <th>sampleNum</th>\n",
       "      <th>HR_sampleTypes</th>\n",
       "      <th>HR_sampleNum</th>\n",
       "      <th>HR_cat</th>\n",
       "      <th>HR_type</th>\n",
       "      <th>CP_pres</th>\n",
       "      <th>CP_sampleNum</th>\n",
       "      <th>CP_sampleTypes</th>\n",
       "      <th>captiveNames</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASV_2425</td>\n",
       "      <td>[cp.bon.COLZ.2.16s, cp.ora.COLZ.1.16s, cp.ora....</td>\n",
       "      <td>7</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>Unique_CP</td>\n",
       "      <td>Unique_CP</td>\n",
       "      <td>True</td>\n",
       "      <td>7</td>\n",
       "      <td>[captive_orangutan, captive_bonobo, captive_ch...</td>\n",
       "      <td>[cp.bon.COLZ.2.16s, cp.ora.COLZ.1.16s, cp.ora....</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Lachnospiraceae_NK4A136_group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASV_905</td>\n",
       "      <td>[cp.ora.ATLZ.121.16s, wd.chi.GM.1.16s, wd.chi....</td>\n",
       "      <td>101</td>\n",
       "      <td>[human, wild_gorilla, wild_bonobo, wild_chimp]</td>\n",
       "      <td>99</td>\n",
       "      <td>MX</td>\n",
       "      <td>MX_human_wild_apes</td>\n",
       "      <td>True</td>\n",
       "      <td>2</td>\n",
       "      <td>[captive_orangutan, captive_gorilla]</td>\n",
       "      <td>[cp.ora.ATLZ.121.16s, cp.gor.COLZ.5.16s]</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Lachnospiraceae_AC2044_group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASV_1085</td>\n",
       "      <td>[wd.chi.GM.106.16s, wd.chi.GM.111.16s, wd.chi....</td>\n",
       "      <td>31</td>\n",
       "      <td>[human, wild_chimp, wild_gorilla]</td>\n",
       "      <td>31</td>\n",
       "      <td>MX</td>\n",
       "      <td>MX_human_wild_apes</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>[]</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Lachnospiraceae_NK4A136_group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASV_1261</td>\n",
       "      <td>[cp.ora.ATLZ.121.16s, cp.chi.PC.A2.16S, cp.chi...</td>\n",
       "      <td>15</td>\n",
       "      <td>[human]</td>\n",
       "      <td>11</td>\n",
       "      <td>HR</td>\n",
       "      <td>HR_human</td>\n",
       "      <td>True</td>\n",
       "      <td>4</td>\n",
       "      <td>[captive_orangutan, captive_chimp]</td>\n",
       "      <td>[cp.ora.ATLZ.121.16s, cp.chi.PC.A2.16S, cp.chi...</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Lachnospiraceae_NK4A136_group</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASV_1111</td>\n",
       "      <td>[cp.ora.ATLZ.114.16s, cp.ora.ATLZ.117.16s, cp....</td>\n",
       "      <td>39</td>\n",
       "      <td>[human, wild_gorilla, wild_bonobo, wild_chimp]</td>\n",
       "      <td>26</td>\n",
       "      <td>MX</td>\n",
       "      <td>MX_human_wild_apes</td>\n",
       "      <td>True</td>\n",
       "      <td>13</td>\n",
       "      <td>[captive_orangutan, captive_bonobo, captive_go...</td>\n",
       "      <td>[cp.ora.ATLZ.114.16s, cp.ora.ATLZ.117.16s, cp....</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>unclassified</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        ASV                                        sampleNames  sampleNum  \\\n",
       "0  ASV_2425  [cp.bon.COLZ.2.16s, cp.ora.COLZ.1.16s, cp.ora....          7   \n",
       "1   ASV_905  [cp.ora.ATLZ.121.16s, wd.chi.GM.1.16s, wd.chi....        101   \n",
       "2  ASV_1085  [wd.chi.GM.106.16s, wd.chi.GM.111.16s, wd.chi....         31   \n",
       "3  ASV_1261  [cp.ora.ATLZ.121.16s, cp.chi.PC.A2.16S, cp.chi...         15   \n",
       "4  ASV_1111  [cp.ora.ATLZ.114.16s, cp.ora.ATLZ.117.16s, cp....         39   \n",
       "\n",
       "                                   HR_sampleTypes  HR_sampleNum     HR_cat  \\\n",
       "0                                              []             0  Unique_CP   \n",
       "1  [human, wild_gorilla, wild_bonobo, wild_chimp]            99         MX   \n",
       "2               [human, wild_chimp, wild_gorilla]            31         MX   \n",
       "3                                         [human]            11         HR   \n",
       "4  [human, wild_gorilla, wild_bonobo, wild_chimp]            26         MX   \n",
       "\n",
       "              HR_type  CP_pres  CP_sampleNum  \\\n",
       "0           Unique_CP     True             7   \n",
       "1  MX_human_wild_apes     True             2   \n",
       "2  MX_human_wild_apes    False             0   \n",
       "3            HR_human     True             4   \n",
       "4  MX_human_wild_apes     True            13   \n",
       "\n",
       "                                      CP_sampleTypes  \\\n",
       "0  [captive_orangutan, captive_bonobo, captive_ch...   \n",
       "1               [captive_orangutan, captive_gorilla]   \n",
       "2                                                 []   \n",
       "3                 [captive_orangutan, captive_chimp]   \n",
       "4  [captive_orangutan, captive_bonobo, captive_go...   \n",
       "\n",
       "                                        captiveNames      Phylum  \\\n",
       "0  [cp.bon.COLZ.2.16s, cp.ora.COLZ.1.16s, cp.ora....  Firmicutes   \n",
       "1           [cp.ora.ATLZ.121.16s, cp.gor.COLZ.5.16s]  Firmicutes   \n",
       "2                                                 []  Firmicutes   \n",
       "3  [cp.ora.ATLZ.121.16s, cp.chi.PC.A2.16S, cp.chi...  Firmicutes   \n",
       "4  [cp.ora.ATLZ.114.16s, cp.ora.ATLZ.117.16s, cp....  Firmicutes   \n",
       "\n",
       "           Order           Family                          Genus  \n",
       "0  Clostridiales  Lachnospiraceae  Lachnospiraceae_NK4A136_group  \n",
       "1  Clostridiales  Lachnospiraceae   Lachnospiraceae_AC2044_group  \n",
       "2  Clostridiales  Lachnospiraceae  Lachnospiraceae_NK4A136_group  \n",
       "3  Clostridiales  Lachnospiraceae  Lachnospiraceae_NK4A136_group  \n",
       "4  Clostridiales  Lachnospiraceae                   unclassified  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "asv_16s = asv_hr_table(asv_table_file,metadata_file,tax_table_file) #load from function.ipynb\n",
    "print(len(asv_16s),'total 16S ASVs')\n",
    "asv_16s_captive = asv_16s[asv_16s['CP_pres']==True]\n",
    "print(len(asv_16s_captive),'16S ASVs found in captive apes')\n",
    "print(asv_16s_captive['HR_type'].value_counts())\n",
    "asv_16s.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ASV</th>\n",
       "      <th>sampleNames</th>\n",
       "      <th>sampleNum</th>\n",
       "      <th>HR_sampleTypes</th>\n",
       "      <th>HR_sampleNum</th>\n",
       "      <th>HR_cat</th>\n",
       "      <th>HR_type</th>\n",
       "      <th>CP_pres</th>\n",
       "      <th>CP_sampleNum</th>\n",
       "      <th>CP_sampleTypes</th>\n",
       "      <th>...</th>\n",
       "      <th>ATLZ_captive_orangutan</th>\n",
       "      <th>HOUZ_captive_gorilla</th>\n",
       "      <th>COLZ_captive_gorilla</th>\n",
       "      <th>HOUZ_captive_orangutan</th>\n",
       "      <th>PC_captive_chimp</th>\n",
       "      <th>HOUZ_captive_chimp</th>\n",
       "      <th>COMZ_captive_gorilla</th>\n",
       "      <th>CP_sp_loc</th>\n",
       "      <th>numEnclosure</th>\n",
       "      <th>multi_site_sp</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ASV_617</td>\n",
       "      <td>[wd.gor.BI.a172.16s, wd.gor.BI.a173.16s, wd.go...</td>\n",
       "      <td>108</td>\n",
       "      <td>[wild_gorilla]</td>\n",
       "      <td>108</td>\n",
       "      <td>HR</td>\n",
       "      <td>HR_wild_gorilla</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>single_site_single_sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ASV_982</td>\n",
       "      <td>[wd.gor.BI.a175.16s, wd.gor.BI.a177.16s, wd.go...</td>\n",
       "      <td>64</td>\n",
       "      <td>[wild_gorilla]</td>\n",
       "      <td>64</td>\n",
       "      <td>HR</td>\n",
       "      <td>HR_wild_gorilla</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>single_site_single_sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ASV_453</td>\n",
       "      <td>[wd.chi.GM.10.16s, wd.chi.GM.100.16s, wd.chi.G...</td>\n",
       "      <td>86</td>\n",
       "      <td>[wild_gorilla, wild_chimp]</td>\n",
       "      <td>86</td>\n",
       "      <td>MX</td>\n",
       "      <td>MX_2_wild_apes</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>single_site_single_sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ASV_932</td>\n",
       "      <td>[wd.bon.ML.a207.16s, wd.bon.ML.a208.16s, wd.bo...</td>\n",
       "      <td>48</td>\n",
       "      <td>[wild_bonobo]</td>\n",
       "      <td>48</td>\n",
       "      <td>HR</td>\n",
       "      <td>HR_wild_bonobo</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>single_site_single_sp</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ASV_1398</td>\n",
       "      <td>[wd.bon.LA.a187.16s, wd.bon.LA.a189.16s, wd.bo...</td>\n",
       "      <td>18</td>\n",
       "      <td>[wild_bonobo, wild_gorilla]</td>\n",
       "      <td>18</td>\n",
       "      <td>MX</td>\n",
       "      <td>MX_2_wild_apes</td>\n",
       "      <td>False</td>\n",
       "      <td>0</td>\n",
       "      <td>[]</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>[]</td>\n",
       "      <td>0</td>\n",
       "      <td>single_site_single_sp</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 28 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        ASV                                        sampleNames  sampleNum  \\\n",
       "0   ASV_617  [wd.gor.BI.a172.16s, wd.gor.BI.a173.16s, wd.go...        108   \n",
       "1   ASV_982  [wd.gor.BI.a175.16s, wd.gor.BI.a177.16s, wd.go...         64   \n",
       "2   ASV_453  [wd.chi.GM.10.16s, wd.chi.GM.100.16s, wd.chi.G...         86   \n",
       "3   ASV_932  [wd.bon.ML.a207.16s, wd.bon.ML.a208.16s, wd.bo...         48   \n",
       "4  ASV_1398  [wd.bon.LA.a187.16s, wd.bon.LA.a189.16s, wd.bo...         18   \n",
       "\n",
       "                HR_sampleTypes  HR_sampleNum HR_cat          HR_type  CP_pres  \\\n",
       "0               [wild_gorilla]           108     HR  HR_wild_gorilla    False   \n",
       "1               [wild_gorilla]            64     HR  HR_wild_gorilla    False   \n",
       "2   [wild_gorilla, wild_chimp]            86     MX   MX_2_wild_apes    False   \n",
       "3                [wild_bonobo]            48     HR   HR_wild_bonobo    False   \n",
       "4  [wild_bonobo, wild_gorilla]            18     MX   MX_2_wild_apes    False   \n",
       "\n",
       "   CP_sampleNum CP_sampleTypes  ... ATLZ_captive_orangutan  \\\n",
       "0             0             []  ...                    0.0   \n",
       "1             0             []  ...                    0.0   \n",
       "2             0             []  ...                    0.0   \n",
       "3             0             []  ...                    0.0   \n",
       "4             0             []  ...                    0.0   \n",
       "\n",
       "  HOUZ_captive_gorilla COLZ_captive_gorilla HOUZ_captive_orangutan  \\\n",
       "0                  0.0                  0.0                    0.0   \n",
       "1                  0.0                  0.0                    0.0   \n",
       "2                  0.0                  0.0                    0.0   \n",
       "3                  0.0                  0.0                    0.0   \n",
       "4                  0.0                  0.0                    0.0   \n",
       "\n",
       "  PC_captive_chimp  HOUZ_captive_chimp  COMZ_captive_gorilla  CP_sp_loc  \\\n",
       "0              0.0                 0.0                   0.0         []   \n",
       "1              0.0                 0.0                   0.0         []   \n",
       "2              0.0                 0.0                   0.0         []   \n",
       "3              0.0                 0.0                   0.0         []   \n",
       "4              0.0                 0.0                   0.0         []   \n",
       "\n",
       "   numEnclosure          multi_site_sp  \n",
       "0             0  single_site_single_sp  \n",
       "1             0  single_site_single_sp  \n",
       "2             0  single_site_single_sp  \n",
       "3             0  single_site_single_sp  \n",
       "4             0  single_site_single_sp  \n",
       "\n",
       "[5 rows x 28 columns]"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def multi_site_sp(cp_desc):\n",
    "    \"\"\"Takes in a list of sample site_Descriptions and designate ASVs \n",
    "    based on whether they are found across multiple locations or \n",
    "    a single location, observed in a single host species or multiple \n",
    "    host species\"\"\"\n",
    "    sites = list(set([x.split('_')[0] for x in cp_desc]))\n",
    "    sitesNum = len(sites)\n",
    "    sitesMulti = 'multi_site' if sitesNum > 1 else 'single_site'      \n",
    "    species = list(set([x.split('_')[2] for x in cp_desc]))\n",
    "    speciesNum = len(species)\n",
    "    speciessMulti = 'multi_sp' if speciesNum > 1 else 'single_sp' \n",
    "    return(sitesMulti + '_' + speciessMulti)\n",
    "\n",
    "metadata = pd.read_csv(metadata_file,sep='\\t',index_col=None)\n",
    "metadata['Description_site'] = metadata['site_code']+'_'+metadata['Description']\n",
    "sample_type_site_dict = dict(zip(metadata['X.SampleID'], metadata['Description_site']))\n",
    "description_df = asv_16s['sampleNames'].apply(lambda l: pd.Series(\n",
    "    [sample_type_site_dict[name] for name in l]).value_counts()) \n",
    "description_df = description_df.fillna(0) \n",
    "capt_desc = list(set(metadata['Description_site'][metadata['captivity_status']=='captive'])) \n",
    "description_df = description_df[capt_desc]\n",
    "description_df['CP_sp_loc'] = description_df.apply(lambda row: list(row.index[row>0]),axis=1)\n",
    "description_df['numEnclosure'] = description_df['CP_sp_loc'].apply(lambda x: len(x))\n",
    "description_df['multi_site_sp'] = description_df['CP_sp_loc'].apply(lambda x:  \n",
    "                                                                    multi_site_sp(x))\n",
    "asv_16s_description = asv_16s.merge(description_df,left_index=True,right_index=True)\n",
    "asv_16s_description.to_csv(f'{OUTDIR}/tables/16S_ASVs_summary.txt',sep='\\t',index=None)\n",
    "asv_16s_description.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python [conda env:ete]",
   "language": "python",
   "name": "conda-env-ete-py"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
